---
title: "Blastocyst_vs_Blastoid"
author: "Tara Urselmann"
date: "2024-03-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup

```{r Library, results='hide', echo = TRUE, error=FALSE, warning=FALSE, message=FALSE}
#loading of all necessary libraries
library(workflowr) 
library(here) 
library(Seurat) 
library(SeuratDisk) 
library(RColorBrewer) 
library(ggplot2) 
library(dplyr) 
library(scran) 
library(kableExtra)
library(progeny)
library(ggplot2)
library(tidyr)
library(readr)
library(pheatmap)
library(tibble)
library(ggrepel)
library(MAYA)
library(biomaRt)
library(biomartr)
library(seqinr)
library(limma)
library(edgeR)
library(org.Hs.eg.db)
library(decoupleR)
library(OmnipathR)
library(EnvStats)
library(EnhancedVolcano)
library(ggpmisc)
```

```{r Work directory, results='hide', echo = TRUE, error=FALSE, warning=FALSE, message=FALSE}
#wflow_git_config(user.name = "taraurselmann",user.email = "taraurselmann@gmail.com") 
wflow_start("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/", existing = TRUE)
set_here(path = "C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/")
```

## Functions

```{r limma pipeline}
#Write a function for analyzing this data set with limma pipeline to obtain DEGs
#Note: The input of DGElist is count, since that is not available I use the TPM data.
#used this thread as guideline: https://support.bioconductor.org/p/98820/
DEG_analysis <- function(df, group, loglimit, adj_p_value, output_file_name){
  group <- as.character(group)
  design <- model.matrix(~group)
  dge <- DGEList(counts = log(df+1))
  fit <- lmFit(as.matrix(dge), design)
  fit <- eBayes(fit, trend = T)
  table_dge <- topTable(fit, coef=ncol(design), number = length(rownames(df)))
  table_dge$diff <- "NO"
  # if log2Foldchange > loglimit and pvalue < p_value, set as "UP" 
  table_dge$diff[table_dge$logFC > loglimit & table_dge$adj.P.Val < adj_p_value] <- "UP"
  # if log2Foldchange < -loglimit and pvalue < p_value, set as "DOWN"
  table_dge$diff[table_dge$logFC < -loglimit & table_dge$adj.P.Val < adj_p_value] <- "DOWN"
  
  table_dge$genes <- rownames(table_dge)
  table_dge$delabel <- NA
  table_dge$delabel[table_dge$diff != "NO"] <- table_dge$genes[table_dge$diff != "NO"]

  table_dge_up <- subset(table_dge, diff == "UP")
  table_dge_down <- subset(table_dge, diff == "DOWN")
  
  write.table(table_dge, file.path("output", output_file_name))
  p <- ggplot(data = table_dge, aes(x= logFC, y=-log10(adj.P.Val), col= diff, label = delabel, show.legend = F))+
    NoLegend()+
    geom_point()+
    theme_minimal()+
    geom_text_repel(col = "black", na.rm = TRUE, hjust = 1)+
    geom_vline(xintercept=c(-loglimit, loglimit), col="red") +
    geom_hline(yintercept=-log10(adj_p_value), col="red") +
    scale_color_manual(values = c("black", "lightgrey", "black"), guide = "none")
  return(p)
}


```

### 

# Loading data

## Blastocyst data

```{r load data in directory}
#Get all separate files in directory, every file contains a sample
all_files <- list.files("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/GSE133200_RAW/", full.names=TRUE, recursive = T)
ldf <- lapply(all_files, read.csv)
all_files_list <- lapply(ldf, summary)
names(all_files_list) <- substr(all_files, 1, 1000000000)
```

```{r reference}
#This loop construction is used to open all files in directory
for (i in names(all_files_list)){
  if (i == names(all_files_list[1])){
  file1 <- read.csv(i, sep = "")
  TPM_df <- file1
  }
  else{
  file2 <- read.csv(i, sep = "")
  TPM_df <- full_join(TPM_df, file2, by = join_by(Gene))
  }
}

rownames(TPM_df) <- TPM_df[,1]
TPM_df <- TPM_df[-c(1)]
Blastocyst_data <- TPM_df


#The reference list is used to link the right name to the sample
reference_list <- readr::read_tsv("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_tpm/data/reference_list.txt")

cell_type_df <- TPM_df
for (i in 1:length(colnames(cell_type_df))){
  if (colnames(cell_type_df[i]) %in% reference_list$old){
    row_index <- which(reference_list == colnames(cell_type_df[i]))
    new_name <- reference_list$new[row_index]
    names(cell_type_df)[names(cell_type_df) == colnames(cell_type_df[i])] <- new_name}
    else {}
    
}

#Samples are deleted if not in reference file 
TPM_df2 <- TPM_df
col_delete <- c()
for (i in 1:length(colnames(cell_type_df))){
  if (grepl("hE", colnames(cell_type_df[i])) == T){
    col_delete <- append(col_delete, i)}}
ncol(cell_type_df)
TPM_df2 <- TPM_df2[,-col_delete]

cell_type_df <- cell_type_df[,-col_delete]
ncol(cell_type_df)
Blastocyst_data <- cell_type_df

```

```{r}
nrow(Blastocyst_data)
#Because this file contains ALIAS symbols the symbols need to be converted to official gene symbols to compare later
geneSymbols <- AnnotationDbi::mapIds(org.Hs.eg.db, keys=rownames(Blastocyst_data), column="SYMBOL", keytype="ALIAS", multiVals="first")
geneSymbols <- as.data.frame(geneSymbols)
Blastocyst_data$new_gene_names <- geneSymbols$geneSymbols
Blastocyst_data = Blastocyst_data[!duplicated(Blastocyst_data$new_gene_names),]
#If no official gene names are available. Exclude those from data set
Blastocyst_data <- na.omit(Blastocyst_data)
rownames(Blastocyst_data) <- Blastocyst_data$new_gene_names
Blastocyst_data <- subset(Blastocyst_data, select = -c(new_gene_names))
nrow(Blastocyst_data)
```

## Blastoid data

```{r Import Blastoid data}
#Import blastoid data of Kagawa
Blastoid_raw <- readRDS("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/blastoid.H9.okae_bts5__scRNAseq.unfiltered__counts.RDS")
Blastoid_meta <- read_tsv("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/blastoid.H9.okae_bts5__scRNAseq.unfiltered__metadata.tsv")

Blastoid_raw <- column_to_rownames(Blastoid_raw, "gene_id")
Blastoid_raw <- rownames_to_column(Blastoid_raw, var = "gene_id")

#Choose to delete rows with the same gene name, but different ensembl ids
Blastoid_raw[c("Ensembl_id", "gene_name")] <- stringr::str_split_fixed(Blastoid_raw$gene_id, '-', 2)
Blastoid_raw = Blastoid_raw[!duplicated(Blastoid_raw$gene_name),]
rownames(Blastoid_raw) <- Blastoid_raw$gene_name
nrow(Blastoid_raw)
#remove redundant columns
Blastoid_raw <- Blastoid_raw[, !names(Blastoid_raw) %in% c("gene_id", "gene_name", "Ensembl_id"), drop = F]

#Add meta data 
Blastoid_meta <- Blastoid_meta %>%
                  dplyr::mutate(Blastoid_meta, time = case_when(
                  grepl("24h",samplename) ~ "24h",
                  grepl("60h",samplename) ~ "60h",
                  grepl("96h",samplename) ~ "96h",
                  TRUE ~ gsub("-.*","",samplename)
             )) %>% data.frame()
rownames(Blastoid_meta) <- Blastoid_meta$sampleid
mtname="^ENSG[[:digit:]]+-MT-"

#Remove unviable cells. Use the same requirements used previously. Pretty standard. 
Blastoid_meta_filtered <- subset(Blastoid_meta, nFeature_RNA > 2000 & percent.mt <12.5)
ncol(Blastoid_raw)
Blastoid_raw <- Blastoid_raw[, rownames(Blastoid_meta_filtered)]
Blastoid_raw <- Blastoid_raw[rowSums(Blastoid_raw) >5, ]
#Retrieve the annotation provided by Cheyenne de Vriend
annotated_list <- LoadH5Seurat("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/BlastoidTutorial/data/SeuratProject_Blastoid_annotated.h5Seurat")
Blastoid_data <- Blastoid_raw
annotation <- as.data.frame(annotated_list@active.ident)
colnames(annotation) <- "cluster_name"
#Use a loop to annotate the right cells
col_delete <- c()
for (i in 1:length(colnames(Blastoid_data))){
  if (colnames(Blastoid_data[i]) %in% rownames(annotation)){
  row_index <- which(rownames(annotation) == colnames(Blastoid_data[i]))
  new_name <- as.character(annotation$cluster_name[row_index])
  names(Blastoid_data)[names(Blastoid_data) == colnames(Blastoid_data[i])] <- new_name}
  else{
  col_delete <- append(col_delete, i)
  }
}
Blastoid_data <- Blastoid_data[,-col_delete]
```

```{r}
Blastoid_data_new <- Blastoid_raw
feature_table_genes <- read.csv("output/Features_genes.csv")
feature_table_genes_2 <- as.data.frame(feature_table_genes$external_gene_name)
colnames(feature_table_genes_2) <- "gene_name"
#transcript length is in bps
feature_table_genes_2$length <- (feature_table_genes$transcript_length/1000)
list_of_fails <- list()
#create df for transformation
ensembl_id_lengths_blastoid <- Blastoid_data_new
lengths_vector <- setNames(feature_table_genes_2$length, feature_table_genes_2$gene_name)
# Identify common IDs
common_ids <- intersect(rownames(ensembl_id_lengths_blastoid), names(lengths_vector))
# Subset data frame to common IDs and perform the transformation
ensembl_id_lengths_blastoid[common_ids,] <- ensembl_id_lengths_blastoid[common_ids,] / lengths_vector[common_ids]
ensembl_id_lengths_blastoid <- ensembl_id_lengths_blastoid[common_ids,]
Blastoid_data_new <- ensembl_id_lengths_blastoid
for (i in 1:length(colnames(Blastoid_data_new))){
  sum_of_sample <- sum(Blastoid_data_new[,i])
  factor <- sum_of_sample/1000000
  Blastoid_data_new[,i] <- (Blastoid_data_new[,i]/factor)
  
}



Blastoid_data_new <- CreateSeuratObject(counts = Blastoid_data_new, meta.data = Blastoid_meta_filtered)
Blastoid_data_new <- SetAssayData(object=Blastoid_data_new, slot="data", new.data = log1p(Blastoid_data_new@assays$RNA@layers$counts)) 
Blastoid_data_new <- FindVariableFeatures(object = Blastoid_data_new, nfeatures = 3000)
Blastoid_data_new <- ScaleData(Blastoid_data_new)
Blastoid_data_new <- RunPCA(Blastoid_data_new)
Blastoid_data_new <- FindNeighbors(Blastoid_data_new)
Blastoid_data_new <- FindClusters(Blastoid_data_new, resolution = 1.5, verbose=FALSE)
Blastoid_data_new <- RunUMAP(Blastoid_data_new, dims = 1:20, n.components = 3L, min.dist = 0.5, verbose=FALSE)
Blastoid_data_new$time <- Blastoid_data_new@meta.data$time
DimPlot(Blastoid_data_new, reduction = "umap", group.by = "time")
DimPlot(Blastoid_data_new, reduction = "umap")
features <- c("GATA2", "GATA3", "POU5F1", "KLF17", "GATA4", "SOX17")
DotPlot(Blastoid_data_new, features = features) + RotatedAxis()+scale_colour_gradient2(low="blue", mid="white", high="red")
```

```{r Gene length information}
#Get feature lengths/ ensembl id
# command for downgrading
#devtools::install_version("dbplyr", version = "2.3.4")

#biomart <- 'ENSEMBL_MART_ENSEMBL'
#ensembl106 <- useEnsembl(biomart = 'genes', 
#                         dataset = 'hsapiens_gene_ensembl',
#                         version = 106)

#mart = useEnsembl("ENSEMBL_MART_ENSEMBL")
#mart=useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl")
#attributes = listAttributes(ensembl106) # Overview of all available features
#write.csv(attributes, "output/Available_attributes_biomaRt.csv")
#ensembl106 <- mart
#attributes <-  read.csv("output/Available_attributes_biomaRt.csv")

#features <- getBM(attributes = c#("ensembl_transcript_id"3,"transcript_is_canonical",
#                                 "ensembl_gene_id", "external_gene_name", 
#                                 "chromosome_name","percentage_gene_gc_content"),
#                  mart = ensembl106)

#transcr_lengths <- getBM(attributes = c#("ensembl_transcript_id",
#                                        "transcript_length", "cds_length"),
#                         filters = "ensembl_transcript_id",
#                         values = features$ensembl_transcript_id,
#                         mart = ensembl106)
#feature_table <- merge(features, transcr_lengths, by = "ensembl_transcript_id")
#feature_table_genes <- feature_table[!is.na(feature_table$transcript_is_canonical),]

#write.csv(feature_table_genes, "output/Features_genes.csv")
feature_table_genes <- read.csv("output/Features_genes.csv")
feature_table_genes_2 <- as.data.frame(feature_table_genes$external_gene_name)
colnames(feature_table_genes_2) <- "gene_name"
#transcript length is in bps
feature_table_genes_2$length <- (feature_table_genes$transcript_length/1000)
list_of_fails <- list()
#create df for transformation
ensembl_id_lengths_blastoid <- Blastoid_data
```

## RPK

```{r}
#Create RPK, so reads per kilobase
# Create a named vector of lengths for quick lookup
lengths_vector <- setNames(feature_table_genes_2$length, feature_table_genes_2$gene_name)
# Identify common IDs
common_ids <- intersect(rownames(ensembl_id_lengths_blastoid), names(lengths_vector))
# Subset data frame to common IDs and perform the transformation
ensembl_id_lengths_blastoid[common_ids,] <- ensembl_id_lengths_blastoid[common_ids,] / lengths_vector[common_ids]
ensembl_id_lengths_blastoid <- ensembl_id_lengths_blastoid[common_ids,]
Blastoid_data <- ensembl_id_lengths_blastoid

```

## TPM

```{r}
#Normalize data for sequencing depth
for (i in 1:length(colnames(Blastoid_data))){
  sum_of_sample <- sum(Blastoid_data[,i])
  factor <- sum_of_sample/1000000
  Blastoid_data[,i] <- (Blastoid_data[,i]/factor)
  
}


Blastoid_tpm <- Blastoid_data
Blastocyst_tpm <- Blastocyst_data
#Histogram to check for distribution in both data sets
#Histogram Blastoid
hist_blastoid_df <- as.data.frame(rowSums(Blastoid_data))
hist_blastoid <- ggplot(hist_blastoid_df, aes(x =`rowSums(Blastoid_data)`)) +
  geom_histogram()+
  scale_y_continuous(trans = "log2")

#Histogram Blastocyst
hist_blastocyst_df <- as.data.frame(rowSums(Blastocyst_data))
hist_blastocyst <- ggplot(hist_blastocyst_df, aes(x = `rowSums(Blastocyst_data)`)) +
  geom_histogram()+
  scale_y_continuous(trans = "log2")

  
hist_blastocyst + hist_blastoid
```

```{r}
#combining the data set is required to compare later on
#It is important that the data sets compare only the available genes as the Blastocyst data already excluded some genes --> inner_join
#Add column of rownames in both data sets to join them later on
Blastocyst_tpm$gene <- rownames(Blastocyst_tpm)
Blastoid_tpm$gene <- rownames(Blastoid_tpm)
Blast_tpm <- inner_join(Blastocyst_tpm, Blastoid_tpm)
Blast_tpm <- column_to_rownames(Blast_tpm, var = "gene")
Blast_tpm <- Blast_tpm
#Create a subset of both data set containing only trophectoderm

#Blastoid data only 60 and 96h as they form a cluster together
#Blastocyst data only S2 as coculturing with ISK cells could potentially alter results
col_delete <- c()
for (i in 1:length(colnames(Blast_tpm))){
  if (grepl("PTE 96h", colnames(Blast_tpm[i])) ==F & grepl("TE 96h", colnames(Blast_tpm[i])) ==F & grepl("S2.MTE", colnames(Blast_tpm[i])) == F & grepl("S2.PTE", colnames(Blast_tpm[i])) == F){
    col_delete <- append(col_delete, i)}
}
Blast_tpm_te <- Blast_tpm[,-col_delete]


```

```{r}
VIRMA_RIP <- readxl::read_excel(path = "C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/VIRMA_RIP.xlsx")
ensembl <- useMart("ensembl")
geneSymbols <- AnnotationDbi::mapIds(org.Hs.eg.db, keys=VIRMA_RIP$`RIP and m6A-seq overlap genes`, column="SYMBOL", keytype="ALIAS", multiVals="first")
geneSymbols <- as.data.frame(geneSymbols)
VIRMA_RIP$new_gene_names <- geneSymbols$geneSymbols
VIRMA_RIP = VIRMA_RIP[!duplicated(VIRMA_RIP$new_gene_names),]
#If no official gene names are available. Exclude those from data set
VIRMA_RIP <- na.omit(VIRMA_RIP)
rownames(VIRMA_RIP) <- VIRMA_RIP$new_gene_names
VIRMA_RIP <- rownames(VIRMA_RIP)
VIRMA_RIP <- unlist(VIRMA_RIP)
# Select the appropriate dataset
ensembl <- useDataset("hsapiens_gene_ensembl", mart = ensembl) # For human species, change to appropriate species if needed

# Retrieve 3'UTR sequences for the genes in your dataframe
get_3utr_sequences <- function(gene_names) {
  utr <- getBM(attributes = c("hgnc_symbol", "3utr"),
               filters = "external_gene_name",
               values = gene_names,
               mart = ensembl)
  return(utr)
}

# Replace "VIRMA_RIP$Gene_Name_Column" with the actual column name in your dataframe that contains gene names
gene_names <- VIRMA_RIP
utr_sequences <- get_3utr_sequences(gene_names)

# View the results
head(utr_sequences)

ensembl106 <- useEnsembl(biomart = 'genes', 
                        dataset = 'hsapiens_gene_ensembl',
                        version = 106)

# command for downgrading
devtools::install_version("dbplyr", version = "2.3.4")


# Extract attributes:
# ID variables: transcript_is_canonical,external_gene_name, ensembl_gene_id,
#   ensembl_transcript_id, chromosome_name
# Sequences: 5utr and 3utr, coding
# Length: transcript_length, cds_length
# GC content: percentage_gene_gc_content, per sequence section (calculate)



# Note: you cannot extract all features at once

# Start building by transcript ID, extract all available 
features <- getBM(attributes = c("ensembl_transcript_id","transcript_is_canonical",
                                 "ensembl_gene_id", "external_gene_name", 
                                 "chromosome_name","percentage_gene_gc_content"),
                  mart = ensembl106)

utr3 <- getBM(attributes = c("ensembl_transcript_id",
                               "3utr"),
                filters = "ensembl_transcript_id",
                values = features$ensembl_transcript_id,
                mart = ensembl106)

# Merge by transcript ID
feature_table2 <- merge(features, utr3, by = "ensembl_transcript_id")
write.csv(feature_table2, "output/table_3utr")
feature_table2 <- read.csv("output/table_3utr")
```

```{r}
VIRMA_binding_RNAs <- readxl::read_xlsx("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/mRNA_VIRMA.xlsx")
rownames(VIRMA_binding_RNAs) <- VIRMA_binding_RNAs$mRNA
#Because this file contains ALIAS symbols the symbols need to be converted to official gene symbols to compare later
geneSymbols <- AnnotationDbi::mapIds(org.Hs.eg.db, keys=rownames(VIRMA_binding_RNAs), column="SYMBOL", keytype="ALIAS", multiVals="first")
geneSymbols <- as.data.frame(geneSymbols)
VIRMA_binding_RNAs$new_gene_names <- geneSymbols$geneSymbols
VIRMA_binding_RNAs = VIRMA_binding_RNAs[!duplicated(VIRMA_binding_RNAs$new_gene_names),]
#If no official gene names are available. Exclude those from data set
VIRMA_binding_RNAs <- na.omit(VIRMA_binding_RNAs)
rownames(VIRMA_binding_RNAs) <- VIRMA_binding_RNAs$new_gene_names
VIRMA_binding_RNAs <- rownames(VIRMA_binding_RNAs)
VIRMA_binding_RNAs <- unlist(VIRMA_binding_RNAs)



DoHeatmap(Blast_integrated, features = VIRMA_binding_RNAs, group.by = "cell_layers", size = 3, slot = "data")+theme(text = element_text(size = 5))+NoLegend()

write.csv(VIRMA_binding_RNAs, "output/VIRMA_RNA_correct.csv") 
```

### Randomize and cluster

```{r}

new_name = list()
for (i in 1:length(colnames(Blast_tpm_te))){
  if (grepl("S2", colnames(Blast_tpm_te)[i]) == T){
    new_name <- append(new_name, "Blastocyst")
  }else{
    new_name <- append(new_name, "Blastoid")
  }
}
Blast_regression <- Blast_tpm_te
colnames(Blast_regression) <- new_name
Blastoid_regression <- Blast_regression[ , grep("^Blastoid"   , names(Blast_regression))]
Blastocyst_regression <-  Blast_regression[ , grep("^Blastocyst"   , names(Blast_regression))]

regression_df <- data.frame(Blastoid = rowMeans(log(Blastoid_regression+1)), Blastocyst = rowMeans(log(Blastocyst_regression+1)))
rownames(regression_df) <- rownames(Blast_regression)
regression_df$delabel <- rownames(regression_df)

regression_te <- ggplot(regression_df, aes(x= Blastoid, y = Blastocyst))+
  geom_point(size = 1, show.legend = T)+
  stat_summary(fun.data= regression_df) + 
  geom_smooth(method='lm')+
  scale_x_continuous(expand = c(0, 0)) + 
  scale_y_continuous(expand = c(0, 0), limits = c(0, 10))+
  stat_poly_eq()+
  theme_bw()+
  xlab("Blastoid gene expression")+
  ylab("Blastocyst gene expression")+
  ggtitle("Regression Trophectoderm")
regression_te  

```

```{r EPI/ICM}
col_delete <- c()
for (i in 1:length(colnames(Blast_tpm))){
  if (grepl("EPI", colnames(Blast_tpm[i])) ==F & grepl("S2.EPI", colnames(Blast_tpm[i])) == F & grepl("ICM", colnames(Blast_tpm[i])) == F | grepl("AF", colnames(Blast_tpm[i]) ==F)){
    col_delete <- append(col_delete, i)}
}
Blast_tpm_icm <- Blast_tpm[,-col_delete]

new_name = list()
for (i in 1:length(colnames(Blast_tpm_icm))){
  if (grepl("S2", colnames(Blast_tpm_icm)[i]) == T | grepl("S1", colnames(Blast_tpm_icm)[i]) == T | grepl("S3", colnames(Blast_tpm_icm)[i]) == T){
    new_name <- append(new_name, "Blastocyst")
  }else{
    new_name <- append(new_name, "Blastoid")
  }
}

Blast_regression_icm <- Blast_tpm_icm
colnames(Blast_regression_icm) <- new_name
Blastoid_regression_icm <- Blast_regression_icm[ , grep("^Blastoid"   , names(Blast_regression_icm))]
Blastocyst_regression_icm <-  Blast_regression_icm[ , grep("^Blastocyst"   , names(Blast_regression_icm))]

regression_df_icm <- data.frame(Blastoid = rowMeans(log(Blastoid_regression_icm+1)), Blastocyst = rowMeans(log(Blastocyst_regression_icm+1)))
rownames(regression_df_icm) <- rownames(Blast_regression_icm)

regression_icm <- ggplot(regression_df_icm, aes(x= regression_df_icm$Blastoid, y = regression_df_icm$Blastocyst))+
  geom_point(size = 1)+
  stat_summary(fun.data= regression_df_icm) + 
  geom_smooth(method='lm')+
  scale_x_continuous(expand = c(0, 0)) + 
  scale_y_continuous(expand = c(0, 0), limits = c(0, 10))+
  stat_poly_eq()+
  theme_bw()+
  xlab("Blastoid gene expression")+
  ylab("Blastocyst gene expression")+
  ggtitle("Regression Inner cell mass")


regression_icm
```

```{r}
col_delete <- c()
for (i in 1:length(colnames(Blast_tpm))){
  if (grepl("PRE", colnames(Blast_tpm[i])) ==F & grepl("PE", colnames(Blast_tpm[i])) == F){
    col_delete <- append(col_delete, i)}
}
Blast_tpm_pre <- Blast_tpm[,-col_delete]

new_name = list()
for (i in 1:length(colnames(Blast_tpm_pre))){
  if (grepl("PE", colnames(Blast_tpm_pre)[i]) == T){
    new_name <- append(new_name, "Blastocyst")
  }else{
    new_name <- append(new_name, "Blastoid")
  }
}

Blast_regression_pre <- Blast_tpm_pre
colnames(Blast_regression_pre) <- new_name
Blastoid_regression_pre <- Blast_regression_pre[ , grep("^Blastoid"   , names(Blast_regression_pre))]
Blastocyst_regression_pre <-  Blast_regression_pre[ , grep("^Blastocyst"   , names(Blast_regression_pre))]

regression_df_pre <- data.frame(Blastoid = rowMeans(log(Blastoid_regression_pre+1)), Blastocyst = rowMeans(log(Blastocyst_regression_pre+1)))
rownames(regression_df_pre) <- rownames(Blast_regression_pre)

regression_pre <- ggplot(regression_df_pre, aes(x= regression_df_pre$Blastoid, y = regression_df_pre$Blastocyst))+
  geom_point(size = 1)+
  stat_summary(fun.data= regression_df_pre) + 
  geom_smooth(method='lm')+
  scale_x_continuous(expand = c(0, 0)) + 
  scale_y_continuous(expand = c(0, 0), limits = c(0, 10))+
  stat_poly_eq()+
  theme_bw()+
  xlab("Blastoid gene expression")+
  ylab("Blastocyst gene expression")+
  ggtitle("Regression Primitive endoderm")

regression_pre

```

```{r warning=FALSE, fig.width = 12}
regression_te + regression_icm + regression_pre
```

```{r}
hist_te_blastoid <- qqPlot(regression_df$Blastoid)
hist_te_blastocyst <- qqPlot(regression_df$Blastocyst)
hist_icm_blastoid <- qqPlot(regression_df_icm$Blastoid)
hist_icm_blastocyst <- qqPlot(regression_df_icm$Blastocyst)
hist_pre_blastoid <- qqPlot(regression_df_pre$Blastoid)
hist_pre_blastocyst <- qqPlot(regression_df_pre$Blastocyst)
```

## Limma

```{r 10 groups of 10}

#To minimize noise the blastocyst and blastoid will be separated by TE only. Subsequently the columns will be randomized to minimize bias, and than they will be clustered in ~10 groups. 
#Obtain only TE from Blastocyst
col_delete <- c()
for (i in 1:length(colnames(Blastocyst_tpm))){
  if (grepl("S2.MTE", colnames(Blastocyst_tpm[i])) == F & grepl("S2.PTE", colnames(Blastocyst_tpm[i])) == F){
    col_delete <- append(col_delete, i)}
}
Blastocyst_tpm_te <- Blastocyst_tpm[,-col_delete]


#Randomize column order
v <- c(1:length(Blastocyst_tpm_te))
set.seed(001)
v <- sample(v)
Blastocyst_tpm_random <- Blastocyst_tpm_te[v]
#Divide in to ~10 groups by making 10 groups

Blastocyst_tpm_10 <- as.data.frame(rowMeans(Blastocyst_tpm_random[0:47]))
colnames(Blastocyst_tpm_10) <- c("1a")
Blastocyst_tpm_10$'2a' <- rowMeans(Blastocyst_tpm_random[48:94])
Blastocyst_tpm_10$'3a' <- rowMeans(Blastocyst_tpm_random[95:141])
Blastocyst_tpm_10$'4a' <- rowMeans(Blastocyst_tpm_random[142:188])
Blastocyst_tpm_10$'5a' <- rowMeans(Blastocyst_tpm_random[189:235])
Blastocyst_tpm_10$'6a' <- rowMeans(Blastocyst_tpm_random[236:282])
Blastocyst_tpm_10$'7a' <- rowMeans(Blastocyst_tpm_random[283:329])
Blastocyst_tpm_10$'8a' <- rowMeans(Blastocyst_tpm_random[330:376])
Blastocyst_tpm_10$'9a' <- rowMeans(Blastocyst_tpm_random[377:423])
Blastocyst_tpm_10$'10a' <- rowMeans(Blastocyst_tpm_random[424:470])
Blastocyst_tpm_10$"gene" <- rownames(Blastocyst_tpm_10)

#Obtain only TE from Blastoid
col_delete <- c()
for (i in 1:length(colnames(Blastoid_tpm))){
  if (grepl("PTE 96h", colnames(Blastoid_tpm[i])) ==F & grepl("TE 96h", colnames(Blastoid_tpm[i])) ==F){
    col_delete <- append(col_delete, i)}
}
Blastoid_tpm_te <- Blastoid_tpm[,-col_delete]


v <- c(1:length(Blastoid_tpm_te))
v <- sample(v)
Blastoid_tpm_random <- Blastoid_tpm_te[v]
Blastoid_tpm_10 <- as.data.frame(rowMeans(Blastoid_tpm_random[1:37]))
colnames(Blastoid_tpm_10) <- c("1b")
Blastoid_tpm_10$'2b' <- rowMeans(Blastoid_tpm_random[38:74])
Blastoid_tpm_10$'3b' <- rowMeans(Blastoid_tpm_random[75:111])
Blastoid_tpm_10$'4b' <- rowMeans(Blastoid_tpm_random[112:148])
Blastoid_tpm_10$'5b' <- rowMeans(Blastoid_tpm_random[149:185])
Blastoid_tpm_10$'6b' <- rowMeans(Blastoid_tpm_random[186:222])
Blastoid_tpm_10$'7b' <- rowMeans(Blastoid_tpm_random[223:259])
Blastoid_tpm_10$'8b' <- rowMeans(Blastoid_tpm_random[260:296])
Blastoid_tpm_10$'9b' <- rowMeans(Blastoid_tpm_random[297:333])
Blastoid_tpm_10$'10b' <- rowMeans(Blastoid_tpm_random[334:375])
Blastoid_tpm_10$"gene" <- rownames(Blastoid_tpm_10)

Blast_tpm_te_10 <- inner_join(Blastocyst_tpm_10, Blastoid_tpm_10)
Blast_tpm_te_10 <- column_to_rownames(Blast_tpm_te_10, var = "gene")
Blast_tpm_te_10 <- log((Blast_tpm_te_10+1))
group <- c(rep(c("Blastocyst"), times = 10 ), rep(c("Blastoid"), times =10))
design <- model.matrix(~group)

DEG_plot_te <- DEG_analysis(Blast_tpm_te_10, group =group, loglimit = 1.0, adj_p_value = 0.05, output_file_name = "DEG_TE")
DEG_plot_te
```

```{r 10 groups of 10, EPIBLAST}

#To minimize noise the blastocyst and blastoid will be separated by TE only. Subsequently the columns will be randomized to minimize bias, and than they will be clustered in ~10 groups. 
#Obtain only TE from Blastocyst
col_delete <- c()
for (i in 1:length(colnames(Blastocyst_tpm))){
  if (grepl("ICM", colnames(Blastocyst_tpm[i])) == F & grepl("EPI", colnames(Blastocyst_tpm[i])) == F){
    col_delete <- append(col_delete, i)}
}
Blastocyst_tpm_icm <- Blastocyst_tpm[,-col_delete]


#Randomize column order
v <- c(1:length(Blastocyst_tpm_icm))
set.seed(001)
v <- sample(v)
Blastocyst_tpm_icm_random <- Blastocyst_tpm_icm[v]
#Divide in to ~10 groups by making 10 groups

Blastocyst_tpm_icm_10 <- as.data.frame(rowMeans(Blastocyst_tpm_icm_random[1:8]))
Blastocyst_tpm_icm_10$'2a' <- rowMeans(Blastocyst_tpm_icm_random[9:16])
Blastocyst_tpm_icm_10$'3a' <- rowMeans(Blastocyst_tpm_icm_random[17:24])
Blastocyst_tpm_icm_10$'4a' <- rowMeans(Blastocyst_tpm_icm_random[25:32])
Blastocyst_tpm_icm_10$'5a' <- rowMeans(Blastocyst_tpm_icm_random[33:40])
Blastocyst_tpm_icm_10$'6a' <- rowMeans(Blastocyst_tpm_icm_random[41:48])
Blastocyst_tpm_icm_10$'7a' <- rowMeans(Blastocyst_tpm_icm_random[49:56])
Blastocyst_tpm_icm_10$'8a' <- rowMeans(Blastocyst_tpm_icm_random[57:64])
Blastocyst_tpm_icm_10$'9a' <- rowMeans(Blastocyst_tpm_icm_random[65:72])
Blastocyst_tpm_icm_10$"gene" <- rownames(Blastocyst_tpm_icm_10)

#Obtain only TE from Blastoid
col_delete <- c()
for (i in 1:length(colnames(Blastoid_tpm))){
  if (grepl("EPI", colnames(Blastoid_tpm[i])) ==F & grepl("ICM", colnames(Blastoid_tpm[i])) ==F){
    col_delete <- append(col_delete, i)}
}
Blastoid_tpm_icm <- Blastoid_tpm[,-col_delete]


v <- c(1:length(Blastoid_tpm_icm))
v <- sample(v)
Blastoid_tpm_icm_random <- Blastoid_tpm_icm[v]
Blastoid_tpm_icm_10 <- as.data.frame(rowMeans(Blastoid_tpm_icm_random[1:55]))
Blastoid_tpm_icm_10$'2b' <- rowMeans(Blastoid_tpm_icm_random[56:110])
Blastoid_tpm_icm_10$'3b' <- rowMeans(Blastoid_tpm_icm_random[111:165])
Blastoid_tpm_icm_10$'4b' <- rowMeans(Blastoid_tpm_icm_random[166:220])
Blastoid_tpm_icm_10$'5b' <- rowMeans(Blastoid_tpm_icm_random[221:275])
Blastoid_tpm_icm_10$'6b' <- rowMeans(Blastoid_tpm_icm_random[276:330])
Blastoid_tpm_icm_10$'7b' <- rowMeans(Blastoid_tpm_icm_random[331:385])
Blastoid_tpm_icm_10$'8b' <- rowMeans(Blastoid_tpm_icm_random[386:440])
Blastoid_tpm_icm_10$'9b' <- rowMeans(Blastoid_tpm_icm_random[441:495])
Blastoid_tpm_icm_10$'10b' <- rowMeans(Blastoid_tpm_icm_random[496:550])
Blastoid_tpm_icm_10$"gene" <- rownames(Blastoid_tpm_icm_10)

Blast_tpm_icm_10 <- inner_join(Blastocyst_tpm_icm_10, Blastoid_tpm_icm_10)
Blast_tpm_icm_10 <- column_to_rownames(Blast_tpm_icm_10, var = "gene")
Blast_tpm_icm_10 <- log((Blast_tpm_icm_10+1))
group <- c(rep(c("Blastocyst"), times = 9 ), rep(c("Blastoid"), times =10))
design <- model.matrix(~group)

DEG_plot_icm <- DEG_analysis(Blast_tpm_icm_10, group =group, loglimit = 1.0, adj_p_value = 0.05, output_file_name = "DEG_ICM")
DEG_plot_icm
```

```{r 10 groups of 10, PrE}

#To minimize noise the blastocyst and blastoid will be separated by TE only. Subsequently the columns will be randomized to minimize bias, and than they will be clustered in ~10 groups. 
#Obtain only TE from Blastocyst
col_delete <- c()
for (i in 1:length(colnames(Blastocyst_tpm))){
  if (grepl("PE", colnames(Blastocyst_tpm[i])) == F){
    col_delete <- append(col_delete, i)}
}
Blastocyst_tpm_pre <- Blastocyst_tpm[,-col_delete]


#Randomize column order
v <- c(1:length(Blastocyst_tpm_pre))
set.seed(001)
v <- sample(v)
Blastocyst_tpm_pre_random <- Blastocyst_tpm_pre[v]
#Divide in to ~10 groups by making 10 groups

Blastocyst_tpm_pre_10 <- as.data.frame(rowMeans(Blastocyst_tpm_pre_random[1:9]))
Blastocyst_tpm_pre_10$'2a' <- rowMeans(Blastocyst_tpm_pre_random[10:18])
Blastocyst_tpm_pre_10$'3a' <- rowMeans(Blastocyst_tpm_pre_random[19:27])
Blastocyst_tpm_pre_10$'4a' <- rowMeans(Blastocyst_tpm_pre_random[28:36])
Blastocyst_tpm_pre_10$'5a' <- rowMeans(Blastocyst_tpm_pre_random[37:45])
Blastocyst_tpm_pre_10$"gene" <- rownames(Blastocyst_tpm_pre_10)

#Obtain only TE from Blastoid
col_delete <- c()
for (i in 1:length(colnames(Blastoid_tpm))){
  if (grepl("PRE", colnames(Blastoid_tpm[i])) ==F & grepl("PE", colnames(Blastoid_tpm[i])) ==F){
    col_delete <- append(col_delete, i)}
}
Blastoid_tpm_pre <- Blastoid_tpm[,-col_delete]


v <- c(1:length(Blastoid_tpm_pre))
v <- sample(v)
Blastoid_tpm_pre_random <- Blastoid_tpm_pre[v]
Blastoid_tpm_pre_10 <- as.data.frame(rowMeans(Blastoid_tpm_pre_random[1:26]))
Blastoid_tpm_pre_10$'2b' <- rowMeans(Blastoid_tpm_pre_random[27:52])
Blastoid_tpm_pre_10$'3b' <- rowMeans(Blastoid_tpm_pre_random[53:78])
Blastoid_tpm_pre_10$'4b' <- rowMeans(Blastoid_tpm_pre_random[79:104])
Blastoid_tpm_pre_10$'5b' <- rowMeans(Blastoid_tpm_pre_random[105:130])
Blastoid_tpm_pre_10$'6b' <- rowMeans(Blastoid_tpm_pre_random[131:156])
Blastoid_tpm_pre_10$'7b' <- rowMeans(Blastoid_tpm_pre_random[157:182])
Blastoid_tpm_pre_10$'8b' <- rowMeans(Blastoid_tpm_pre_random[183:208])
Blastoid_tpm_pre_10$'9b' <- rowMeans(Blastoid_tpm_pre_random[209:234])
Blastoid_tpm_pre_10$'10b' <- rowMeans(Blastoid_tpm_pre_random[235:260])
Blastoid_tpm_pre_10$"gene" <- rownames(Blastoid_tpm_pre_10)

Blast_tpm_pre_10 <- inner_join(Blastocyst_tpm_pre_10, Blastoid_tpm_pre_10)
Blast_tpm_pre_10 <- column_to_rownames(Blast_tpm_pre_10, var = "gene")
Blast_tpm_pre_10 <- log((Blast_tpm_pre_10+1))
group <- c(rep(c("Blastocyst"), times = 5 ), rep(c("Blastoid"), times =10))
design <- model.matrix(~group)

DEG_plot_pre <- DEG_analysis(Blast_tpm_pre_10, group =group, loglimit = 1.0, adj_p_value = 0.05, output_file_name = "DEG_PRE")
DEG_plot_pre
```

```{r, fig.width = 12, fig.height= 5}
regression_te + regression_icm + regression_pre + DEG_plot_te + DEG_plot_icm + DEG_plot_pre
```

```{r}
test_te_blastocyst <- subset(Blastocyst_tpm_10, rowMeans(Blastocyst_tpm_te[1:10]) > 0)
test_te_blastoid <- subset(Blastoid_tpm_10, rowMeans(Blastoid_tpm_te[1:10]) > 0)
test_te_blast <- inner_join(test_te_blastocyst, test_te_blastoid)
test_te_blast <- column_to_rownames(test_te_blast, var = "gene")
test_te_blast <- log((test_te_blast+1))
group <- c(rep(c("Blastocyst"), times = 10 ), rep(c("Blastoid"), times =10))
DEG_analysis(test_te_blast, group = group, loglimit = 1, adj_p_value = 0.05, output_file_name = "DEG_0")
```

```{r}
Blastocyst_tpm_all_1 <- Blastocyst_tpm_icm_random
Blastocyst_tpm_all_1$"gene" <- rownames(Blastocyst_tpm_icm_random)
Blastocyst_tpm_all_2 <- Blastocyst_tpm_pre_random
Blastocyst_tpm_all_2$"gene" <- rownames(Blastocyst_tpm_pre_random)
Blastocyst_tpm_all_3 <- Blastocyst_tpm_random
Blastocyst_tpm_all_3$"gene" <- rownames(Blastocyst_tpm_random)
Blastocyst_tpm_all <- inner_join(Blastocyst_tpm_all_1, Blastocyst_tpm_all_2)
Blastocyst_tpm_all <- inner_join(Blastocyst_tpm_all, Blastocyst_tpm_all_3)
Blastocyst_tpm_all <- column_to_rownames(Blastocyst_tpm_all, var = "gene")

set.seed(001)
v <- c(1:length(Blastocyst_tpm_all))
v <- sample(v)
Blastocyst_tpm_all <- Blastocyst_tpm_all[v]
Blastocyst_tpm_all_10 <- as.data.frame(rowMeans(Blastocyst_tpm_all[1:59]))
Blastocyst_tpm_all_10$'2a' <- rowMeans(Blastocyst_tpm_all[60:118])
Blastocyst_tpm_all_10$'3a' <- rowMeans(Blastocyst_tpm_all[119:177])
Blastocyst_tpm_all_10$'4a' <- rowMeans(Blastocyst_tpm_all[178:236])
Blastocyst_tpm_all_10$'5a' <- rowMeans(Blastocyst_tpm_all[237:295])
Blastocyst_tpm_all_10$'6a' <- rowMeans(Blastocyst_tpm_all[296:354])
Blastocyst_tpm_all_10$'7a' <- rowMeans(Blastocyst_tpm_all[355:413])
Blastocyst_tpm_all_10$'8a' <- rowMeans(Blastocyst_tpm_all[414:472])
Blastocyst_tpm_all_10$'9a' <- rowMeans(Blastocyst_tpm_all[473:531])
Blastocyst_tpm_all_10$'10a' <- rowMeans(Blastocyst_tpm_all[532:590])
Blastocyst_tpm_all_10$"gene" <- rownames(Blastocyst_tpm_all)


Blastoids_tpm_all_1 <- Blastoid_tpm_icm_random
Blastoids_tpm_all_1$"gene" <- rownames(Blastoid_tpm_icm_random)
Blastoids_tpm_all_2 <- Blastoid_tpm_pre_random
Blastoids_tpm_all_2$"gene" <- rownames(Blastoid_tpm_pre_random)
Blastoids_tpm_all_3 <- Blastoid_tpm_random
Blastoids_tpm_all_3$"gene" <- rownames(Blastoid_tpm_random)
Blastoids_tpm_all <- inner_join(Blastoids_tpm_all_1, Blastoids_tpm_all_2)
Blastoids_tpm_all <- inner_join(Blastoids_tpm_all, Blastoids_tpm_all_3)
Blastoids_tpm_all <- column_to_rownames(Blastoids_tpm_all, var = "gene")
v <- c(1:length(Blastoids_tpm_all))

v <- sample(v)
Blastoids_tpm_all <- Blastoids_tpm_all[v]
#Obtain only TE from Blastoid
Blastoid_tpm_all_10 <- as.data.frame(rowMeans(Blastoids_tpm_all[1:119]))
Blastoid_tpm_all_10$'2b' <- rowMeans(Blastoids_tpm_all[120:238])
Blastoid_tpm_all_10$'3b' <- rowMeans(Blastoids_tpm_all[239:357])
Blastoid_tpm_all_10$'4b' <- rowMeans(Blastoids_tpm_all[358:476])
Blastoid_tpm_all_10$'5b' <- rowMeans(Blastoids_tpm_all[477:595])
Blastoid_tpm_all_10$'6b' <- rowMeans(Blastoids_tpm_all[596:714])
Blastoid_tpm_all_10$'7b' <- rowMeans(Blastoids_tpm_all[715:833])
Blastoid_tpm_all_10$'8b' <- rowMeans(Blastoids_tpm_all[834:952])
Blastoid_tpm_all_10$'9b' <- rowMeans(Blastoids_tpm_all[953:1071])
Blastoid_tpm_all_10$'10b' <- rowMeans(Blastoids_tpm_all[1072:1190])
Blastoid_tpm_all_10$"gene" <- rownames(Blastoids_tpm_all)


Blast_tpm_all_10 <- inner_join(Blastocyst_tpm_all_10, Blastoid_tpm_all_10)
Blast_tpm_all_10 <- column_to_rownames(Blast_tpm_all_10, var = "gene")
Blast_tpm_all_10 <- log((Blast_tpm_all_10+1))
group <- c(rep(c("Blastocyst"), times = 10 ), rep(c("Blastoid"), times =10))
design <- model.matrix(~group)

DEG_plot_all <- DEG_analysis(Blast_tpm_all_10, group =group, loglimit = 1.0, adj_p_value = 0.05, output_file_name = "DEG_ALL")
DEG_plot_all
```

```{r}
#To minimize noise the blastocyst and blastoid will be separated by TE only. Subsequently the columns will be randomized to minimize bias, and than they will be clustered in ~10 groups. 
#Obtain only TE from Blastocyst
col_delete <- c()
for (i in 1:length(colnames(Blastocyst_tpm))){
  if (grepl("S2.MTE", colnames(Blastocyst_tpm[i])) == F){
    col_delete <- append(col_delete, i)}
}
Blastocyst_tpm_mte <- Blastocyst_tpm[,-col_delete]


#Randomize column order
v <- c(1:length(Blastocyst_tpm_mte))
set.seed(001)
v <- sample(v)
Blastocyst_tpm_mte_random <- Blastocyst_tpm_mte[v]
#Divide in to ~10 groups by making 10 groups

Blastocyst_tpm_mte_10 <- as.data.frame(rowMeans(Blastocyst_tpm_mte_random[1:38]))
colnames(Blastocyst_tpm_mte_10) <- c("1a")
Blastocyst_tpm_mte_10$'2a' <- rowMeans(Blastocyst_tpm_mte_random[39:76])
Blastocyst_tpm_mte_10$'3a' <- rowMeans(Blastocyst_tpm_mte_random[77:114])
Blastocyst_tpm_mte_10$'4a' <- rowMeans(Blastocyst_tpm_mte_random[115:152])
Blastocyst_tpm_mte_10$'5a' <- rowMeans(Blastocyst_tpm_mte_random[153:190])
Blastocyst_tpm_mte_10$'6a' <- rowMeans(Blastocyst_tpm_mte_random[191:228])
Blastocyst_tpm_mte_10$'7a' <- rowMeans(Blastocyst_tpm_mte_random[229:266])
Blastocyst_tpm_mte_10$'8a' <- rowMeans(Blastocyst_tpm_mte_random[267:304])
Blastocyst_tpm_mte_10$'9a' <- rowMeans(Blastocyst_tpm_mte_random[305:342])
Blastocyst_tpm_mte_10$'10a' <- rowMeans(Blastocyst_tpm_mte_random[343:380])
Blastocyst_tpm_mte_10$"gene" <- rownames(Blastocyst_tpm_mte_10)

#Obtain only TE from Blastoid
col_delete <- c()
for (i in 1:length(colnames(Blastocyst_tpm))){
  if (grepl("S2.PTE", colnames(Blastocyst_tpm[i])) == F){
    col_delete <- append(col_delete, i)}
}
Blastocyst_tpm_pte <- Blastocyst_tpm[,-col_delete]


v <- c(1:length(Blastocyst_tpm_pte))
v <- sample(v)
Blastocyst_tpm_pte_random <- Blastocyst_tpm_pte[v]
Blastocyst_tpm_pte_10 <- as.data.frame(rowMeans(Blastocyst_tpm_pte_random[1:16]))
colnames(Blastocyst_tpm_pte_10) <- c("1b")
Blastocyst_tpm_pte_10$'2b' <- rowMeans(Blastocyst_tpm_pte_random[17:32])
Blastocyst_tpm_pte_10$'3b' <- rowMeans(Blastocyst_tpm_pte_random[33:48])
Blastocyst_tpm_pte_10$'4b' <- rowMeans(Blastocyst_tpm_pte_random[49:64])
Blastocyst_tpm_pte_10$'5b' <- rowMeans(Blastocyst_tpm_pte_random[65:80])
Blastocyst_tpm_pte_10$"gene" <- rownames(Blastocyst_tpm_pte_10)

Blast_tpm_te_ptemte_10 <- inner_join(Blastocyst_tpm_mte_10, Blastocyst_tpm_pte_10)
Blast_tpm_te_ptemte_10 <- column_to_rownames(Blast_tpm_te_ptemte_10, var = "gene")
Blast_tpm_te_ptemte_10 <- log((Blast_tpm_te_ptemte_10+1))
group <- c(rep(c("MTE"), times = 10 ), rep(c("PTE"), times =5))
design <- model.matrix(~group)

DEG_plot_MTE_PTE_blastocyst <- DEG_analysis(Blast_tpm_te_ptemte_10, group =group, loglimit = 0.5, adj_p_value = 0.05, output_file_name = "DEG_MTE_PTE_blastocyst")
DEG_plot_MTE_PTE_blastocyst
```

```{r AF/AS}

#To minimize noise the blastocyst and blastoid will be separated by TE only. Subsequently the columns will be randomized to minimize bias, and than they will be clustered in ~10 groups. 
#Obtain only TE from Blastocyst
col_delete <- c()
for (i in 1:length(colnames(Blastocyst_tpm))){
  if (grepl("AF MTE", colnames(Blastocyst_tpm[i])) == F){
    col_delete <- append(col_delete, i)}
}
Blastocyst_tpm_af <- Blastocyst_tpm[,-col_delete]

col_delete <- c()
for (i in 1:length(colnames(Blastocyst_tpm))){
  if (grepl("S3.MTE", colnames(Blastocyst_tpm[i])) == F){
    col_delete <- append(col_delete, i)}
}
Blastocyst_tpm_as <- Blastocyst_tpm[,-col_delete]


#Randomize column order
v <- c(1:length(Blastocyst_tpm_as))
set.seed(001)
v <- sample(v)
Blastocyst_tpm_as_random <- Blastocyst_tpm_as[v]
#Divide in to ~10 groups by making 10 groups

Blastocyst_tpm_as_10 <- as.data.frame(rowMeans(Blastocyst_tpm_as_random[1:13]))
Blastocyst_tpm_as_10$'2a' <- rowMeans(Blastocyst_tpm_as_random[14:26])
Blastocyst_tpm_as_10$'3a' <- rowMeans(Blastocyst_tpm_as_random[27:39])
Blastocyst_tpm_as_10$'4a' <- rowMeans(Blastocyst_tpm_as_random[40:52])
Blastocyst_tpm_as_10$'5a' <- rowMeans(Blastocyst_tpm_as_random[53:65])
Blastocyst_tpm_as_10$'6a' <- rowMeans(Blastocyst_tpm_as_random[66:78])
Blastocyst_tpm_as_10$'7a' <- rowMeans(Blastocyst_tpm_as_random[79:91])
Blastocyst_tpm_as_10$'8a' <- rowMeans(Blastocyst_tpm_as_random[92:104])
Blastocyst_tpm_as_10$'9a' <- rowMeans(Blastocyst_tpm_as_random[105:117])
Blastocyst_tpm_as_10$'10a' <- rowMeans(Blastocyst_tpm_as_random[118:130])
Blastocyst_tpm_as_10$"gene" <- rownames(Blastocyst_tpm_as_10)

#Randomize column order
v <- c(1:length(Blastocyst_tpm_af))
set.seed(001)
v <- sample(v)
Blastocyst_tpm_af_random <- Blastocyst_tpm_af[v]
#Divide in to ~10 groups by making 10 groups

Blastocyst_tpm_af_10 <- as.data.frame(rowMeans(Blastocyst_tpm_af_random[1:8]))
Blastocyst_tpm_af_10$'2b' <- rowMeans(Blastocyst_tpm_af_random[9:16])
Blastocyst_tpm_af_10$'3b' <- rowMeans(Blastocyst_tpm_af_random[17:24])
Blastocyst_tpm_af_10$'4b' <- rowMeans(Blastocyst_tpm_af_random[25:32])
Blastocyst_tpm_af_10$'5b' <- rowMeans(Blastocyst_tpm_af_random[33:40])
Blastocyst_tpm_af_10$'6b' <- rowMeans(Blastocyst_tpm_af_random[41:48])
Blastocyst_tpm_af_10$'7b' <- rowMeans(Blastocyst_tpm_af_random[49:56])
Blastocyst_tpm_af_10$'8b' <- rowMeans(Blastocyst_tpm_af_random[57:64])
Blastocyst_tpm_af_10$'9b' <- rowMeans(Blastocyst_tpm_af_random[65:72])
Blastocyst_tpm_af_10$"gene" <- rownames(Blastocyst_tpm_af_10)

Blastocyst_tpm_af_as <- inner_join(Blastocyst_tpm_af_10, Blastocyst_tpm_as_10)
Blastocyst_tpm_af_as <- column_to_rownames(Blastocyst_tpm_af_as, var = "gene")
Blastocyst_tpm_af_as <- log((Blastocyst_tpm_af_as+1))
group <- c(rep(c("Attachment failure"), times = 9), rep(c("Attachment success"), times = 10))

DEG_analysis(Blastocyst_tpm_af_as, group = group, loglimit = 1, adj_p_value = 0.05, output_file_name = "DEG_AF_AS")
```

```{r}

DEGS_blast <- readxl::read_xlsx("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/output/DEGS_blastocyst_blastoids.xlsx")
DEGS_blast_blastocysts <- DEGS_blast[c(1,3,5)]
degs_blastocysts <- intersect(DEGS_blast_blastocysts$Blastocyst_icm, DEGS_blast_blastocysts$Blastocyst_pre)
degs_blastocysts <- intersect(degs_blastocysts, DEGS_blast_blastocysts$Blastocyst_te)

DEGS_blast_blastoids <- DEGS_blast[c(2,4,6)]
degs_blastoids <- intersect(DEGS_blast_blastoids$Blastoid_icm, DEGS_blast_blastoids$Blastoid_pre)
degs_blastoids <- intersect(degs_blastoids, DEGS_blast_blastoids$Blastoid_te)

DEGS_blastoids_te <- subset(DEGS_blast_blastoids$Blastoid_te, !(DEGS_blast_blastoids$Blastoid_te %in% degs_blastoids))
DEGS_blastocysts_te <- subset(DEGS_blast_blastocysts$Blastocyst_te, !(DEGS_blast_blastocysts$Blastocyst_te %in% degs_blastocysts))


```

## Seurat

```{r}
#When combining the data set it is essential 

Blast_tpm_seurat <- CreateSeuratObject(Blast_tpm)
name_list <- rownames(as.data.frame(Blast_tpm_seurat@active.ident))
Blast_tpm_seurat@meta.data$names <- name_list



cell_layers = list()
cystoid <- list()
for(i in Blast_tpm_seurat@meta.data$names){
  if (grepl("S1", i) | grepl("S2", i) | grepl("S3", i) | grepl("ISK", i) | grepl("AF", i)== T){
    cystoid <- append(cystoid, "Blastocyst")
    if (grepl("AF", i) == T){
      cell_layers <- append(cell_layers, "Blastocyst-AF")}
    else if (grepl("MTE", i) == T){
      cell_layers <- append(cell_layers, "Blastocyst-MTE")}
    else if (grepl("EPI", i) == T){
      cell_layers <- append(cell_layers, "Blastocyst-EPI")}
    else if (grepl("ICM", i) == T){
      cell_layers <- append(cell_layers, "Blastocyst-ICM")}    
    else if (grepl("ISK", i) == T){
      cell_layers <-append(cell_layers, "Blastocyst_ISK")}
    else if (grepl("PTE", i) == T){
      cell_layers <-append(cell_layers, "Blastocyst_PTE")}
    else if (grepl("PE", i) == T){
      cell_layers <- append(cell_layers, "Blastocyst-PrE")}
    else if (grepl("S1.TE", i) == T){
      cell_layers <-append(cell_layers, "Blastocyst_TE")}}
  else{
    cystoid <- append(cystoid, "Blastoid")
    if (grepl("PTE", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-PTE")}
    else if (grepl("TE", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-TE")}    
    else if (grepl("ICM", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-ICM")}
    else if (grepl("Naive", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-Naive")}
    else if (grepl("PRE", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-PrE")}
    else if (grepl("EPI", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-EPI")}
     else if (grepl("Unknown", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-Unknown")}
    else{print(i)}}}

cell_layers <- unlist(cell_layers)
Blast_tpm_seurat@meta.data$cell_layers <- cell_layers

set.seed(1234)
varGene = 3000
resolutionwanted=c(0.02,0.5,1,1.5,2,2.5,3,3.5,4,4.5,5,5.5,6)
Blast_tpm_seurat2 <- Blast_tpm_seurat
Blast_tpm_seurat <- SetAssayData(Blast_tpm_seurat, slot = 'data', new.data = log1p(Blast_tpm))
Blast_tpm_seurat <- FindVariableFeatures(object = Blast_tpm_seurat, nfeatures = varGene)
Blast_tpm_seurat <- ScaleData(Blast_tpm_seurat)
Blast_tpm_seurat <- RunPCA(Blast_tpm_seurat)
Blast_tpm_seurat <- FindNeighbors(Blast_tpm_seurat)
Blast_tpm_seurat <- FindClusters(Blast_tpm_seurat, resolution = 0.5, verbose=FALSE)
Blast_tpm_seurat <- RunUMAP(Blast_tpm_seurat, dims = 1:20, n.components = 3L, min.dist = 0.5, verbose=FALSE)


#plot of Stage names 
Idents(Blast_tpm_seurat) <- Blast_tpm_seurat@meta.data$cell_layers
DimPlot(Blast_tpm_seurat)
Blast_tpm_seurat
```

```{r}
FeaturePlot(Blast_tpm_seurat, features = c("GATA2", "GATA3"))
```

```{r}
features <- c("GATA2", "GATA3", "POU5F1", "KLF17", "GATA4", "SOX17")

DotPlot(Blast_tpm_seurat, features = features) + RotatedAxis()+scale_colour_gradient2(low="blue", mid="white", high="red")
```

```{r}

Blast_tpm_seurat_te <-  subset(Blast_tpm_seurat, cell_layers == "Blastocyst-MTE" | cell_layers == "Blastocyst_PTE" | cell_layers == "Blastoid-PTE"| cell_layers == "Blastoid-TE")
Blast_tpm_seurat_te_clusters <- data.frame(Cell = names(Idents(Blast_tpm_seurat_te)), CellType = as.character(Idents(Blast_tpm_seurat_te)), stringsAsFactors = F)
Blast_tpm_seurat_te_progeny <- progeny(Blast_tpm_seurat_te, scale = F, organism = "Human", top = 500, perm = 1, return_assay = T)

Blast_tpm_seurat_te_progeny <- ScaleData(Blast_tpm_seurat_te_progeny, assay = "progeny")

BCP_df <- as.data.frame(t(GetAssayData(Blast_tpm_seurat_te_progeny, slot = "scale.data", assay = "progeny")))%>%
  rownames_to_column("Cell") %>%
  gather(Pathway, Activity, -Cell)

BCP_df <- inner_join(BCP_df, Blast_tpm_seurat_te_clusters)
sum_bcp_df <- BCP_df %>%
  group_by(Pathway, CellType) %>%
  summarise(avg = mean(Activity), std=sd(Activity))

sum_bcp_df <- sum_bcp_df %>%
  dplyr::select(-std) %>%
  spread(Pathway, avg) %>%
  data.frame(row.names = 1, check.names = F, stringsAsFactors = F)


paletteLength = 100
myColor = colorRampPalette(c("Darkblue", "white","red"))(paletteLength)

progenyBreaks = c(seq(min(sum_bcp_df), 0, 
                      length.out=ceiling(paletteLength/2) + 1),
                  seq(max(sum_bcp_df)/paletteLength, 
                      max(sum_bcp_df), 
                      length.out=floor(paletteLength/2)))
progeny_hmap = pheatmap(t(sum_bcp_df),fontsize=14, 
                        fontsize_row = 10, 
                        color=myColor, breaks = progenyBreaks, 
                        main = "Pathway analysis Blastoid vs. Blastocyst", angle_col = 45,
                        treeheight_col = 0,  border_color = NA)
```

```{r fig.width=10, fig.height =6}
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

Blast_tpm_seurat <- CellCycleScoring(Blast_tpm_seurat, s.features = s.genes, g2m.features = g2m.genes, set.ident =T)

Idents(Blast_tpm_seurat) <- Blast_tpm_seurat@meta.data$cell_layers
Cluster_dim <- DimPlot(Blast_tpm_seurat)
Idents(Blast_tpm_seurat) <- Blast_tpm_seurat@meta.data$Phase
Cell_cycle_dim <- DimPlot(Blast_tpm_seurat)
Cluster_dim + Cell_cycle_dim
```

### Heatmap

```{r}
library("rio")
new_vector = c()
for (i in DEGS_blast_blastocysts){
  new_vector <- c(new_vector, i)
}
DEG_names_blastocyst <- unique(new_vector)

new_vector = c()
for (i in DEGS_blast){
  new_vector <- c(new_vector, i)
}
DEG_names <- unique(new_vector)
Blast_heatmap_DEGS <-Blast_tpm_seurat

DoHeatmap(Blast_heatmap_DEGS, features = DEGS_blastocysts_te, group.by = "cell_layers", size = 3)+theme(text = element_text(size = 5))+NoLegend()
```

## Integration

```{r}
data_group = list()
for (i in Blast_tpm_seurat2$cell_layers){
  if (grepl("Blastoid", i) == T){
    data_group = append(data_group, "Blastoid")
  } else if (grepl("Blastocyst", i) == T){
    data_group = append(data_group, "Blastocyst")
  } else (print(i))
}
data_group = unlist(data_group)
Blast_tpm_seurat2@meta.data$data_set = data_group

cell_type = list()
for (i in Blast_tpm_seurat2$cell_layers){
  if (grepl("AF", i) == T){
    cell_type = append(cell_type, "AF")
  } else if (grepl("MTE", i) == T){
  cell_type = append(cell_type, "MTE")
  } else if (grepl("PTE", i) == T){
    cell_type = append(cell_type, "PTE")
  } else if (grepl("TE", i) == T){
    cell_type = append(cell_type, "TE")
  } else if (grepl("PrE", i) == T){
    cell_type = append(cell_type, "PrE")
  } else if (grepl("Naive", i) == T){
    cell_type = append(cell_type, "Naive")
  } else if (grepl("EPI", i) == T){
    cell_type = append(cell_type, "EPI")
  } else if (grepl("ICM", i) == T){
    cell_type = append(cell_type, "ICM")
  } else if (grepl("ISK", i) == T){
    cell_type = append(cell_type, "ISK")
  } else if (grepl("Unknown", i) == T){
    cell_type = append(cell_type, "Unknown")

  }}

cell_type = unlist(cell_type)
Blast_tpm_seurat2@meta.data$cell_type = cell_type


Blast_test <- Blast_tpm_seurat2
Blast_integrated <- Blast_tpm_seurat2
Blast_integrated[["RNA"]] <- split(Blast_integrated[["RNA"]], f =Blast_integrated$data_set)
Blast_integrated <- NormalizeData(Blast_integrated, scale.factor = 1)

Blast_integrated@assays[["RNA"]]@layers[["data.Blastocyst"]] <- as(log1p(Blast_integrated@assays[["RNA"]]@layers[["counts.Blastocyst"]]), "sparseMatrix")
Blast_integrated@assays[["RNA"]]@layers[["data.Blastoid"]] <- as(log1p(Blast_integrated@assays[["RNA"]]@layers[["counts.Blastoid"]]), "sparseMatrix")


Blast_integrated <- FindVariableFeatures(Blast_integrated)
Blast_integrated <- ScaleData(Blast_integrated)
Blast_integrated <- RunPCA(Blast_integrated)
Blast_integrated <- FindNeighbors(Blast_integrated, dims = 1:30, reduction = "pca")
Blast_integrated <- FindClusters(Blast_integrated, resolution = 2, cluster.name = "unintegrated_clusters")
Blast_integrated <- RunUMAP(Blast_integrated, dims = 1:30, reduction = "pca", reduction.name = "umap.unintegrated")
DimPlot(Blast_integrated, reduction = "umap.unintegrated", group.by = "data_set")
Blast_integrated <- Blast_integrated
Blast_integrated <- IntegrateLayers(object = Blast_integrated, method = CCAIntegration, orig.reduction = "pca", new.reduction = "integrated.cca",
    verbose = FALSE, assay = "RNA")

Blast_integrated[["RNA"]] <- JoinLayers(Blast_integrated[["RNA"]])
Blast_integrated <- RunPCA(Blast_integrated)
Blast_integrated <- FindNeighbors(Blast_integrated, reduction = "integrated.cca", dims = 1:30)
Blast_integrated <- FindClusters(Blast_integrated, resolution = 1)
Blast_integrated <- RunUMAP(Blast_integrated, dims = 1:30, reduction = "integrated.cca")
plot_integration <- DimPlot(Blast_integrated, reduction = "umap", group.by = "data_set")
#plot_integration_cells <- DimPlot(Blast_integrated, reduction = "umap", group.by = "cell_type")
plot_integration_cells <-  DimPlot(Blast_integrated, reduction = "umap", group.by = "cell_type", cols = c( "#c5ca30", "#ff924c", "#ffca3a","#1982c4" ,  "#36949d","#8ac926","#ff595e","#565aa0" ,"cornflowerblue","#6a4c99" ), shape.by = "data_set", pt.size = 1)
plot_integration + plot_integration_cells

DimPlot(Blast_integrated, reduction = "umap", group.by = "data_set")

```

```{r fig.width = 10, fig.height = 4}
DimPlot(Blast_integrated, reduction = "umap" , group.by = "cell_type")
time_list = list()
for (i in Blast_integrated$names){
  if (grepl("96h", i) == T){
    time_list = append(time_list, "96h")
  } else if (grepl("S1", i) == T){
    time_list = append(time_list, "S1")
  } else if (grepl("S2", i) == T){
    time_list = append(time_list, "S2")
  } else if (grepl("S3", i) == T){
    time_list = append(time_list, "S3")
  } else if (grepl("AF", i) == T){
    time_list = append(time_list, "AF")
  } else if (grepl("60h", i) == T){
    time_list = append(time_list, "60h")
  } else if (grepl("24h", i) == T){
    time_list = append(time_list, "24h")
  } else if (grepl("transitioning", i) == T){
    time_list = append(time_list, "transitioning")
  } else if (grepl("ISK", i) == T){
    time_list = append(time_list, "ISK")
  } else if (grepl("Naive", i) == T){
    time_list = append(time_list, "Naive")
  } else if (grepl("Unknown", i) == T){
    time_list = append(time_list, "Unknown")}
    else{print(i)}
}
time_list = unlist(time_list)
Blast_integrated@meta.data$times = time_list


plot_1 <- DimPlot(Blast_integrated, reduction = "umap", group.by = "cell_type", cols = c( "#c5ca30", "#ff924c", "#ffca3a","#1982c4" ,  "#36949d","#8ac926","#ff595e","purple" ,"cornflowerblue","#6a4c99" ), shape.by = "data_set", pt.size = 0.5)
plot_2 <- DimPlot(Blast_integrated, reduction = "umap", group.by = "times", cols = c( "#c5ca30", "#ff924c", "#ffca3a","#1982c4" ,  "#36949d","#8ac926","#ff595e","purple" ,"cornflowerblue","#6a4c99", "grey" ), shape.by = "data_set", pt.size = 0.5)
plot_1+ plot_2

```

```{r}
do_want <- c("EPI", "ICM", "MTE", "PTE", "TE", "PrE", "Naive")
Blast_integrated <- subset(Blast_integrated, cell_type == "EPI" | cell_type == "ICM" | cell_type == "MTE" | cell_type == "PTE" | cell_type == "TE" | cell_type == "PrE" | cell_type == "Naive")
Blast_integrated <- FindClusters(Blast_integrated, resolution = 1)
DimPlot(Blast_integrated, reduction ="umap")


features = c("GATA2", "GATA3", "POU5F1", "KLF17", "GATA4", "SOX17")
DotPlot(Blast_integrated, features = features) + RotatedAxis()+scale_colour_gradient2(low="blue", mid="white", high="red")
plot_1 <- DimPlot(Blast_integrated, reduction = "umap", group.by = "cell_type", cols = c( "#6a4c99", "#ff924c", "#ffca3a","#1982c4" ,  "#36949d","#8ac926","#ff595e","purple" ,"cornflowerblue","#6a4c99" ), shape.by = "data_set", pt.size = 0.5)
plot_1



Blast_integrated_blastocyst <- subset(Blast_integrated, data_set == "Blastocyst")
plot_2 <- DimPlot(Blast_integrated_blastocyst, reduction = "umap", group.by = "cell_type", cols = c( "#6a4c99", "#ff924c", "#ffca3a","#ff595e" ,  "#36949d","#8ac926","cornflowerblue","purple" ,"#6a4c99" ), shape.by = "data_set", pt.size = 0.5)


Blast_integrated_blastoid <- subset(Blast_integrated, data_set == "Blastoid")
plot_3 <- DimPlot(Blast_integrated_blastoid, reduction = "umap", group.by = "cell_type", cols = c( "#6a4c99", "cornflowerblue", "#ff595e", "#36949d",  "#8ac926", "#ff924c", "#ffca3a","purple"), shape.by = "data_set", pt.size = 0.5)
plot_1+ plot_2 + plot_3
```

```{r}
DimPlot(Blast_integrated, reduction = "umap", split.by = "data_set", pt.size = 0.8)
```

```{r}
FeaturePlot(Blast_integrated, features = c("SOX17", "GATA4"), split.by = "data_set", max.cutoff = 3, cols = c("grey",
    "red"), reduction = "umap")
FeaturePlot(Blast_integrated, features = c("GATA2", "GATA3"), split.by = "data_set", max.cutoff = 3, cols = c("grey",
    "red"), reduction = "umap")
FeaturePlot(Blast_integrated, features = c("KLF17", "POU5F1"), split.by = "data_set", max.cutoff = 3, cols = c("grey",
    "red"), reduction = "umap")


```

```{r}
FeaturePlot(Blast_integrated, features = c("VIRMA"), pt.size = 0.5, split.by = "data_set", max.cutoff = 3, reduction = "umap")
```

```{r fig.width =10}

VIRMA_list = list(c(
"AKAP11",
"ANKHD1",
"ANKRD11",
"ANKRD36",
"ANKRD36B",
"ARHGAP11B",
"ARHGAP35",
"BCL9L",
"BRCA2",
"C1orf226",
"CASP8AP2",
"CCDC144CP",
"CCDC85C",
"CDC42EP2",
"CLCN5",
"CREBBP",
"CTTNBP2",
"ERCC4",
"FBXL18",
"FEM1A",
"H2AC1",
"HEATR5A",
"HERC2",
"HERC3",
"HOOK1",
"HSPA1A",
"ITPR3",
"JUN",
"KIAA0232",
"KIF13B",
"KLF13",
"LINC00842",
"LRP4",
"MRPL23",
"MTUS2",
"NFIC",
"PFKFB3",
"PPP1R12B",
"REL",
"RPS6KA5",
"SACS",
"SATB2",
"SHROOM3",
"SIK1",
"SMG1P2",
"SNRPD3",
"SNX33",
"SORBS1",
"SPEN",
"SPTB",
"SRCAP",
"SYNE2",
"TACC2",
"TBC1D4",
"TET2",
"TLN2",
"TNRC6B",
"TRIM2",
"TUBB3",
"ZBED6",
"ZNF12",
"ZNF236",
"ZNF445",
"ZNF704",
"ZNF800"	
))

Blast_integrated <- AddModuleScore(
  object = Blast_integrated,
 features = VIRMA_list,
 ctrl = 5,
 name = "VIRMA_features"
)

feature_VIRMA <- FeaturePlot(Blast_integrated, features = "VIRMA_features1", label = F, repel = T, reduction = "umap") +  scale_colour_gradientn(colours = rev(RColorBrewer::brewer.pal(n=11, name = "RdBu")))
Dim_VIRMA <- DimPlot(Blast_integrated, reduction = "umap", group.by = "data_set")
library("gridExtra")
grid.arrange(feature_VIRMA, Dim_VIRMA)
```

```{r}

Blast_integrated <- AddModuleScore(
 object = Blast_integrated,
 features = as.list(intersect(rownames(Blast_integrated), VIRMA_RIP)),
 ctrl = 5,
 name = "VIRMA_RIP_features"
)


feature_VIRMA <- FeaturePlot(Blast_integrated, features = "VIRMA_RIP_features1", label = F, repel = T, reduction = "umap", ) +  scale_colour_gradientn(colours = rev(RColorBrewer::brewer.pal(n=11, name = "RdBu")))
feature_VIRMA + Dim_VIRMA
```

```{r}
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

Blast_integrated <- CellCycleScoring(Blast_integrated, s.features = s.genes, g2m.features = g2m.genes, set.ident =T)

Idents(Blast_integrated) <- Blast_integrated@meta.data$cell_layers
Cluster_dim <- DimPlot(Blast_integrated, reduction = "umap")
Idents(Blast_integrated) <- Blast_integrated@meta.data$Phase
Cell_cycle_dim <- DimPlot(Blast_integrated, reduction = "umap")
Cell_cycle_dim
```

```{r}
FeaturePlot(Blast_integrated, pt.size = 0.8,  reduction = "umap", features = c("CPSF5", "CPSF6"), split.by = "data_set")
FeaturePlot(Blast_integrated, pt.size = 0.8,  reduction = "umap", features = c("IGF2BP1", "IGF2BP2", "IGF2BP3"), split.by = "data_set")
```

```{r}
FeaturePlot(Blast_integrated, pt.size = 0.8,  reduction = "umap", features = c("CBLL1", "WTAP", "ZC3H13", "RBM15", "RBM15B"), split.by = "data_set")
```

```{r}


IGF2BP_list = list(c("IGF2BP1", "IGF2BP2", "IGF2BP3"))

Blast_integrated <- AddModuleScore(
  object = Blast_integrated,
 features = IGF2BP_list,
 ctrl = 5,
 name = "IGF2BP_features"
)

feature_IGF2BP <- FeaturePlot(Blast_integrated, features = "IGF2BP_features1", label = F, repel = T, reduction = "umap") +  scale_colour_gradientn(colours = rev(RColorBrewer::brewer.pal(n=11, name = "RdBu")))
Dim_IGF2BP <- DimPlot(Blast_integrated, reduction = "umap", group.by = "data_set")
library("gridExtra")
grid.arrange(feature_IGF2BP, Dim_IGF2BP)

feature_IGF2BP +feature_VIRMA
```

```{r}
VlnPlot(Blast_integrated, group.by = "cell_type", features = "VIRMA", split.by = "data_set")
VlnPlot(Blast_integrated, group.by = "cell_type", features = "IGF2BP3", split.by = "data_set")


features <- c("VIRMA", "IGF2BP3", "METTL3", "METTL14")
Idents(Blast_integrated) <- Blast_integrated@meta.data$cell_layers
DotPlot(Blast_integrated, features = features) + RotatedAxis() +scale_colour_gradient2(low="blue", mid="white", high="red")

FeaturePlot(Blast_integrated, features = "METTL3", split.by = "data_set", reduction = "umap", pt.size = 0.8)

```

```{r fig.height= 30}
Potential_virma_TFs <- read.csv("output/TFBS.csv")
Potential_virma_TFs <- unlist(Potential_virma_TFs)
DoHeatmap(Blast_integrated, features = Potential_virma_TFs, group.by = "cell_layers", size = 3, slot = "data")+theme(text = element_text(size = 5))+NoLegend()

DEGS_All <- read.csv("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/output/DEG_ALL", sep = " ")

TF_VIRMA <- DEGS_All[Potential_virma_TFs, ]
TF_VIRMA <- TF_VIRMA[order(TF_VIRMA$logFC, decreasing = F, na.last = T),]
TF_VIRMA <- rownames(head(TF_VIRMA, 200))

Blastocyst_TF_virma <- as.data.frame(rowMeans(Blastocyst_tpm[1:1108]))
Blastocyst_TF_virma$"gene" <- rownames(Blastocyst_TF_virma)
Blastoid_TF_virma <- as.data.frame(rowMeans(Blastoid_tpm[1:1816]))
Blastoid_TF_virma$"gene" <- rownames(Blastoid_TF_virma)
Blast_TF_virma <- inner_join(Blastocyst_TF_virma, Blastoid_TF_virma)
Blast_TF_virma <- column_to_rownames(Blast_TF_virma, "gene")
Blast_TF_virma <- Blast_TF_virma[Potential_virma_TFs,]
Blast_TF_virma$"diff" <- (Blast_TF_virma$`rowMeans(Blastocyst_tpm[1:1108])` - Blast_TF_virma$`rowMeans(Blastoid_tpm[1:1816])`)
Blast_TF_virma <- Blast_TF_virma[order(Blast_TF_virma$diff, decreasing = T),]
Top_TF_virma <- rownames(head(Blast_TF_virma, 100))

VlnPlot(Blast_integrated, features = Top_TF_virma, group.by = "cell_layers", split.by = "data_set", )
FeaturePlot(Blast_integrated, features = Top_TF_virma, split.by = "data_set", reduction = "umap")

```

```{r}
for (gene in TF_VIRMA){
  if (gene %in% rownames(Blastocyst_tpm_te)){
    df <- data.frame(t(Blastocyst_tpm_te[c("VIRMA", gene),]))

    df <- as.data.frame(sapply(df, as.numeric))
    df[df==0] <- NA
    df[complete.cases(df), ]
    df <- na.omit(df)
    
    plot <- ggplot(df, aes(x= df$VIRMA, y=df[,2]))+
                geom_point()+
                stat_summary(fun.data= df) + 
                geom_smooth(method='lm')+
                stat_poly_eq()+
                ylab(gene)
    print(plot)}
    
   else{print(gene)}}

```

```{r}
FeaturePlot(Blast_integrated, features = c("VIRMA", "METTL3", "METTL16", "METTL14", "RBM15", "CBLL1", "WTAP"), split.by = "data_set", reduction = "umap")
VlnPlot(Blast_integrated, features = c("VIRMA", "METTL3", "METTL16", "METTL14", "RBM15", "CBLL1", "WTAP"), group.by = "cell_layers", split.by = "data_set", )
```

```{r}
FeaturePlot(Blast_integrated, features = c("VIRMA", "NLRP2", "NLRP5", "NLRP7", "PADI6", "KHDC3L", "TLE6", "OOEP", "BTG4", "PATL2", "TRIP13", "TUBB8"), split.by = "data_set", reduction = "umap", )
VlnPlot(Blast_integrated, features = c("VIRMA", "NLRP2", "NLRP5", "NLRP7", "PADI6", "KHDC3L", "TLE6", "OOEP", "BTG4", "PATL2", "TRIP13", "TUBB8"), group.by = "cell_layers", split.by = "data_set", )
```

```{r}
FeaturePlot(Blast_integrated, features = c("VIRMA", "DOT1L", "MEN1"), split.by = "data_set", reduction = "umap")
VlnPlot(Blast_integrated, features = c("VIRMA", "DOT1L", "MEN1"), group.by = "cell_layers", split.by = "data_set", )
```

```{r}
VIRMA_TF <- c("AML1A", "TFAP2C", "ATF", "CEBPA", "BPTF", "FOXL1", "MEIS1", "MEIS1A", "STAT3")
VlnPlot(Blast_integrated,  features = VIRMA_TF, split.by = "data_set")
FeaturePlot(Blast_integrated, features = c("VIRMA", "CEBPA", "MYC"), reduction = "umap", split.by = "data_set")
VlnPlot(Blast_integrated, features = c("VIRMA", "CEBPA", "MYC"), group.by = "cell_layers")
```

```{r fig.height= 15}
TOP_TFs_VIRMA <- c("RUNX1", "TFAP2C", "ATF1", "CEBPA", "BPTF", "FOXL1", "MEIS1", 'MEIS1A', "STAT3", "VIRMA")
FeaturePlot(Blast_integrated, features = TOP_TFs_VIRMA, reduction = "umap", split.by = "data_set")
VlnPlot(Blast_integrated, features = TOP_TFs_VIRMA, group.by = "cell_layers", split.by = "data_set")
```

```{r fig.height = 20}
FeaturePlot(Blast_integrated, features = c("DNMT1", "DNMT3A", "DNMT3B", "TET1", "TET2", "TET3", "CPSF6", "NUDT21", "VIRMA", "YAP1"), reduction = "umap", split.by = "data_set")
VlnPlot(Blast_integrated, features = c("DNMT1", "DNMT3A", "DNMT3B", "TET1", "TET2", "TET3", "CPSF6", "NUDT21", "VIRMA", "YAP1"), group.by = "cell_layers", split.by = "data_set")
FeaturePlot(Blast_integrated, features = c("VIRMA", "YAP1", "CHST11", "CDX2", "MOB1B", "CDK1", "BCL2", "LATS1", "LATS2", "AMOT", "NF2"), reduction = "umap", split.by = "data_set")
VlnPlot(Blast_integrated, features = c("VIRMA", "YAP1", "CHST11", "CDX2", "MOB1B", "CDK1", "BCL2", "LATS1", "LATS2", "AMOT", "NF2"), group.by = "cell_layers", split.by = "data_set")
 
```

```{r fig.height= 6}
FeaturePlot(Blast_integrated, features = c("VIRMA", "METTL3", "METTL14", "METTL16", "WTAP", "CBLL1", "ZC3H13", "RBM15"), split.by = "data_set", reduction = "umap")
VlnPlot(Blast_integrated, features =  c("VIRMA", "METTL3", "METTL14", "METTL16", "WTAP", "CBLL1", "ZC3H13", "RBM15"), group.by = "cell_layers", split.by = "data_set", )
DoHeatmap(Blast_integrated, features =  c("VIRMA", "METTL3", "METTL14", "METTL16", "WTAP", "CBLL1", "ZC3H13", "RBM15"), group.by = "cell_layers", slot = "data", label = F)+scale_fill_gradient(low = "lightyellow", high = "red")
```

```{r}
DoHeatmap(Blast_integrated, features =  c("VIRMA", "HNRNPA2B1", "HNRNPC", "YTHDC1", "YTHDF2", "YTHDF1", "YTHDF3", "IGF2BP1", "IGF2BP2", "IGF2BP3", "FTO", "ALKBH5", "SRSF3", "SRSF1", "SRSF10", "HNRNP6", "ELF3", "ELAVL1", "PABPC1", "FMR1", "MATR3"), group.by = "cell_layers", slot = "data", , label = F)+scale_fill_gradient(low = "lightyellow", high = "darkblue")
VlnPlot(Blast_integrated, features =  c("VIRMA", "HNRNPA2B1", "HNRNPC", "YTHDC1", "YTHDF2", "YTHDF1", "YTHDF3", "IGF2BP1", "IGF2BP2", "IGF2BP3", "FTO", "ALKBH5", "SRSF3", "SRSF1", "SRSF10", "HNRNP6", "ELF3", "ELAVL1", "PABPC1", "FMR1", "MATR3"), group.by = "cell_layers", split.by = "data_set")
```

```{r}
FeaturePlot(Blast_integrated, features = c("VIRMA", "SNAI1", "SMC1A", "CDH1", "CDH2"), split.by = "data_set", reduction = "umap")
VlnPlot(Blast_integrated, features =  c("VIRMA", "SNAI1", "SMC1A", "CDH1", "CDH2"), group.by = "cell_layers", split.by = "data_set", )
```

```{r}
METTL_gene_list_nm <- c("METTL1", "METTL2A", "METTL2B", "METTL3", "METTL4", "METTL5", "METTL6", "METTL8", "METTL14", "METTL15", "METTL16", "METTL19", "METTL25B")
METTL_gene_list_pm <- c("METTL9", "METTL10", "METTL11A", "METTL11B", "METTL12", "METTL13", "METTL18", "METTL20",  "METTL21A", "METTL21B", "METTL21C", "METTL21D", "METTL21E", "METTL22", "METTL23")
METTL_gene_list_other <- c("METTL7A", "METTL7B", "METTL17", "METTL25", "METTL26", "METTL27")

DoHeatmap(Blast_integrated, features = METTL_gene_list_nm, group.by = "cell_layers", size = 3, slot = "data")+theme(text = element_text(size = 5))+NoLegend()

DoHeatmap(Blast_integrated, features = METTL_gene_list_pm, group.by = "cell_layers", size = 3, slot = "data")+theme(text = element_text(size = 5))+NoLegend()

DoHeatmap(Blast_integrated, features = METTL_gene_list_other, group.by = "cell_layers", size = 3, slot = "data")+theme(text = element_text(size = 5))+NoLegend()
```

```{r}
Paternal_histone_features <- c("KDM3B", "KDM2A", "KDM6A", "KAT6B", "KAT6A", "MSL3", "KMT5C", "PHF8", "NSD1", "KMT5A", "KMT5C", "SUPT7L", "HDAC9", "ASH1L", "NCOA1")

DoHeatmap(Blast_integrated, features = Paternal_histone_features, group.by = "cell_layers", size = 3, slot = "data")+theme(text = element_text(size = 5))+NoLegend()
VlnPlot(Blast_integrated, features = Paternal_histone_features, group.by = "cell_layers", split.by = "data_set", )
```

```{r}
FeaturePlot(Blast_integrated, features = c("VIRMA", "DCP1", "DCP2", "CHAF1A", "CHAF1B", "XRN1", "ZFP36", "BRF1", "TIA1"), split.by = "data_set", reduction = "umap")
VlnPlot(Blast_integrated, features = c("VIRMA", "DCP1", "DCP2", "CHAF1A", "CHAF1B", "XRN1", "ZFP36", "BRF1", "TIA1"), group.by = "cell_layers", split.by = "data_set", )
```

```{r}
VIRMA_tfs <-  c("REST", "IKZF1", "ELK4", "ELK1", "VIRMA")

FeaturePlot(Blast_integrated, features = VIRMA_tfs, split.by = "data_set", reduction = "umap")
VlnPlot(Blast_integrated, features = VIRMA_tfs, group.by = "cell_layers", split.by = "data_set", )

```

```{r fig.height=30}
library("scHelper")
VIRMA_RNA_df <- data.frame(GetAssayData(Blast_integrated), check.names = F)
colnames(VIRMA_RNA_df) <- Blast_integrated@meta.data$data_set

row_delete <- c()
for (i in 1:length(rownames(VIRMA_RNA_df))){
  if(rownames(VIRMA_RNA_df[i,]) %in% VIRMA_binding_RNAs){
    next}
  else{
    row_delete <- append(row_delete, i)}
}
VIRMA_RNA_df <- VIRMA_RNA_df[-row_delete,]
VIRMA_RNA_df_sum <- as.data.frame(rowSums(VIRMA_RNA_df["Blastocyst"]))
colnames(VIRMA_RNA_df_sum) <- "Blastocyst"
VIRMA_RNA_df_sum$"Blastoid" <- rowSums(VIRMA_RNA_df["Blastoid"]) 

heatmap_virma_all <- pheatmap(VIRMA_RNA_df_sum, color = colorRampPalette(rev(brewer.pal(n = 7, name =
  "RdYlBu")))(100), fontsize_row = 1, height = 20)
virma.clust <- cbind(VIRMA_RNA_df_sum, cluster = cutree(heatmap_virma_all$tree_row, k = 10))
pheatmap(virma.clust)
print(virma.clust)
write.csv(virma.clust, "output/virma_clusters.csv")
cluster_1 <- subset(virma.clust, virma.clust$cluster == 1)
cluster_2 <- subset(virma.clust, virma.clust$cluster == 2)
cluster_3 <- subset(virma.clust, virma.clust$cluster == 3)
cluster_4 <- subset(virma.clust, virma.clust$cluster == 4)
cluster_5 <- subset(virma.clust, virma.clust$cluster == 5)
cluster_6 <- subset(virma.clust, virma.clust$cluster == 6)
cluster_7 <- subset(virma.clust, virma.clust$cluster == 7)
cluster_8 <- subset(virma.clust, virma.clust$cluster == 8)
cluster_9 <- subset(virma.clust, virma.clust$cluster == 9)
cluster_10 <- subset(virma.clust, virma.clust$cluster == 10)
ph_1 <- pheatmap(cluster_1[,c(1,2)], fontsize_row = 10)
ph_2 <- pheatmap(cluster_2[,c(1,2)],)
ph_3 <- pheatmap(cluster_3[,c(1,2)],)
ph_4 <- pheatmap(cluster_4[,c(1,2)],)
ph_5 <- pheatmap(cluster_5[,c(1,2)],)
ph_6 <- pheatmap(cluster_6[,c(1,2)],)
ph_7 <- pheatmap(cluster_7[,c(1,2)],)
ph_8 <- pheatmap(cluster_8[,c(1,2)],)
ph_9 <- pheatmap(cluster_9[,c(1,2)],)
ph_10 <- pheatmap(cluster_10[,c(1,2)])

virma.clust_sum <- virma.clust
virma.clust_sum$sum <- (virma.clust_sum$Blastocyst-virma.clust_sum$Blastoid)
virma.clust_sum$log2fc <- (log2((virma.clust_sum$Blastocyst+0.0000001)/(virma.clust_sum$Blastoid+0.0000001)))
virma_targets_blastocyst_up <- subset(virma.clust_sum, virma.clust_sum$log2fc > 2)
write.csv(virma.clust_sum, "output/virma_binding_blastocyst_up")
```

```{r}
VIRMA_RNA_df_celltypes <- data.frame(GetAssayData(Blast_integrated), check.names = F)
colnames(VIRMA_RNA_df_celltypes) <- Blast_integrated@meta.data$cell_layers

row_delete <- c()
for (i in 1:length(rownames(VIRMA_RNA_df_celltypes))){
  if (rownames(VIRMA_RNA_df_celltypes[i,]) %in% VIRMA_binding_RNAs){
    if (rowSums(VIRMA_RNA_df_celltypes[i,]) <30){
     row_delete <- append(row_delete, i)}
    else {next}}
  else{
    row_delete <- append(row_delete, i)}
}
VIRMA_RNA_df_celltypes <- VIRMA_RNA_df_celltypes[-row_delete,]
hist(rowSums(VIRMA_RNA_df_celltypes), breaks = 5000, xlim = c(0, 50))

# Transpose the data
VIRMA_RNA_df_celltypes <- as.data.frame(VIRMA_RNA_df_celltypes)
VIRMA_RNA_df_celltypes_ss <- VIRMA_RNA_df_celltypes[0:1000, 0:1000]
# Define sample groups (you can modify this based on your actual sample names)

# Define sample group colors
sample_group_colors <- c(
  "Blastocyst.MTE" = "blue", 
  "Blastocyst_PTE" = "red",
  "Blastocyst.EPI" = "green",
  "Blastocyst.ICM" = "orange",
  "Blastocyst.PrE" = "purple",
  "Blastocyst_TE" = "yellow",
  "Blastoid.EPI" = "cyan",
  "Blastoid.TE" = "magenta",
  "Blastoid.Naive" = "brown",
  "Blastoid.PrE" = "pink"
)

# Create the heatmap without row and column labels
heatmap <- pheatmap(
  VIRMA_RNA_df_celltypes,
  cluster_rows = TRUE, # Cluster genes
  cluster_cols = TRUE, # Cluster samples
  scale = "row", # Scale rows (genes)
  main = "Gene Expression Heatmap",
  show_colnames = FALSE, # Do not show column names (sample names)
  show_rownames = TRUE, # Show row names (gene names)
  color_bar = TRUE, # Display color bar legend
)
```

```{r}
ELAVL_targets <- as.data.frame(readxl::read_excel(path = "C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/ELAVL_targets.xlsx"))

geneSymbols <- AnnotationDbi::mapIds(org.Hs.eg.db, keys=ELAVL_targets$`ELAVL mRNA targets`, column="SYMBOL", keytype="ALIAS", multiVals="first")
geneSymbols <- as.data.frame(geneSymbols)
ELAVL_targets$new_gene_names <- geneSymbols$geneSymbols
ELAVL_targets = ELAVL_targets[!duplicated(ELAVL_targets$new_gene_names),]
#If no official gene names are available. Exclude those from data set
ELAVL_targets <- na.omit(ELAVL_targets)
rownames(ELAVL_targets) <- ELAVL_targets$new_gene_names
ELAVL_targets <- rownames(ELAVL_targets)
ELAVL_targets <- unlist(ELAVL_targets)
```

```{r}
DEGS_VIRMA_table <- readxl::read_excel("output/VIRMA_targets_DEGs.xlsx")
Blastocyst_3utr_virma <- Blastocyst_tpm[DEGS_VIRMA_table$`Blastocyst TE VIRMA`,]
Blastocyst_3utr_virma <- na.omit(Blastocyst_3utr_virma)
matching_values <- feature_table2$external_gene_name[feature_table2$external_gene_name %in% rownames(Blastocyst_3utr_virma)]
Blastocyst_3utr_virma <- feature_table2[feature_table2$external_gene_name %in% rownames(Blastocyst_3utr_virma), ]
Blastocyst_3utr_virma <- na.omit(Blastocyst_3utr_virma)
Blastocyst_3utr_virma <- subset(Blastocyst_3utr_virma, nchar(Blastocyst_3utr_virma[,8]) > 8 & Blastocyst_3utr_virma[,8] != "Sequence unavailable" & nchar(Blastocyst_3utr_virma[,8]) <3000)
Blastocyst_3utr_virma <- subset(Blastocyst_3utr_virma, grepl("CHR_", Blastocyst_3utr_virma$chromosome_name) ==F)
rownames(Blastocyst_3utr_virma) <- make.unique(Blastocyst_3utr_virma$external_gene_name)
Blastocyst_3utr_virma_fasta <- as.data.frame(Blastocyst_3utr_virma$X3utr)
rownames(Blastocyst_3utr_virma_fasta) <- make.unique(Blastocyst_3utr_virma$external_gene_name)
seqRFLP::dataframe2fas(x = Blastocyst_3utr_virma_fasta, file = "output/virma_utr_DEGS")
```

```{r}
DEG_analysis_VIRMA <- function(df, group, loglimit, adj_p_value, output_file_name){
  group <- as.character(group)
  design <- model.matrix(~group)
  dge <- DGEList(counts = log(df+1))
  fit <- lmFit(as.matrix(dge), design)
  fit <- eBayes(fit, trend = T)
  table_dge <- topTable(fit, coef=ncol(design), number = length(rownames(df)))

  table_dge$VIRMA <- "others"

  table_dge$VIRMA[rownames(table_dge) %in% VIRMA_RIP] <- "Target of VIRMA"
  table_dge$VIRMA[rownames(table_dge) %in% ELAVL_targets] <- "Target of zELAVL"
  table_dge$VIRMA[rownames(table_dge) == "VIRMA"] <- "VIRMA"

  print(table_dge)
  table_dge$genes <- rownames(table_dge)
  table_dge$delabel <- NA
  table_dge$delabel[table_dge$VIRMA != "others"] <- table_dge$genes[table_dge$VIRMA != "others"]
  table_dge <- table_dge[order(table_dge$VIRMA, decreasing = F, na.last = T),]
  print(table_dge)
  write.table(table_dge, file.path("output", output_file_name))
  p <- ggplot(data = table_dge, aes(x= logFC, y=-log10(adj.P.Val), col= VIRMA, show.legend = F))+
    NoLegend()+
    geom_point()+
    theme_minimal()+
    #geom_text_repel(col = "black", na.rm = TRUE, hjust = 1)+
    geom_vline(xintercept=c(-loglimit, loglimit), col="grey") +
    geom_hline(yintercept=-log10(adj_p_value), col="grey") +
    scale_color_manual(values = c("grey", "darkorange", "red", "blue"))

  return(p)
}



group <- c(rep(c("Blastocyst"), times = 10 ), rep(c("Blastoid"), times =10))
design <- model.matrix(~group)
DEG_analysis_VIRMA(Blast_tpm_te_10, group =group, loglimit = 1.0, adj_p_value = 0.05, output_file_name = "DEG_TE_VIRMA")

group <- c(rep(c("Blastocyst"), times = 9 ), rep(c("Blastoid"), times =10))
design <- model.matrix(~group)
DEG_analysis_VIRMA(Blast_tpm_icm_10, group =group, loglimit = 1.0, adj_p_value = 0.05, output_file_name = "DEG_ICM_VIRMA")



group <- c(rep(c("Blastocyst"), times = 5 ), rep(c("Blastoid"), times =10))
design <- model.matrix(~group)
DEG_analysis_VIRMA(Blast_tpm_pre_10, group =group, loglimit = 1.0, adj_p_value = 0.05, output_file_name = "DEG_PrE_VIRMA")
```

```{r}
VIRMA_IGF2BP3_blastocyst <- data.frame(t(Blastocyst_tpm[c("VIRMA", "IGF2BP3"),]))
VIRMA_IGF2BP3_blastocyst[VIRMA_IGF2BP3_blastocyst==0 | VIRMA_IGF2BP3_blastocyst=="VIRMA" | VIRMA_IGF2BP3_blastocyst== "IGF2BP3"] <- NA
VIRMA_IGF2BP3_blastocyst<-VIRMA_IGF2BP3_blastocyst[complete.cases(VIRMA_IGF2BP3_blastocyst),]
VIRMA_IGF2BP3_blastocyst[VIRMA_IGF2BP3_blastocyst==0] <- NA
ggplot(VIRMA_IGF2BP3_blastocyst, aes(x=VIRMA, y=IGF2BP3))+
  geom_point()

VIRMA_IGF2BP3_blastoid <- data.frame(t(Blastoid_tpm["NANOG",]))
VIRMA_IGF2BP3_blastoid$"IGF2BP3" <- t(Blastoid_tpm["IGF2BP3",])
ggplot(VIRMA_IGF2BP3_blastoid, aes(x= NANOG, y=IGF2BP3))+
  geom_point()


```

```{r}
VIRMA_IGF2BP3 <- data.frame(t(Blast_tpm["VIRMA",]))
VIRMA_IGF2BP3$"IGF2BP3" <- t(Blast_tpm["IGF2BP3",])
ggplot(VIRMA_IGF2BP3, aes(x= VIRMA, y=IGF2BP3))+
  geom_point()+
  stat_summary(fun.data= VIRMA_IGF2BP3) + 
  geom_smooth(method='lm')+
  stat_poly_eq()

VIRMA_IGF2BP2 <- data.frame(t(Blast_tpm[c("VIRMA", "IGF2BP2"),]))
ggplot(VIRMA_IGF2BP2,aes(x= VIRMA, y=IGF2BP2))+
  geom_point()+
  stat_summary(fun.data= VIRMA_IGF2BP2) + 
  geom_smooth(method='lm')+
  stat_poly_eq()

VIRMA_IGF2BP1 <- data.frame(t(Blast_tpm[c("VIRMA", "IGF2BP1"),]))
ggplot(VIRMA_IGF2BP1, aes(x= VIRMA, y=IGF2BP1))+
  geom_point()+
  stat_summary(fun.data= VIRMA_IGF2BP1) + 
  geom_smooth(method='lm')+
  stat_poly_eq()

VIRMA_YAP1 <- data.frame(t(Blast_tpm[c("VIRMA", "YAP1"),]))
ggplot(VIRMA_YAP1, aes(x= VIRMA, y=YAP1))+
  geom_point()+
  stat_summary(fun.data= VIRMA_YAP1) + 
  geom_smooth(method='lm')+
  stat_poly_eq()

VIRMA_IGF2BPs <- data.frame(t(Blast_tpm[c("VIRMA", "IGF2BP1", "IGF2BP2", "IGF2BP3"),]))
VIRMA_IGF2BPs$sum <- rowMeans(VIRMA_IGF2BPs[c(2,3,4)])
ggplot(VIRMA_IGF2BPs, aes(x= VIRMA, y=sum))+
  geom_point()+
  stat_summary(fun.data= VIRMA_IGF2BPs) + 
  geom_smooth(method='lm')+
  stat_poly_eq()




IGF2BP3_MATR3 <- data.frame(t(Blast_tpm[c("IGF2BP3", "MATR3"),]))
ggplot(IGF2BP3_MATR3, aes(x= IGF2BP3, y=MATR3))+
  geom_point()+
  stat_summary(fun.data= IGF2BP3_MATR3) + 
  geom_smooth(method='lm')+
  stat_poly_eq()


VIRMA_ELAVL1 <- data.frame(t(Blast_tpm[c("VIRMA", "ELAVL1"),]))
ggplot(VIRMA_ELAVL1, aes(x= VIRMA, y=ELAVL1))+
  geom_point()+
  stat_summary(fun.data= VIRMA_ELAVL1) + 
  geom_smooth(method='lm')+
  stat_poly_eq()
```

```{r}

```

```{r}
library("seqRFLP")
utr_virma <- seqinr::read.fasta(file = "C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/output/mart_export.txt", as.string = T)
utr_virma <- as.data.frame(utr_virma)
utr_virma <- t(utr_virma)
sequence_lengths <- nchar(utr_virma[,1])
utr_virma <- subset(utr_virma, nchar(utr_virma[,1]) > 8 & utr_virma[,1] != "sequence unavailable")
utr_virma_upper <- utr_virma
utr_virma_upper$V1 <- toupper(utr_virma$V1)
seqRFLP::dataframe2fas(x = utr_virma_upper, file = "output/virma_utr")
devtools::install_github("https://github.com/helixcn/seqRFLP")

```

```{r}
utr_virma2 <- seqinr::read.fasta(file = "C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/output/mart_export2.txt", as.string = T)
utr_virma2 <- as.data.frame(utr_virma2)
utr_virma2 <- t(utr_virma2)
sequence_lengths <- nchar(utr_virma2[,1])
utr_virma2 <- subset(utr_virma2, nchar(utr_virma2[,1]) > 8 & utr_virma2[,1] != "sequence unavailable")
utr_virma_upper2 <- utr_virma2
utr_virma_upper2$V1 <- toupper(utr_virma2$V1)
seqRFLP::dataframe2fas(x = utr_virma_upper2, file = "output/virma_utr2")
```

```{r}
degs_tfs <- intersect(Potential_virma_TFs, degs_blastocysts)


for 


```

```{r}

mouse_genes_ZygoteMII <- readxl::read_excel("data/mouse_embryonic_ZygoteMII.xlsx")
mouse_genes_ZygoteMII <- mouse_genes_ZygoteMII[!is.na(mouse_genes_ZygoteMII$...2), ]

mouse_genes_E2CZygote <- readxl::read_excel("data/mouse_embryonic_E2CZygote.xlsx")
mouse_genes_E2CZygote <- mouse_genes_E2CZygote[!is.na(mouse_genes_E2CZygote$...2), ]

mouse_genes_M2CE2C <- readxl::read_excel("data/mouse_embryonic_M2CE2C.xlsx")
mouse_genes_M2CE2C <- mouse_genes_M2CE2C[!is.na(mouse_genes_M2CE2C$...2), ]

mouse_genes_L2CM2C <- readxl::read_excel("data/mouse_embryonic_L2CM2C.xlsx")
mouse_genes_L2CM2C <- mouse_genes_L2CM2C[!is.na(mouse_genes_L2CM2C$...2), ]

mouse_genes_4CL2C <- readxl::read_excel("data/mouse_embryonic_4C2L2C.xlsx")
mouse_genes_4CL2C <- mouse_genes_4CL2C[!is.na(mouse_genes_4CL2C$...2), ]

#Load libraries containing gene information of human and mouse
library(org.Hs.eg.db)
library(org.Mm.eg.db)
library(Orthology.eg.db)

mapIt <- function(mouseids, horg, morg, orth){
    mouseg <- mapIds(morg, mouseids, "ENTREZID", "SYMBOL")
    mapped <- select(orth, mouseg, "Homo_sapiens","Mus_musculus")
    names(mapped) <- c("Mus_egid","Homo_egid")
    husymb <- select(horg, as.character(mapped[,2]), "SYMBOL","ENTREZID")
    return(data.frame(Mus_symbol = mouseids,
                      mapped,
                      Homo_symbol = husymb[,2]))
}

humGenes <- mapIt(mouse_genes_ZygoteMII$...2, org.Hs.eg.db, org.Mm.eg.db, Orthology.eg.db)
mouse_genes_ZygoteMII$...2 <- humGenes$Homo_symbol
humGenes <- mapIt(mouse_genes_E2CZygote$...2, org.Hs.eg.db, org.Mm.eg.db, Orthology.eg.db)
mouse_genes_E2CZygote$...2<- humGenes$Homo_symbol
humGenes <- mapIt(mouse_genes_M2CE2C$...2, org.Hs.eg.db, org.Mm.eg.db, Orthology.eg.db)
mouse_genes_M2CE2C$...2 <- humGenes$Homo_symbol
humGenes <- mapIt(mouse_genes_L2CM2C$...2, org.Hs.eg.db, org.Mm.eg.db, Orthology.eg.db)
mouse_genes_L2CM2C$...2 <- humGenes$Homo_symbol
humGenes <- mapIt(mouse_genes_4CL2C$...2, org.Hs.eg.db, org.Mm.eg.db, Orthology.eg.db)
mouse_genes_4CL2C$...2 <- humGenes$Homo_symbol

DEG_analysis_mouse_stages <- function(df, loglimit, FDR, output_file_name){
  colnames(df) <- c("Gene_mouse", "Gene_human", "logFC", "FDR", "Methylation_status")
  df <- as.data.frame(df[!is.na(df$Gene_human), ])
  rownames(df) <- make.unique(df$"Gene_human")
  print(df)
  table_dge <- df
  table_dge$VIRMA <- "others"
  table_dge$VIRMA[rownames(table_dge) %in% VIRMA_RIP] <- "Target of VIRMA"
  #table_dge$VIRMA[rownames(table_dge) %in% ELAVL_targets] <- "Target of zELAVL"
  table_dge$VIRMA[rownames(table_dge) == "VIRMA"] <- "VIRMA"
  table_dge$genes <- rownames(table_dge)
  table_dge$delabel <- NA
  table_dge$delabel[table_dge$VIRMA != "others"] <- table_dge$genes[table_dge$VIRMA != "others"]
  table_dge <- table_dge[order(table_dge$VIRMA, decreasing = F, na.last = T),]
  write.table(table_dge, file.path("output", output_file_name))
  p <- ggplot(data = table_dge, aes(x= logFC, y=-log10(as.numeric(table_dge$FDR)), col= VIRMA, show.legend = F))+
    NoLegend()+
    geom_point()+
    theme_minimal()+
    #geom_text_repel(col = "black", na.rm = TRUE, hjust = 1)+
    geom_vline(xintercept=c(-loglimit, loglimit), col="grey") +
    geom_hline(yintercept=-log10(FDR), col="grey") +
    scale_color_manual(values = c("grey", "darkorange", "blue"))+
    xlim(-10, 10)+
    ylim(0, 5)

  return(p)
}
DEG_analysis_mouse_stages(mouse_genes_ZygoteMII, loglimit = 1.0, FDR = 0.05, output_file_name = "DEG_mouse_ZygoteMII")
DEG_analysis_mouse_stages(mouse_genes_E2CZygote, loglimit = 1.0, FDR = 0.05, output_file_name = "DEG_mouse_E2CZygote")
DEG_analysis_mouse_stages(mouse_genes_M2CE2C, loglimit = 1.0, FDR = 0.05, output_file_name = "DEG_mouse_M2CE2C")
DEG_analysis_mouse_stages(mouse_genes_L2CM2C, loglimit = 1.0, FDR = 0.05, output_file_name = "DEG_mouse_L2CM2C")
DEG_analysis_mouse_stages(mouse_genes_4CL2C, loglimit = 1.0, FDR = 0.05, output_file_name = "DEG_mouse_4CL2C")

```

# CollecTri

```{r}
net <- get_collectri(organism='human', split_complexes=FALSE)
Collectri_blast_integrated <- readRDS("output/acts_ulm2.RDS")
# Extract ulm and store it in tfsulm in pbmc
Blast_integrated[['tfsulm']] <- Collectri_blast_integrated %>%
  pivot_wider(id_cols = 'source', names_from = 'condition',
              values_from = 'score') %>%
  column_to_rownames('source') %>%
  Seurat::CreateAssayObject(.)



# Scale the data
Blast_integrated <- ScaleData(Blast_integrated, assay = "tfsulm")
Blast_integrated@assays$tfsulm@data <- Blast_integrated@assays$tfsulm@scale.data



```

```{r fig.width = 12}
p1 <- DimPlot(Blast_integrated, reduction = "umap", label = TRUE, pt.size = 0.5, group.by = "data_set") + 
  NoLegend() + ggtitle('Cell types')
DefaultAssay(object = Blast_integrated) <- "tfsulm"
p2 <- (FeaturePlot(Blast_integrated, features = c("YAP1"), reduction = "umap") & 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) +
  ggtitle('YAP1 activity')
DefaultAssay(object = Blast_integrated) <- "RNA"
p3 <- FeaturePlot(Blast_integrated, features = c("YAP1"), reduction = "umap") + ggtitle('YAP1 expression')
p1 | p3 | p2
```

```{r}
p1 <- DimPlot(Blast_integrated, reduction = "umap", label = TRUE, pt.size = 0.5, group.by = "data_set") + 
  NoLegend() + ggtitle('Cell types')
DefaultAssay(object = Blast_integrated) <- "tfsulm"
p2 <- (FeaturePlot(Blast_integrated, features = c("BCOR"), reduction = "umap") & 
  scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) +
  ggtitle('BCOR activity')
DefaultAssay(object = Blast_integrated) <- "RNA"
p3 <- FeaturePlot(Blast_integrated, features = c("BCOR"), reduction = "umap") + ggtitle('BCOR expression')

p4 <- FeaturePlot(Blast_integrated, features = c("VIRMA"), reduction = "umap") + ggtitle('VIRMA expression')
p1 + p4 + p3 + p2 
vp1 <- VlnPlot(Blast_integrated, assay = "tfsulm", features = c("BCOR"), split.by = "data_set") + scale_y_continuous(limits = c(0, NA))
vp2 <- VlnPlot(Blast_integrated, assay = "RNA", features = c("VIRMA", "BCOR"), split.by = "data_set")
vp1 + vp2
```

```{r}
n_tfs <- 50
#Extract activities from object as a long dataframe
df <- t(as.matrix(Blast_integrated@assays$tfsulm@data)) %>%
  as.data.frame() %>%
  mutate(cluster = Blast_integrated@meta.data$data_set) %>%
  pivot_longer(cols = -cluster, names_to = "source", values_to = "score") %>%
  group_by(cluster, source) %>%
  summarise(mean = mean(score))

# Get top tfs with more variable means across clusters
tfs <- df %>%
  group_by(source) %>%
  summarise(std = sd(mean)) %>%
  arrange(-abs(std)) %>%
  head(n_tfs) %>%
  pull(source)

# Subset long data frame to top tfs and transform to wide matrix
top_acts_mat <- df %>%
  filter(source %in% tfs) %>%
  pivot_wider(id_cols = 'cluster', names_from = 'source',
              values_from = 'mean') %>%
  column_to_rownames('cluster') %>%
  as.matrix()

# Choose color palette
palette_length = 100
my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)

my_breaks <- c(seq(-3, 0, length.out=ceiling(palette_length/2) + 1),
               seq(0.05, 3, length.out=floor(palette_length/2)))

# Plot
pheatmap(top_acts_mat, border_color = NA, color=my_color, breaks = my_breaks, fontsize = 8) 
```

```{r}
Blast_integrated_te <- subset(Blast_integrated, cell_type == "MTE" | cell_type == "TE" | cell_type == "PTE" )
#Extract activities from object as a long dataframe
df <- t(as.matrix(Blast_integrated_te@assays$tfsulm@data)) %>%
  as.data.frame() %>%
  mutate(cluster = Blast_integrated_te@meta.data$data_set) %>%
  pivot_longer(cols = -cluster, names_to = "source", values_to = "score") %>%
  group_by(cluster, source) %>%
  summarise(mean = mean(score))

# Get top tfs with more variable means across clusters
tfs <- df %>%
  group_by(source) %>%
  summarise(std = sd(mean)) %>%
  arrange(-abs(std)) %>%
  head(n_tfs) %>%
  pull(source)

# Subset long data frame to top tfs and transform to wide matrix
top_acts_mat <- df %>%
  filter(source %in% tfs) %>%
  pivot_wider(id_cols = 'cluster', names_from = 'source',
              values_from = 'mean') %>%
  column_to_rownames('cluster') %>%
  as.matrix()

# Choose color palette
palette_length = 100
my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)

my_breaks <- c(seq(-3, 0, length.out=ceiling(palette_length/2) + 1),
               seq(0.05, 3, length.out=floor(palette_length/2)))

# Plot
pheatmap(top_acts_mat, border_color = NA, color=my_color, breaks = my_breaks, fontsize = 8, main = "Activity of TFs in trophectoderm") 

Blast_integrated_icm <- subset(Blast_integrated, cell_type == "EPI" | cell_type == "ICM" )
#Extract activities from object as a long dataframe
df <- t(as.matrix(Blast_integrated_icm@assays$tfsulm@data)) %>%
  as.data.frame() %>%
  mutate(cluster = Blast_integrated_icm@meta.data$data_set) %>%
  pivot_longer(cols = -cluster, names_to = "source", values_to = "score") %>%
  group_by(cluster, source) %>%
  summarise(mean = mean(score))

# Get top tfs with more variable means across clusters
tfs <- df %>%
  group_by(source) %>%
  summarise(std = sd(mean)) %>%
  arrange(-abs(std)) %>%
  head(n_tfs) %>%
  pull(source)

# Subset long data frame to top tfs and transform to wide matrix
top_acts_mat <- df %>%
  filter(source %in% tfs) %>%
  pivot_wider(id_cols = 'cluster', names_from = 'source',
              values_from = 'mean') %>%
  column_to_rownames('cluster') %>%
  as.matrix()

# Choose color palette
palette_length = 100
my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)

my_breaks <- c(seq(-3, 0, length.out=ceiling(palette_length/2) + 1),
               seq(0.05, 3, length.out=floor(palette_length/2)))

# Plot
pheatmap(top_acts_mat, border_color = NA, color=my_color, breaks = my_breaks, fontsize = 8,  main = "Activity of TFs in ICM")
```
