---
title: "Blastocyst_vs_Blastoid"
author: "Tara Urselmann"
date: "2024-03-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup

## Intro

```{r}
#This research will focus on uncovering major limitations in the blastoid model compared to the blastocyst. 
#Enjoy. 
#If any questions arise, please don't hesitate to contact me : taraurselmann@gmail.com
```

## Libraries

```{r Library, results='hide', echo = TRUE, error=FALSE, warning=FALSE, message=FALSE}
#loading of all necessary libraries
library(workflowr) 
library(here) 
library(Seurat) 
library(SeuratDisk) 
library(RColorBrewer) 
library(ggplot2) 
library(dplyr) 
library(scran) 
library(kableExtra)
library(progeny)
library(ggplot2)
library(tidyr)
library(readr)
library(pheatmap)
library(tibble)
library(ggrepel)
library(biomaRt)
library(biomartr)
library(seqinr)
library(limma)
library(edgeR)
library(org.Hs.eg.db)
library(decoupleR)
library(OmnipathR)
library(EnvStats)
library(ggpmisc)
library(GOplot)
library(ggpubr)

```

```{r Work directory, results='hide', echo = TRUE, error=FALSE, warning=FALSE, message=FALSE}
#wflow_git_config(user.name = "taraurselmann",user.email = "taraurselmann@gmail.com") 
#fill in your own path to directory

wflow_start("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/", existing = TRUE)
set_here(path = "C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/")
```

## Functions

```{r limma pipeline}

#Write a function for analyzing this data set with limma pipeline to obtain DEGs
#Note: The input of DGElist is count, since that is not available I use the TPM data.
#used this thread as guideline: https://support.bioconductor.org/p/98820/
DEG_analysis <- function(df, group, loglimit, adj_p_value, output_file_name, main){
  group <- as.character(group) #Very important that group is defined correctly by user, as it will compare both groups later
  design <- model.matrix(~group) 
  dge <- DGEList(counts = log1p(df)) #A log transformation used following te guidline
  fit <- lmFit(as.matrix(dge), design) #Linear fit
  fit <- eBayes(fit, trend = T) 
  table_dge <- topTable(fit, coef=ncol(design), number = length(rownames(df)))
  table_dge$diff <- "NO" # Make new column 
  
  # if log2Foldchange > loglimit and pvalue < p_value, set as "UP" 
  table_dge$diff[table_dge$logFC > loglimit & table_dge$adj.P.Val < adj_p_value] <- "UP"
  # if log2Foldchange < -loglimit and pvalue < p_value, set as "DOWN"
  table_dge$diff[table_dge$logFC < -loglimit & table_dge$adj.P.Val < adj_p_value] <- "DOWN"
  
  table_dge$genes <- rownames(table_dge)
  table_dge$delabel <- NA
  table_dge$delabel[table_dge$diff != "NO"] <- table_dge$genes[table_dge$diff != "NO"] #Only use labels of DEG genes

  write.table(table_dge, file.path("output", output_file_name)) #Write an excel file containing DEGs to output file
  p <- ggplot(data = table_dge, aes(x= logFC, y=-log10(adj.P.Val), col= diff, show.legend = F))+
    NoLegend()+
    geom_point()+
    theme_minimal()+
    geom_vline(xintercept=c(-loglimit, loglimit), col="red") +
    geom_hline(yintercept=-log10(adj_p_value), col="red") +
    scale_color_manual(values = c("darkblue", "lightgrey", "darkred"), guide = "none")+
    ggtitle(main)+
    xlab("logFC(Blastoid/Blastocyst)")
  return(p)
}


```

```{r}
#Make a similar volcanoplot as described above, but highlight VIRMA targets
DEG_analysis_VIRMA <- function(df, group, loglimit, adj_p_value, main){
  group <- as.character(group)
  design <- model.matrix(~group)
  dge <- DGEList(counts = log1p(df)) #Log transform according to guideline
  fit <- lmFit(as.matrix(dge), design)
  fit <- eBayes(fit, trend = T)
  table_dge <- topTable(fit, coef=ncol(design), number = length(rownames(df))) #Get complete table
  table_dge$diff <- "NO"
  # if log2Foldchange > loglimit and pvalue < p_value, set as "UP" 
  table_dge$VIRMA <- "others"

  table_dge$VIRMA[rownames(table_dge) %in% VIRMA_binding_RNAs] <- "Target of VIRMA"
  table_dge$VIRMA[rownames(table_dge) == "VIRMA"] <- "VIRMA"
  table_dge$genes <- rownames(table_dge)
  table_dge$delabel <- NA
  table_dge$delabel[table_dge$VIRMA == "VIRMA"] <- table_dge$genes[table_dge$VIRMA == "VIRMA"] #Make new column that highlights VIRMA targets
  table_dge <- table_dge[order(table_dge$VIRMA, decreasing = F, na.last = T),]
  print(table_dge)
  p_virma <- ggplot(data = table_dge, aes(x= logFC, y=-log10(adj.P.Val), col= VIRMA, show.legend = F, label = delabel))+
    NoLegend()+
    geom_point()+
    theme_minimal()+
    geom_text_repel(col = "black", na.rm = TRUE, hjust = 3, size = 3)+
    geom_vline(xintercept=c(-loglimit, loglimit), col="grey") +
    geom_hline(yintercept=-log10(adj_p_value), col="grey") +
    scale_color_manual(values = c("grey", "coral", "cornflowerblue", "red"), guide = "none")+
    #xlab("logFC (Blastoid/Blastocyst)")+
    ggtitle(main)

  return(p_virma)
}

```

# Loading data

## Blastocyst data

```{r load data in directory}
#Get all separate files in directory, every file contains a sample
all_files <- list.files("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/GSE133200_RAW/", full.names=TRUE, recursive = T)

ldf <- lapply(all_files, read.csv) #Read all files in directory
all_files_list <- lapply(ldf, summary)
names(all_files_list) <- substr(all_files, 1, 1000000000) #Get all names of files in directory 
```

```{r reference}
#This loop construction is used to open all files in directory
for (i in names(all_files_list)){
  if (i == names(all_files_list[1])){ #Use the first file to make a new dataframe
  file1 <- read.csv(i, sep = "")
  TPM_df <- file1}
  
  else { #For all subsequent files add them to the previously generated dataframe
  file2 <- read.csv(i, sep = "")
  TPM_df <- full_join(TPM_df, file2, by = join_by(Gene))}
}

rownames(TPM_df) <- TPM_df[,1] #Use correct rownames
TPM_df <- TPM_df[-c(1)] #Delete column used for rownames
Blastocyst_data <- TPM_df 


#The reference list is used to link the right name to the sample
reference_list <- readr::read_tsv("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_tpm/data/reference_list.txt")

cell_type_df <- TPM_df
for (i in 1:length(colnames(cell_type_df))){
  if (colnames(cell_type_df[i]) %in% reference_list$old){
    row_index <- which(reference_list == colnames(cell_type_df[i]))
    new_name <- reference_list$new[row_index]
    names(cell_type_df)[names(cell_type_df) == colnames(cell_type_df[i])] <- new_name}
    else {}
    
}

#Samples are deleted if not in reference file 
col_delete <- c()
for (i in 1:length(colnames(cell_type_df))){ #Loop through dataframe and check if names are in reference file
  if (grepl("hE", colnames(cell_type_df[i])) == T){
    col_delete <- append(col_delete, i)}}

cell_type_df <- cell_type_df[,-col_delete] #Delete names that are not in reference file

#Rename file to Blastocyst data as the data containing only blastocyst data will be named with that term
Blastocyst_data <- cell_type_df

```

### Rename gene symbols

```{r renaming aliases to gene symbols}
#Because this file contains ALIAS symbols the symbols need to be converted to official gene symbols to compare later
geneSymbols <- AnnotationDbi::mapIds(org.Hs.eg.db, keys=rownames(Blastocyst_data), column="SYMBOL", keytype="ALIAS", multiVals="first")
geneSymbols <- as.data.frame(geneSymbols)
Blastocyst_data$new_gene_names <- geneSymbols$geneSymbols #Make a new column in dataframe with correct symbols
Blastocyst_data = Blastocyst_data[!duplicated(Blastocyst_data$new_gene_names),] #Remove rows that have duplicated symbols
Blastocyst_data <- na.omit(Blastocyst_data) #If no official gene names are available. Exclude those from data set
rownames(Blastocyst_data) <- Blastocyst_data$new_gene_names #Change rownames to new gene symbols
Blastocyst_data <- subset(Blastocyst_data, select = -c(new_gene_names)) #Remove column with gene names

Blastocyst_data <- Blastocyst_data[rowSums(Blastocyst_data) >5, ] 
```

## Blastoid data

```{r Import Blastoid data}
#Import blastoid data of Kagawa
Blastoid_raw <- readRDS("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/blastoid.H9.okae_bts5__scRNAseq.unfiltered__counts.RDS")
Blastoid_meta <- read_tsv("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/blastoid.H9.okae_bts5__scRNAseq.unfiltered__metadata.tsv")

#Choose to delete rows with the same gene name, but different ensembl ids
Blastoid_raw[c("Ensembl_id", "gene_name")] <- stringr::str_split_fixed(Blastoid_raw$gene_id, '-', 2)
Blastoid_raw = Blastoid_raw[!duplicated(Blastoid_raw$gene_name),]
rownames(Blastoid_raw) <- Blastoid_raw$gene_name
Blastoid_raw <- Blastoid_raw[, !names(Blastoid_raw) %in% c("gene_id", "gene_name", "Ensembl_id"), drop = F] #remove redundant columns

#Add meta data 
Blastoid_meta <- Blastoid_meta %>%
                  dplyr::mutate(Blastoid_meta, time = case_when(
                  grepl("24h",samplename) ~ "24h",
                  grepl("60h",samplename) ~ "60h",
                  grepl("96h",samplename) ~ "96h",
                  TRUE ~ gsub("-.*","",samplename)
             )) %>% data.frame()
rownames(Blastoid_meta) <- Blastoid_meta$sampleid
mtname="^ENSG[[:digit:]]+-MT-"

#Remove unviable cells. Use the same requirements used previously. Pretty standard. 
Blastoid_meta_filtered <- subset(Blastoid_meta, nFeature_RNA > 2000 & percent.mt <12.5)
Blastoid_raw <- Blastoid_raw[, rownames(Blastoid_meta_filtered)]
Blastoid_raw <- Blastoid_raw[rowSums(Blastoid_raw) >5, ] #Genes should be measured in at least 5 times in all cells

#Retrieve the annotation provided by Cheyenne de Vriend
annotated_list <- LoadH5Seurat("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/BlastoidTutorial/data/SeuratProject_Blastoid_annotated.h5Seurat")
Blastoid_data <- Blastoid_raw
annotation <- as.data.frame(annotated_list@active.ident) #Transfer annotation from seurat object
colnames(annotation) <- "active_id"
annotation$active_id <- as.character(annotation$active_id)


annotation$"cluster_name" <- annotated_list@meta.data$samplename


for (row in 1:length(rownames(annotation))){
  if (grepl("transitioning", annotation$active_id[row]) == T){
    if (grepl("24h", annotation$cluster_name[row]) == T){
      annotation$active_id[row] <- paste(annotation$active_id[row],"24h")}
    if (grepl("60h", annotation$cluster_name[row]) == T){
      annotation$active_id[row] <- paste(annotation$active_id[row],"60h")}
    if (grepl("96h", annotation$cluster_name[row]) == T){
      annotation$active_id[row] <- paste(annotation$active_id[row],"96h")}
    if (grepl("naive", annotation$cluster_name[row]) == T){
      annotation$active_id[row] <- paste(annotation$active_id[row],"naive")}
    }
  }



#Use a loop to annotate the right cells
ncol(Blastoid_data)
col_delete <- c()
for (i in 1:length(colnames(Blastoid_data))){
  if (colnames(Blastoid_data[i]) %in% rownames(annotation)){
  row_index <- which(rownames(annotation) == colnames(Blastoid_data[i]))
  new_name <- as.character(annotation$active_id[row_index])
  names(Blastoid_data)[names(Blastoid_data) == colnames(Blastoid_data[i])] <- new_name}
  else{
  col_delete <- append(col_delete, i)
  }
}
Blastoid_data <- Blastoid_data[,-col_delete] #delete columns that aren't annotated 
ncol(Blastoid_data)
```

```{r Gene length information}

#ENSEMBL database is not always working correcly depending on server availability, so the code below is to illustrate how I got all the necessary features. But for future purposes the table can be extracted from the output file. 


#Get feature lengths/ ensembl id
# command for downgrading
#devtools::install_version("dbplyr", version = "2.3.4")



#Retrieve the table of features from output file
feature_table_genes <- read.csv("output/Features_genes.csv")
feature_table_genes_2 <- as.data.frame(feature_table_genes$external_gene_name) #extract only gene names from file
colnames(feature_table_genes_2) <- "gene_name" 


#Add length to dataframe, and make it kbps
feature_table_genes_2$length <- (feature_table_genes$transcript_length/1000)  #transcript length is in bps
ensembl_id_lengths_blastoid <- Blastoid_data #create df for transformation
```

### RPK

```{r transform to RPK}
#Create RPK, so reads per kilobase

lengths_vector <- setNames(feature_table_genes_2$length, feature_table_genes_2$gene_name) # Create a named vector of lengths for quick lookup
common_ids <- intersect(rownames(ensembl_id_lengths_blastoid), names(lengths_vector)) # Identify common IDs


# Subset data frame to common IDs and perform the transformation
ensembl_id_lengths_blastoid[common_ids,] <- ensembl_id_lengths_blastoid[common_ids,] / lengths_vector[common_ids]
ensembl_id_lengths_blastoid <- ensembl_id_lengths_blastoid[common_ids,]
Blastoid_data <- ensembl_id_lengths_blastoid #Rename dataframe to old name

```

### TPM

```{r transform to TPM}
#Normalize data for sequencing depth
for (i in 1:length(colnames(Blastoid_data))){
  sum_of_sample <- sum(Blastoid_data[,i])
  factor <- sum_of_sample/1000000 #Transform for transcripts per million
  Blastoid_data[,i] <- (Blastoid_data[,i]/factor)
  
}

#Now both datasets are in TPM format and will be named as such
Blastoid_tpm <- Blastoid_data
Blastocyst_tpm <- Blastocyst_data
```

## Join datasets

```{r Join both datasets}
#combining the data set is required to compare later on
#It is important that the data sets compare only the available genes as the Blastocyst data already excluded some genes --> inner_join
#Add column of rownames in both data sets to join them later on
Blastocyst_tpm$gene <- rownames(Blastocyst_tpm)
Blastoid_tpm$gene <- rownames(Blastoid_tpm)

#join datasets. If both datasets are present with a datastructure (Blastocyst and Blastoid) the term "Blast" will be used. 
Blast_tpm <- inner_join(Blastocyst_tpm, Blastoid_tpm)
Blast_tpm <- column_to_rownames(Blast_tpm, var = "gene") #Rename rownames

#Create a subset of both data set containing only trophectoderm
#Blastoid data only 60 and 96h as they form a cluster together
#Blastocyst data only S2 as coculturing with ISK cells could potentially alter results
col_delete <- c()
for (i in 1:length(colnames(Blast_tpm))){
  if (grepl("PTE 96h", colnames(Blast_tpm[i])) == F & 
      grepl("TE 96h", colnames(Blast_tpm[i])) == F & 
      grepl("S2.MTE", colnames(Blast_tpm[i])) == F & 
      grepl("S2.PTE", colnames(Blast_tpm[i])) == F){
    col_delete <- append(col_delete, i)}
}

Blast_tpm_te <- Blast_tpm[,-col_delete] #All columns that are not TE are deleted for this new df


```

```{r Retrieve VIRMA binding RNAs}
VIRMA_binding_RNAs <- readxl::read_xlsx("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/mRNA_VIRMA.xlsx") #Get document that has mRNA binding targets from BIOGRID

rownames(VIRMA_binding_RNAs) <- VIRMA_binding_RNAs$mRNA
#Because this file contains ALIAS symbols the symbols need to be converted to official gene symbols to compare later
geneSymbols <- AnnotationDbi::mapIds(org.Hs.eg.db, 
                                     keys=rownames(VIRMA_binding_RNAs), 
                                     column="SYMBOL", 
                                     keytype="ALIAS", 
                                     multiVals="first")
geneSymbols <- as.data.frame(geneSymbols)
VIRMA_binding_RNAs$new_gene_names <- geneSymbols$geneSymbols
VIRMA_binding_RNAs = VIRMA_binding_RNAs[!duplicated(VIRMA_binding_RNAs$new_gene_names),]


#If no official gene names are available. Exclude those from data set
VIRMA_binding_RNAs <- na.omit(VIRMA_binding_RNAs)
rownames(VIRMA_binding_RNAs) <- VIRMA_binding_RNAs$new_gene_names
VIRMA_binding_RNAs <- rownames(VIRMA_binding_RNAs) #Get only the gene names as they are relevant for further research
VIRMA_binding_RNAs <- unlist(VIRMA_binding_RNAs)


write.csv(VIRMA_binding_RNAs, "output/VIRMA_RNA_correct.csv")  #Save the symbols, for easy access
```

## Finding VIRMA motifs

# Analysis of data

## Regression of dataset

### Trophectoderm

```{r regression te}

#Get a list with Blastocyst or Blastoid in Blast_te dataframe
new_name = list()
for (i in 1:length(colnames(Blast_tpm_te))){
  if (grepl("S2", colnames(Blast_tpm_te)[i]) == T){ #Use S2 term as marker for Blastocyst data
    new_name <- append(new_name, "Blastocyst")
  }else{ #If not S2 in name, then name it as Blastoid.
    new_name <- append(new_name, "Blastoid")
  }
}

#Make new regression dataframe
Blast_regression <- Blast_tpm_te
colnames(Blast_regression) <- new_name


Blastoid_regression <- Blast_regression[ , grep("^Blastoid"   , names(Blast_regression))]
Blastocyst_regression <-  Blast_regression[ , grep("^Blastocyst"   , names(Blast_regression))]

#Make dataframe containing the mean value of all genes and log transform
regression_df <- data.frame(Blastoid = rowMeans(log(Blastoid_regression+1)), 
                            Blastocyst = rowMeans(log(Blastocyst_regression+1)))


rownames(regression_df) <- rownames(Blast_regression)
regression_df$delabel <- rownames(regression_df) #Create label names for graph purposes

regression_te <- ggplot(regression_df, aes(x= Blastoid, y = Blastocyst))+
  geom_point(size = 1, show.legend = T)+
  stat_summary(fun.data= regression_df) + 
  geom_smooth(method='lm')+
  scale_x_continuous(expand = c(0, 0)) + 
  scale_y_continuous(expand = c(0, 0), limits = c(0, 10))+
  stat_poly_eq()+
  theme_bw()+
  xlab("Blastoid gene expression")+
  ylab("Blastocyst gene expression")+
  ggtitle("Regression Trophectoderm")
regression_te  

```

### EPI/ICM

```{r EPI/ICM}
#Do the same for EPI/ICM as was done for the TE samples. 

col_delete <- c()
for (i in 1:length(colnames(Blast_tpm))){
  if (grepl("EPI", colnames(Blast_tpm[i])) ==F & 
      grepl("S2.EPI", colnames(Blast_tpm[i])) == F & 
      grepl("ICM", colnames(Blast_tpm[i])) == F | 
      grepl("AF", colnames(Blast_tpm[i]) == F)){
    col_delete <- append(col_delete, i)}
}
Blast_tpm_icm <- Blast_tpm[,-col_delete]

new_name = list()
for (i in 1:length(colnames(Blast_tpm_icm))){
  if (grepl("S2", colnames(Blast_tpm_icm)[i]) == T | grepl("S1", colnames(Blast_tpm_icm)[i]) == T | grepl("S3", colnames(Blast_tpm_icm)[i]) == T){
    new_name <- append(new_name, "Blastocyst")
  }else{
    new_name <- append(new_name, "Blastoid")
  }
}

Blast_regression_icm <- Blast_tpm_icm
colnames(Blast_regression_icm) <- new_name
Blastoid_regression_icm <- Blast_regression_icm[ , grep("^Blastoid"   , names(Blast_regression_icm))]
Blastocyst_regression_icm <-  Blast_regression_icm[ , grep("^Blastocyst"   , names(Blast_regression_icm))]

regression_df_icm <- data.frame(Blastoid = rowMeans(log(Blastoid_regression_icm+1)), Blastocyst = rowMeans(log(Blastocyst_regression_icm+1)))
rownames(regression_df_icm) <- rownames(Blast_regression_icm)

regression_icm <- ggplot(regression_df_icm, aes(x= regression_df_icm$Blastoid, y = regression_df_icm$Blastocyst))+
  geom_point(size = 1)+
  stat_summary(fun.data= regression_df_icm) + 
  geom_smooth(method='lm')+
  scale_x_continuous(expand = c(0, 0)) + 
  scale_y_continuous(expand = c(0, 0), limits = c(0, 10))+
  stat_poly_eq()+
  theme_bw()+
  xlab("Blastoid gene expression")+
  ylab("Blastocyst gene expression")+
  ggtitle("Regression Inner cell mass")


regression_icm
```

### Primitive endoderm

```{r PrE}

#Do the same for the primitive endoderm
col_delete <- c()
for (i in 1:length(colnames(Blast_tpm))){
  if (grepl("PRE", colnames(Blast_tpm[i])) ==F & 
      grepl("PE", colnames(Blast_tpm[i])) == F){
    col_delete <- append(col_delete, i)}
}
Blast_tpm_pre <- Blast_tpm[,-col_delete]

new_name = list()
for (i in 1:length(colnames(Blast_tpm_pre))){
  if (grepl("PE", colnames(Blast_tpm_pre)[i]) == T){
    new_name <- append(new_name, "Blastocyst")
  }else{
    new_name <- append(new_name, "Blastoid")
  }
}

Blast_regression_pre <- Blast_tpm_pre
colnames(Blast_regression_pre) <- new_name
Blastoid_regression_pre <- Blast_regression_pre[ , grep("^Blastoid"   , names(Blast_regression_pre))]
Blastocyst_regression_pre <-  Blast_regression_pre[ , grep("^Blastocyst"   , names(Blast_regression_pre))]

regression_df_pre <- data.frame(Blastoid = rowMeans(log(Blastoid_regression_pre+1)), Blastocyst = rowMeans(log(Blastocyst_regression_pre+1)))
rownames(regression_df_pre) <- rownames(Blast_regression_pre)

regression_pre <- ggplot(regression_df_pre, aes(x= regression_df_pre$Blastoid, y = regression_df_pre$Blastocyst))+
  geom_point(size = 1)+
  stat_summary(fun.data= regression_df_pre) + 
  geom_smooth(method='lm')+
  scale_x_continuous(expand = c(0, 0)) + 
  scale_y_continuous(expand = c(0, 0), limits = c(0, 10))+
  stat_poly_eq()+
  theme_bw()+
  xlab("Blastoid gene expression")+
  ylab("Blastocyst gene expression")+
  ggtitle("Regression Primitive endoderm")

regression_pre

```

```{r overrview all regression warning=FALSE, fig.width = 12}
regression_te + regression_icm + regression_pre #Plot together as an overview
```

## Limma

### Trophectoderm

```{r 10 groups of 10}

#To minimize noise the blastocyst and blastoid will be separated by TE only. Subsequently the columns will be randomized to minimize bias, and than they will be clustered in ~10 groups. 

#Obtain only TE from Blastocyst
col_delete <- c()
for (i in 1:length(colnames(Blastocyst_tpm))){
  if (grepl("S2.MTE", colnames(Blastocyst_tpm[i])) == F & 
      grepl("S2.PTE", colnames(Blastocyst_tpm[i])) == F){
    col_delete <- append(col_delete, i)}
}
Blastocyst_tpm_te <- Blastocyst_tpm[,-col_delete]


#Randomize column order
v <- c(1:length(Blastocyst_tpm_te))
set.seed(001)
v <- sample(v)
Blastocyst_tpm_random <- Blastocyst_tpm_te[v]

#Divide in to ~10 groups by making 10 groups
Blastocyst_tpm_10 <- as.data.frame(rowMeans(Blastocyst_tpm_random[0:47]))
colnames(Blastocyst_tpm_10) <- c("1a")
Blastocyst_tpm_10$'2a' <- rowMeans(Blastocyst_tpm_random[48:94])
Blastocyst_tpm_10$'3a' <- rowMeans(Blastocyst_tpm_random[95:141])
Blastocyst_tpm_10$'4a' <- rowMeans(Blastocyst_tpm_random[142:188])
Blastocyst_tpm_10$'5a' <- rowMeans(Blastocyst_tpm_random[189:235])
Blastocyst_tpm_10$'6a' <- rowMeans(Blastocyst_tpm_random[236:282])
Blastocyst_tpm_10$'7a' <- rowMeans(Blastocyst_tpm_random[283:329])
Blastocyst_tpm_10$'8a' <- rowMeans(Blastocyst_tpm_random[330:376])
Blastocyst_tpm_10$'9a' <- rowMeans(Blastocyst_tpm_random[377:423])
Blastocyst_tpm_10$'10a' <- rowMeans(Blastocyst_tpm_random[424:470])
Blastocyst_tpm_10$"gene" <- rownames(Blastocyst_tpm_10)

#Obtain only TE from Blastoid
col_delete <- c()
for (i in 1:length(colnames(Blastoid_tpm))){
  if (grepl("PTE 96h", colnames(Blastoid_tpm[i])) ==F & 
      grepl("TE 96h", colnames(Blastoid_tpm[i])) ==F){
    col_delete <- append(col_delete, i)}
}
Blastoid_tpm_te <- Blastoid_tpm[,-col_delete]


v <- c(1:length(Blastoid_tpm_te))
v <- sample(v)
Blastoid_tpm_random <- Blastoid_tpm_te[v]
Blastoid_tpm_10 <- as.data.frame(rowMeans(Blastoid_tpm_random[1:37]))
colnames(Blastoid_tpm_10) <- c("1b")
Blastoid_tpm_10$'2b' <- rowMeans(Blastoid_tpm_random[38:74])
Blastoid_tpm_10$'3b' <- rowMeans(Blastoid_tpm_random[75:111])
Blastoid_tpm_10$'4b' <- rowMeans(Blastoid_tpm_random[112:148])
Blastoid_tpm_10$'5b' <- rowMeans(Blastoid_tpm_random[149:185])
Blastoid_tpm_10$'6b' <- rowMeans(Blastoid_tpm_random[186:222])
Blastoid_tpm_10$'7b' <- rowMeans(Blastoid_tpm_random[223:259])
Blastoid_tpm_10$'8b' <- rowMeans(Blastoid_tpm_random[260:296])
Blastoid_tpm_10$'9b' <- rowMeans(Blastoid_tpm_random[297:333])
Blastoid_tpm_10$'10b' <- rowMeans(Blastoid_tpm_random[334:375])
Blastoid_tpm_10$"gene" <- rownames(Blastoid_tpm_10)

Blast_tpm_te_10 <- inner_join(Blastocyst_tpm_10, Blastoid_tpm_10)
Blast_tpm_te_10 <- column_to_rownames(Blast_tpm_te_10, var = "gene")
group <- c(rep(c("Blastocyst"), times = 10 ), 
           rep(c("Blastoid"), times =10))
design <- model.matrix(~group)

DEG_plot_te <- DEG_analysis(Blast_tpm_te_10, group =group, loglimit = 1.5, adj_p_value = 0.05, output_file_name = "DEG_TE", main = "Trophectoderm")
DEG_plot_te
```

### ICM/EPI

```{r 10 groups of 10, EPIBLAST}

#To minimize noise the blastocyst and blastoid will be separated by epi only. Subsequently the columns will be randomized to minimize bias, and than they will be clustered in ~10 groups. 
#Obtain only epi from Blastocyst
col_delete <- c()
for (i in 1:length(colnames(Blastocyst_tpm))){
  if (grepl("ICM", colnames(Blastocyst_tpm[i])) == F & 
      grepl("EPI", colnames(Blastocyst_tpm[i])) == F |
      grepl("AF", colnames(Blastocyst_tpm[i])) == T){
    col_delete <- append(col_delete, i)}
}
Blastocyst_tpm_icm <- Blastocyst_tpm[,-col_delete]


#Randomize column order
v <- c(1:length(Blastocyst_tpm_icm))
set.seed(001)
v <- sample(v)
Blastocyst_tpm_icm_random <- Blastocyst_tpm_icm[v]
#Divide in to ~10 groups by making 10 groups

Blastocyst_tpm_icm_10 <- as.data.frame(rowMeans(Blastocyst_tpm_icm_random[1:7]))
Blastocyst_tpm_icm_10$'2a' <- rowMeans(Blastocyst_tpm_icm_random[8:15])
Blastocyst_tpm_icm_10$'3a' <- rowMeans(Blastocyst_tpm_icm_random[16:23])
Blastocyst_tpm_icm_10$'4a' <- rowMeans(Blastocyst_tpm_icm_random[24:31])
Blastocyst_tpm_icm_10$'5a' <- rowMeans(Blastocyst_tpm_icm_random[32:39])
Blastocyst_tpm_icm_10$'6a' <- rowMeans(Blastocyst_tpm_icm_random[40:47])
Blastocyst_tpm_icm_10$'7a' <- rowMeans(Blastocyst_tpm_icm_random[48:55])
Blastocyst_tpm_icm_10$'8a' <- rowMeans(Blastocyst_tpm_icm_random[56:63])
Blastocyst_tpm_icm_10$'9a' <- rowMeans(Blastocyst_tpm_icm_random[64:71])
Blastocyst_tpm_icm_10$'10a' <- rowMeans(Blastocyst_tpm_icm_random[72:77])
Blastocyst_tpm_icm_10$"gene" <- rownames(Blastocyst_tpm_icm_10)

#Obtain only epi from Blastoid
col_delete <- c()
for (i in 1:length(colnames(Blastoid_tpm))){
  if (grepl("EPI", colnames(Blastoid_tpm[i])) ==F & 
      grepl("ICM", colnames(Blastoid_tpm[i])) ==F){
    col_delete <- append(col_delete, i)}
}
Blastoid_tpm_icm <- Blastoid_tpm[,-col_delete]


v <- c(1:length(Blastoid_tpm_icm))
v <- sample(v)
Blastoid_tpm_icm_random <- Blastoid_tpm_icm[v]
Blastoid_tpm_icm_10 <- as.data.frame(rowMeans(Blastoid_tpm_icm_random[1:55]))
Blastoid_tpm_icm_10$'2b' <- rowMeans(Blastoid_tpm_icm_random[56:110])
Blastoid_tpm_icm_10$'3b' <- rowMeans(Blastoid_tpm_icm_random[111:165])
Blastoid_tpm_icm_10$'4b' <- rowMeans(Blastoid_tpm_icm_random[166:220])
Blastoid_tpm_icm_10$'5b' <- rowMeans(Blastoid_tpm_icm_random[221:275])
Blastoid_tpm_icm_10$'6b' <- rowMeans(Blastoid_tpm_icm_random[276:330])
Blastoid_tpm_icm_10$'7b' <- rowMeans(Blastoid_tpm_icm_random[331:385])
Blastoid_tpm_icm_10$'8b' <- rowMeans(Blastoid_tpm_icm_random[386:440])
Blastoid_tpm_icm_10$'9b' <- rowMeans(Blastoid_tpm_icm_random[441:495])
Blastoid_tpm_icm_10$'10b' <- rowMeans(Blastoid_tpm_icm_random[496:550])
Blastoid_tpm_icm_10$"gene" <- rownames(Blastoid_tpm_icm_10)

Blast_tpm_icm_10 <- inner_join(Blastocyst_tpm_icm_10, Blastoid_tpm_icm_10)
Blast_tpm_icm_10 <- column_to_rownames(Blast_tpm_icm_10, var = "gene")


group <- c(rep(c("Blastocyst"), times = 10), 
           rep(c("Blastoid"), times =10))
design <- model.matrix(~group)

DEG_plot_icm <- DEG_analysis(Blast_tpm_icm_10, group =group, loglimit = 1.5, adj_p_value = 0.05, output_file_name = "DEG_ICM", main = "Epiblast")
DEG_plot_icm
```

### Primitive endoderm

```{r 10 groups of 10, PrE}

#To minimize noise the blastocyst and blastoid will be separated by PRE only. Subsequently the columns will be randomized to minimize bias, and than they will be clustered in ~10 groups. 
#Obtain only PRE from Blastocyst
col_delete <- c()
for (i in 1:length(colnames(Blastocyst_tpm))){
  if (grepl("PE", colnames(Blastocyst_tpm[i])) == F |
      grepl("AF", colnames(Blastocyst_tpm[i])) == T){
    col_delete <- append(col_delete, i)}
}
Blastocyst_tpm_pre <- Blastocyst_tpm[,-col_delete]


#Randomize column order
v <- c(1:length(Blastocyst_tpm_pre))
set.seed(001)
v <- sample(v)
Blastocyst_tpm_pre_random <- Blastocyst_tpm_pre[v]
#Divide in to ~10 groups by making 10 groups

Blastocyst_tpm_pre_10 <- as.data.frame(rowMeans(Blastocyst_tpm_pre_random[1:9]))
Blastocyst_tpm_pre_10$'2a' <- rowMeans(Blastocyst_tpm_pre_random[10:18])
Blastocyst_tpm_pre_10$'3a' <- rowMeans(Blastocyst_tpm_pre_random[19:27])
Blastocyst_tpm_pre_10$'4a' <- rowMeans(Blastocyst_tpm_pre_random[28:36])
Blastocyst_tpm_pre_10$'5a' <- rowMeans(Blastocyst_tpm_pre_random[37:45])
Blastocyst_tpm_pre_10$"gene" <- rownames(Blastocyst_tpm_pre_10)

#Obtain only PRE from Blastoid
col_delete <- c()
for (i in 1:length(colnames(Blastoid_tpm))){
  if (grepl("PRE", colnames(Blastoid_tpm[i])) ==F & 
      grepl("PE", colnames(Blastoid_tpm[i])) ==F){
    col_delete <- append(col_delete, i)}
}
Blastoid_tpm_pre <- Blastoid_tpm[,-col_delete]


v <- c(1:length(Blastoid_tpm_pre))
v <- sample(v)
Blastoid_tpm_pre_random <- Blastoid_tpm_pre[v]
Blastoid_tpm_pre_10 <- as.data.frame(rowMeans(Blastoid_tpm_pre_random[1:26]))
Blastoid_tpm_pre_10$'2b' <- rowMeans(Blastoid_tpm_pre_random[27:52])
Blastoid_tpm_pre_10$'3b' <- rowMeans(Blastoid_tpm_pre_random[53:78])
Blastoid_tpm_pre_10$'4b' <- rowMeans(Blastoid_tpm_pre_random[79:104])
Blastoid_tpm_pre_10$'5b' <- rowMeans(Blastoid_tpm_pre_random[105:130])
Blastoid_tpm_pre_10$'6b' <- rowMeans(Blastoid_tpm_pre_random[131:156])
Blastoid_tpm_pre_10$'7b' <- rowMeans(Blastoid_tpm_pre_random[157:182])
Blastoid_tpm_pre_10$'8b' <- rowMeans(Blastoid_tpm_pre_random[183:208])
Blastoid_tpm_pre_10$'9b' <- rowMeans(Blastoid_tpm_pre_random[209:234])
Blastoid_tpm_pre_10$'10b' <- rowMeans(Blastoid_tpm_pre_random[235:260])
Blastoid_tpm_pre_10$"gene" <- rownames(Blastoid_tpm_pre_10)

Blast_tpm_pre_10 <- inner_join(Blastocyst_tpm_pre_10, Blastoid_tpm_pre_10)
Blast_tpm_pre_10 <- column_to_rownames(Blast_tpm_pre_10, var = "gene")
group <- c(rep(c("Blastocyst"), times = 5 ), 
           rep(c("Blastoid"), times =10))
design <- model.matrix(~group)

DEG_plot_pre <- DEG_analysis(Blast_tpm_pre_10, group =group, loglimit = 1.5, adj_p_value = 0.05, output_file_name = "DEG_PRE", main = "Primitive endoderm")
DEG_plot_pre
```

### All together now

```{r}
Blastocyst_tpm_all_1 <- Blastocyst_tpm_icm_random #Get EPI df
Blastocyst_tpm_all_1$"gene" <- rownames(Blastocyst_tpm_icm_random) #Get gene name as column to combine later
Blastocyst_tpm_all_2 <- Blastocyst_tpm_pre_random #Get PRE df
Blastocyst_tpm_all_2$"gene" <- rownames(Blastocyst_tpm_pre_random)
Blastocyst_tpm_all_3 <- Blastocyst_tpm_random #Get TE df
Blastocyst_tpm_all_3$"gene" <- rownames(Blastocyst_tpm_random)
Blastocyst_tpm_all <- inner_join(Blastocyst_tpm_all_1, Blastocyst_tpm_all_2) #Join the three dfs together using genes
Blastocyst_tpm_all <- inner_join(Blastocyst_tpm_all, Blastocyst_tpm_all_3)
Blastocyst_tpm_all <- column_to_rownames(Blastocyst_tpm_all, var = "gene") #Put gene names back as rows 

#randomize the columns
set.seed(001)
v <- c(1:length(Blastocyst_tpm_all))
v <- sample(v)

#Cluster the columns in to 10 groups
Blastocyst_tpm_all <- Blastocyst_tpm_all[v]
Blastocyst_tpm_all_10 <- as.data.frame(rowMeans(Blastocyst_tpm_all[1:59]))
Blastocyst_tpm_all_10$'2a' <- rowMeans(Blastocyst_tpm_all[60:118])
Blastocyst_tpm_all_10$'3a' <- rowMeans(Blastocyst_tpm_all[119:177])
Blastocyst_tpm_all_10$'4a' <- rowMeans(Blastocyst_tpm_all[178:236])
Blastocyst_tpm_all_10$'5a' <- rowMeans(Blastocyst_tpm_all[237:295])
Blastocyst_tpm_all_10$'6a' <- rowMeans(Blastocyst_tpm_all[296:354])
Blastocyst_tpm_all_10$'7a' <- rowMeans(Blastocyst_tpm_all[355:413])
Blastocyst_tpm_all_10$'8a' <- rowMeans(Blastocyst_tpm_all[414:472])
Blastocyst_tpm_all_10$'9a' <- rowMeans(Blastocyst_tpm_all[473:531])
Blastocyst_tpm_all_10$'10a' <- rowMeans(Blastocyst_tpm_all[532:590])
Blastocyst_tpm_all_10$"gene" <- rownames(Blastocyst_tpm_all)

#Do the same for the blastoid
Blastoids_tpm_all_1 <- Blastoid_tpm_icm_random
Blastoids_tpm_all_1$"gene" <- rownames(Blastoid_tpm_icm_random)
Blastoids_tpm_all_2 <- Blastoid_tpm_pre_random
Blastoids_tpm_all_2$"gene" <- rownames(Blastoid_tpm_pre_random)
Blastoids_tpm_all_3 <- Blastoid_tpm_random
Blastoids_tpm_all_3$"gene" <- rownames(Blastoid_tpm_random)
Blastoids_tpm_all <- inner_join(Blastoids_tpm_all_1, Blastoids_tpm_all_2)
Blastoids_tpm_all <- inner_join(Blastoids_tpm_all, Blastoids_tpm_all_3)
Blastoids_tpm_all <- column_to_rownames(Blastoids_tpm_all, var = "gene")
v <- c(1:length(Blastoids_tpm_all))

v <- sample(v)
Blastoids_tpm_all <- Blastoids_tpm_all[v]

Blastoid_tpm_all_10 <- as.data.frame(rowMeans(Blastoids_tpm_all[1:119]))
Blastoid_tpm_all_10$'2b' <- rowMeans(Blastoids_tpm_all[120:238])
Blastoid_tpm_all_10$'3b' <- rowMeans(Blastoids_tpm_all[239:357])
Blastoid_tpm_all_10$'4b' <- rowMeans(Blastoids_tpm_all[358:476])
Blastoid_tpm_all_10$'5b' <- rowMeans(Blastoids_tpm_all[477:595])
Blastoid_tpm_all_10$'6b' <- rowMeans(Blastoids_tpm_all[596:714])
Blastoid_tpm_all_10$'7b' <- rowMeans(Blastoids_tpm_all[715:833])
Blastoid_tpm_all_10$'8b' <- rowMeans(Blastoids_tpm_all[834:952])
Blastoid_tpm_all_10$'9b' <- rowMeans(Blastoids_tpm_all[953:1071])
Blastoid_tpm_all_10$'10b' <- rowMeans(Blastoids_tpm_all[1072:1190])
Blastoid_tpm_all_10$"gene" <- rownames(Blastoids_tpm_all)


Blast_tpm_all_10 <- inner_join(Blastocyst_tpm_all_10, Blastoid_tpm_all_10)
Blast_tpm_all_10 <- column_to_rownames(Blast_tpm_all_10, var = "gene")
group <- c(rep(c("Blastocyst"), times = 10 ), rep(c("Blastoid"), times =10))
design <- model.matrix(~group)

DEG_plot_all <- DEG_analysis(Blast_tpm_all_10, group =group, loglimit = 1.5, adj_p_value = 0.05, output_file_name = "DEG_ALL", main = "All celltypes" )
DEG_plot_all


```

```{r}
#Make same plot with  indication of VIRMA targets 
DEG_plot_all_VIRMA <- DEG_analysis_VIRMA(Blast_tpm_all_10, 
                                         group =group, 
                                         loglimit = 1.5, 
                                         adj_p_value = 0.05, 
                                         main = "All")
DEG_plot_all_VIRMA
```

```{r MTE/PTE}
#Did a comparison between MTE and PTE. Could be useful in the future, but not necessary for this research

# #To minimize noise the blastocyst and blastoid will be separated by TE only. Subsequently the columns will be randomized to minimize bias, and than they will be clustered in ~10 groups. 
# #Obtain only TE from Blastocyst
# col_delete <- c()
# for (i in 1:length(colnames(Blastocyst_tpm))){
#   if (grepl("S2.MTE", colnames(Blastocyst_tpm[i])) == F){
#     col_delete <- append(col_delete, i)}
# }
# Blastocyst_tpm_mte <- Blastocyst_tpm[,-col_delete]
# 
# 
# #Randomize column order
# v <- c(1:length(Blastocyst_tpm_mte))
# set.seed(001)
# v <- sample(v)
# Blastocyst_tpm_mte_random <- Blastocyst_tpm_mte[v]
# #Divide in to ~10 groups by making 10 groups
# 
# Blastocyst_tpm_mte_10 <- as.data.frame(rowMeans(Blastocyst_tpm_mte_random[1:38]))
# colnames(Blastocyst_tpm_mte_10) <- c("1a")
# Blastocyst_tpm_mte_10$'2a' <- rowMeans(Blastocyst_tpm_mte_random[39:76])
# Blastocyst_tpm_mte_10$'3a' <- rowMeans(Blastocyst_tpm_mte_random[77:114])
# Blastocyst_tpm_mte_10$'4a' <- rowMeans(Blastocyst_tpm_mte_random[115:152])
# Blastocyst_tpm_mte_10$'5a' <- rowMeans(Blastocyst_tpm_mte_random[153:190])
# Blastocyst_tpm_mte_10$'6a' <- rowMeans(Blastocyst_tpm_mte_random[191:228])
# Blastocyst_tpm_mte_10$'7a' <- rowMeans(Blastocyst_tpm_mte_random[229:266])
# Blastocyst_tpm_mte_10$'8a' <- rowMeans(Blastocyst_tpm_mte_random[267:304])
# Blastocyst_tpm_mte_10$'9a' <- rowMeans(Blastocyst_tpm_mte_random[305:342])
# Blastocyst_tpm_mte_10$'10a' <- rowMeans(Blastocyst_tpm_mte_random[343:380])
# Blastocyst_tpm_mte_10$"gene" <- rownames(Blastocyst_tpm_mte_10)
# 
# #Obtain only TE from Blastoid
# col_delete <- c()
# for (i in 1:length(colnames(Blastocyst_tpm))){
#   if (grepl("S2.PTE", colnames(Blastocyst_tpm[i])) == F){
#     col_delete <- append(col_delete, i)}
# }
# Blastocyst_tpm_pte <- Blastocyst_tpm[,-col_delete]
# 
# 
# v <- c(1:length(Blastocyst_tpm_pte))
# v <- sample(v)
# Blastocyst_tpm_pte_random <- Blastocyst_tpm_pte[v]
# Blastocyst_tpm_pte_10 <- as.data.frame(rowMeans(Blastocyst_tpm_pte_random[1:16]))
# colnames(Blastocyst_tpm_pte_10) <- c("1b")
# Blastocyst_tpm_pte_10$'2b' <- rowMeans(Blastocyst_tpm_pte_random[17:32])
# Blastocyst_tpm_pte_10$'3b' <- rowMeans(Blastocyst_tpm_pte_random[33:48])
# Blastocyst_tpm_pte_10$'4b' <- rowMeans(Blastocyst_tpm_pte_random[49:64])
# Blastocyst_tpm_pte_10$'5b' <- rowMeans(Blastocyst_tpm_pte_random[65:80])
# Blastocyst_tpm_pte_10$"gene" <- rownames(Blastocyst_tpm_pte_10)
# 
# Blast_tpm_te_ptemte_10 <- inner_join(Blastocyst_tpm_mte_10, Blastocyst_tpm_pte_10)
# Blast_tpm_te_ptemte_10 <- column_to_rownames(Blast_tpm_te_ptemte_10, var = "gene")
# Blast_tpm_te_ptemte_10 <- log((Blast_tpm_te_ptemte_10+1))
# group <- c(rep(c("MTE"), times = 10 ), rep(c("PTE"), times =5))
# design <- model.matrix(~group)
# 
# DEG_plot_MTE_PTE_blastocyst <- DEG_analysis(Blast_tpm_te_ptemte_10, group =group, loglimit = 0.5, adj_p_value = 0.05, output_file_name = "DEG_MTE_PTE_blastocyst", main = "x")
# DEG_plot_MTE_PTE_blastocyst
```

### VIRMA DEGs

```{r fig.width = 15, fig.height=5}
group <- c(rep(c("Blastocyst"), times = 10), 
           rep(c("Blastoid"), times =10))
design <- model.matrix(~group)

#A plot of virma targets of TE comparison
DEG_plot_te_VIRMA <- DEG_analysis_VIRMA(Blast_tpm_te_10, 
                                        group =group, 
                                        loglimit = 1.5, 
                                        adj_p_value = 0.05, 
                                        main = "Trophectoderm")
DEG_plot_te_VIRMA


#A plot of virma targets of EPI comparison
DEG_plot_icm_VIRMA <- DEG_analysis_VIRMA(Blast_tpm_icm_10, 
                                         group =group, 
                                         loglimit = 1.5, 
                                         adj_p_value = 0.05, 
                                         main = "Epiblast")
DEG_plot_icm_VIRMA


#A plot of virma targets of all cell comparison
DEG_plot_all_VIRMA <- DEG_analysis_VIRMA(Blast_tpm_all_10, 
                                         group =group, 
                                         loglimit = 1.5, 
                                         adj_p_value = 0.05, 
                                         main = "All celltypes")

DEG_plot_all_VIRMA

#A plot of virma targets of PRE comparison
#primitive endoderm has less clusters, so other group specification is needed
group <- c(rep(c("Blastocyst"), times = 5), 
           rep(c("Blastoid"), times =10))
design <- model.matrix(~group)

DEG_plot_pre_VIRMA <- DEG_analysis_VIRMA(Blast_tpm_pre_10, group =group, loglimit = 1.5, adj_p_value = 0.05, main = "Primitive endoderm")
DEG_plot_pre_VIRMA

```

# Seurat

```{r seurat}
Blast_tpm_seurat <- CreateSeuratObject(Blast_tpm) #Create seurat object of blast_tpm

name_list <- rownames(as.data.frame(Blast_tpm_seurat@active.ident)) 
Blast_tpm_seurat@meta.data$names <- name_list #Save names in meta data



cell_layers = list() #List for new names regarding there cell type
cystoid <- list() #Binary list: Blastocyst or Blastoid

#Loop to create new names categories depending on original name (Cell type and data_set(Blastocyst/Blastoid))
for(i in Blast_tpm_seurat@meta.data$names){
  if (grepl("S1", i) | 
      grepl("S2", i) | 
      grepl("S3", i) | 
      grepl("ISK", i) | 
      grepl("AF", i)== T){
    cystoid <- append(cystoid, "Blastocyst") #Blastocyst name, so add to Blastocyst data_set. Further specify name for cell type below
    if (grepl("AF", i) == T){
      cell_layers <- append(cell_layers, "Blastocyst-AF")}
    else if (grepl("MTE", i) == T){
      cell_layers <- append(cell_layers, "Blastocyst-MTE")}
    else if (grepl("EPI", i) == T){
      cell_layers <- append(cell_layers, "Blastocyst-EPI")}
    else if (grepl("ICM", i) == T){
      cell_layers <- append(cell_layers, "Blastocyst-ICM")}    
    else if (grepl("ISK", i) == T){
      cell_layers <-append(cell_layers, "Blastocyst-ISK")}
    else if (grepl("PTE", i) == T){
      cell_layers <-append(cell_layers, "Blastocyst-PTE")}
    else if (grepl("PE", i) == T){
      cell_layers <- append(cell_layers, "Blastocyst-PrE")}
    else if (grepl("S1.TE", i) == T){
      cell_layers <-append(cell_layers, "Blastocyst-TE")}}
  else{
    cystoid <- append(cystoid, "Blastoid") #Blastoid name, so add to Blastocyst data_set. Further specify name for cell type below
    if (grepl("PTE", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-PTE")}
    else if (grepl("TE", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-TE")}    
    else if (grepl("ICM", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-ICM")}
    else if (grepl("PRE", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-PrE")}
    else if (grepl("Naive", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-Naive")}
    else if (grepl("EPI", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-EPI")}
     else if (grepl("Unknown", i) == T){
      cell_layers <- append(cell_layers, "Blastoid-Unknown")}
    else{print(i)}}}

cell_layers <- unlist(cell_layers)
Blast_tpm_seurat@meta.data$cell_layers <- cell_layers

#Set a random seed number, and choose an appropriate amount of variable genes. This is pretty standard, there was no reason to change this.
set.seed(1234)
varGene = 3000

#Because it is a TPM data table. Seurat works a bit differently. Used recommended workflow from Seurat (https://github.com/satijalab/seurat/issues/7496)
Blast_tpm_seurat <- SetAssayData(Blast_tpm_seurat, slot = 'data', new.data = log1p(Blast_tpm))
Blast_tpm_seurat <- FindVariableFeatures(object = Blast_tpm_seurat, nfeatures = varGene)
Blast_tpm_seurat <- ScaleData(Blast_tpm_seurat)
Blast_tpm_seurat <- RunPCA(Blast_tpm_seurat)
Blast_tpm_seurat <- FindNeighbors(Blast_tpm_seurat)
Blast_tpm_seurat <- FindClusters(Blast_tpm_seurat, resolution = 0.5, verbose=FALSE)
Blast_tpm_seurat <- RunUMAP(Blast_tpm_seurat, dims = 1:20, n.components = 3L, min.dist = 0.5, verbose=FALSE)


#plot of cell layer names
Idents(Blast_tpm_seurat) <- Blast_tpm_seurat@meta.data$cell_layers
DimPlot(Blast_tpm_seurat)
```

### Controls

```{r Control of cell markers}
#Check for markers related to trophectoderm, inner cell mass, and primitive endoderm.
cell_type_markers <- c("GATA2", "GATA3", "POU5F1", "KLF17", "GATA4", "SOX17")

#A dotplot is nice visualization tool to quickly see if the groups are as expected.
DotPlot(Blast_tpm_seurat, features = cell_type_markers) + RotatedAxis()+scale_colour_gradient2(low="blue", mid="white", high="red")
```

## PROGENy

### PROGENy of Trophectoderm

```{r PROGENy}
#Progeny can be used to assess the activation or inhibition of different pathways 

#To compare TE between the data set they are subsetted
Blast_tpm_seurat_te <-  subset(Blast_tpm_seurat, 
                               cell_layers == "Blastocyst-MTE" | 
                               cell_layers == "Blastocyst-PTE" | 
                               cell_layers == "Blastoid-PTE"| 
                               cell_layers == "Blastoid-TE")

Blast_tpm_seurat_te_clusters <- data.frame(Cell = names(Idents(Blast_tpm_seurat_te)), 
                                           CellType = as.character(Idents(Blast_tpm_seurat_te)), 
                                           stringsAsFactors = F)

#Use PROGENy workflow as described in their vignette. 
Blast_tpm_seurat_te_progeny <- progeny(Blast_tpm_seurat_te, 
                                       scale = F, 
                                       organism = "Human", 
                                       top = 500, 
                                       perm = 1, 
                                       return_assay = T)

Blast_tpm_seurat_te_progeny <- ScaleData(Blast_tpm_seurat_te_progeny, assay = "progeny") #Make a seperate assay in seurat

#Make a new dataframe for progeny
BCP_df <- as.data.frame(t(GetAssayData(Blast_tpm_seurat_te_progeny, slot = "scale.data", assay = "progeny")))%>%
  rownames_to_column("Cell") %>%
  gather(Pathway, Activity, -Cell)

#Transform data according to vignette
BCP_df <- inner_join(BCP_df, Blast_tpm_seurat_te_clusters)
sum_bcp_df <- BCP_df %>%
  group_by(Pathway, CellType) %>%
  summarise(avg = mean(Activity), std=sd(Activity))

sum_bcp_df <- sum_bcp_df %>%
  dplyr::select(-std) %>%
  spread(Pathway, avg) %>%
  data.frame(row.names = 1, check.names = F, stringsAsFactors = F)

#Choose colors. Standard to PROGENy, can be altered if preferred
paletteLength = 100
myColor = colorRampPalette(c("darkblue", "white","red"))(paletteLength)

progenyBreaks = c(seq(min(sum_bcp_df), 0, 
                      length.out=ceiling(paletteLength/2) + 1),
                  seq(max(sum_bcp_df)/paletteLength, 
                      max(sum_bcp_df), 
                      length.out=floor(paletteLength/2)))
progeny_hmap = pheatmap(t(sum_bcp_df),fontsize=14, 
                        fontsize_row = 10, 
                        color=myColor, breaks = progenyBreaks, 
                        main = "Pathway analysis Blastoid vs. Blastocyst", angle_col = 45,
                        treeheight_col = 0,  border_color = NA)
```

```{r cell cycle, fig.width=10, fig.height =6}
#It could be interesting to check if all cells from one data set or one cell type are within the cell cycle stage. 

#Load cell cycle associated genes
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

#Use cellcyclescoring function from seurat to plot cell cycle scores in the seurat object
Blast_tpm_seurat <- CellCycleScoring(Blast_tpm_seurat, s.features = s.genes, g2m.features = g2m.genes, set.ident =T)

Idents(Blast_tpm_seurat) <- Blast_tpm_seurat@meta.data$cell_layers
Cluster_dim <- DimPlot(Blast_tpm_seurat) #Plot with the cell layers and data set

Idents(Blast_tpm_seurat) <- Blast_tpm_seurat@meta.data$Phase
Cell_cycle_dim <- DimPlot(Blast_tpm_seurat) #plot with cell cycle scores

Cluster_dim + Cell_cycle_dim
```

## Integration

```{r integration of data_set}
data_group = list() #Make a list that can collect either blastocyst or blastoid
#Make a loop that either add blastocyst or blastoid to the list
for (i in Blast_tpm_seurat$cell_layers){
  if (grepl("Blastoid", i) == T){
    data_group = append(data_group, "Blastoid")
  } else if (grepl("Blastocyst", i) == T){
    data_group = append(data_group, "Blastocyst")
  } else (print(i))
}
data_group = unlist(data_group)
Blast_tpm_seurat@meta.data$data_set = data_group #Add new names to meta data in seurat object


cell_type = list() #Make a list to collect the appropriate cell types (e.g. TE, EPI, ISK)
#Loop through cell layers and collect the respective cell type
for (i in Blast_tpm_seurat$cell_layers){
  if (grepl("AF", i) == T){
    cell_type = append(cell_type, "AF")
  } else if (grepl("MTE", i) == T){
  cell_type = append(cell_type, "MTE")
  } else if (grepl("PTE", i) == T){
    cell_type = append(cell_type, "PTE")
  } else if (grepl("TE", i) == T){
    cell_type = append(cell_type, "TE")
  } else if (grepl("PrE", i) == T){
    cell_type = append(cell_type, "PrE")
  } else if (grepl("Naive", i) == T){
    cell_type = append(cell_type, "Naive")
  } else if (grepl("EPI", i) == T){
    cell_type = append(cell_type, "EPI")
  } else if (grepl("ICM", i) == T){
    cell_type = append(cell_type, "ICM")
  } else if (grepl("ISK", i) == T){
    cell_type = append(cell_type, "ISK")
  } else if (grepl("Unknown", i) == T){
    cell_type = append(cell_type, "Unknown")
  }}

cell_type = unlist(cell_type)
Blast_tpm_seurat@meta.data$cell_type = cell_type #Add cell type to meta data of seurat object

Blast_integrated <- Blast_tpm_seurat #Create a new seurat object in which layers will be split

Blast_integrated[["RNA"]] <- split(Blast_integrated[["RNA"]], 
                                   f =Blast_integrated$data_set) #Use data_set annotation to split between to groups
Blast_integrated <- NormalizeData(Blast_integrated, scale.factor = 1) #Because TPM is used, no normalization has been performed. This does not work with layers. So normalization is performed. But will later be overwritten with correct TPM data.

Blast_integrated@assays[["RNA"]]@layers[["data.Blastocyst"]] <- as(log1p(Blast_integrated@assays[["RNA"]]@layers[["counts.Blastocyst"]]), "sparseMatrix") #Overwrite data slot (normalization) with counts slot (TPM data)
Blast_integrated@assays[["RNA"]]@layers[["data.Blastoid"]] <- as(log1p(Blast_integrated@assays[["RNA"]]@layers[["counts.Blastoid"]]), "sparseMatrix") #Overwrite data slot (normalization) with counts slot (TPM data)

#Follow the integration vignette of Seurat
Blast_integrated <- FindVariableFeatures(Blast_integrated)
Blast_integrated <- ScaleData(Blast_integrated)
Blast_integrated <- RunPCA(Blast_integrated)
Blast_integrated <- FindNeighbors(Blast_integrated, dims = 1:30, reduction = "pca")
Blast_integrated <- FindClusters(Blast_integrated, resolution = 2, cluster.name = "unintegrated_clusters")
Blast_integrated <- RunUMAP(Blast_integrated, dims = 1:30, reduction = "pca", reduction.name = "umap.unintegrated")
DimPlot(Blast_integrated, reduction = "umap.unintegrated", group.by = "data_set")
Blast_integrated <- IntegrateLayers(object = Blast_integrated, method = CCAIntegration, orig.reduction = "pca", new.reduction = "integrated.cca", verbose = FALSE, assay = "RNA")

Blast_integrated[["RNA"]] <- JoinLayers(Blast_integrated[["RNA"]])
Blast_integrated <- RunPCA(Blast_integrated)
Blast_integrated <- FindNeighbors(Blast_integrated, reduction = "integrated.cca", dims = 1:30)
Blast_integrated <- FindClusters(Blast_integrated, resolution = 1)
Blast_integrated <- RunUMAP(Blast_integrated, dims = 1:30, reduction = "integrated.cca")
plot_integration <- DimPlot(Blast_integrated, reduction = "umap", group.by = "data_set")
#plot_integration_cells <- DimPlot(Blast_integrated, reduction = "umap", group.by = "cell_type")
plot_integration_cells <-  DimPlot(Blast_integrated, reduction = "umap", group.by = "cell_type", cols = c( "#c5ca30", "#ff924c", "#ffca3a","#1982c4" ,  "#36949d","#8ac926","#ff595e","#565aa0" ,"cornflowerblue","#6a4c99" ), pt.size = 1)

#plots of seurat object split by data set and cell type
plot_integration + plot_integration_cells


```

### Time

```{r split seurat on time, fig.width = 10, fig.height = 4}

time_list = list() #Make a list to collect associated timepoints
#a loop to collect respective timepoints
for (i in Blast_integrated$names){
  if (grepl("96h", i) == T){
    time_list = append(time_list, "96h")
  } else if (grepl("S1", i) == T){
    time_list = append(time_list, "S1")
  } else if (grepl("S2", i) == T){
    time_list = append(time_list, "S2")
  } else if (grepl("S3", i) == T){
    time_list = append(time_list, "S3")
  } else if (grepl("AF", i) == T){
    time_list = append(time_list, "AF")
  } else if (grepl("60h", i) == T){
    time_list = append(time_list, "60h")
  } else if (grepl("24h", i) == T){
    time_list = append(time_list, "24h")
  } else if (grepl("ISK", i) == T){
    time_list = append(time_list, "ISK")
  } else if (grepl("Naive", i) == T){
    time_list = append(time_list, "Naive")
  } else if (grepl("Unknown", i) == T){
    time_list = append(time_list, "Unknown")
  } else if (grepl("transitioning", i) == T){
    time_list = append(time_list, "transitioning")
  } else {print(i)}}
  
  
time_list = unlist(time_list)
Blast_integrated@meta.data$times = time_list #Add time list to add to meta data

#Make two plots that compare times and cell types
plot_1 <- DimPlot(Blast_integrated, reduction = "umap", group.by = "cell_type", cols = c( "#c5ca30", "#ff924c", "#ffca3a","#1982c4" ,  "#36949d","#8ac926","#ff595e","purple" ,"cornflowerblue","#6a4c99" ), shape.by = "data_set", pt.size = 0.5)

plot_2 <- DimPlot(Blast_integrated, reduction = "umap", group.by = "times", cols = c( "#c5ca30", "#ff924c", "#ffca3a","#1982c4" ,  "#36949d","#8ac926","#ff595e","purple" ,"cornflowerblue","#6a4c99", "grey" ), shape.by = "data_set", pt.size = 0.5)
plot_1+ plot_2

```

```{r checks and subsets}
#Subset seurat object with only annotated and usefull cell types
Blast_integrated <- subset(Blast_integrated, 
                           cell_type == "EPI" | 
                           cell_type == "ICM" | 
                           cell_type == "MTE" | 
                           cell_type == "PTE" | 
                           cell_type == "TE" | 
                           cell_type == "PrE" | 
                           cell_type == "Naive")

Blast_integrated <- FindClusters(Blast_integrated, resolution = 1)
DimPlot(Blast_integrated, reduction ="umap")

#Make another dotplot to check markers
DotPlot(Blast_integrated, features = cell_type_markers) + 
  RotatedAxis()+
  scale_colour_gradient2(low="blue", mid="white", high="red")

#Check if the graphs look similar
plot_1 <- DimPlot(Blast_integrated, 
                  reduction = "umap", 
                  group.by = "cell_type", 
                  cols = c( "#6a4c99", "#ff924c", "#ffca3a","#1982c4" ,  "#36949d","#8ac926","#ff595e","purple" ,"cornflowerblue","#6a4c99" ), 
                  pt.size = 0.5)

#Plot the different plots to compare and check
plot_1
DimPlot(Blast_integrated, 
        reduction = "umap", 
        group.by = "data_set", 
        pt.size = 0.8)
```

### Markers cell layers

```{r feature plots markers}

#Feature plot of PrE markers
FeaturePlot(Blast_integrated, 
            features = c("SOX17", "GATA4"), 
            split.by = "data_set", 
            max.cutoff = 3, 
            cols = c("grey", "red"), 
            reduction = "umap")

#Feature plot of TE markers
FeaturePlot(Blast_integrated, 
            features = c("GATA2", "GATA3"), 
            split.by = "data_set", 
            max.cutoff = 3, 
            cols = c("grey", "red"), 
            reduction = "umap")

#Feature plot of inner cell mass markers
FeaturePlot(Blast_integrated, 
            features = c("KLF17", "POU5F1"), 
            split.by = "data_set", 
            max.cutoff = 3, 
            cols = c("grey", "red"), 
            reduction = "umap")


```

# VIRMA

In this section the influence of the VIRMA and its related proteins will be analyzed

## Explorative

### Featureplot VIRMA

```{r featureplot virma}
FeaturePlot(Blast_integrated, features = c("VIRMA"), pt.size = 0.5, split.by = "data_set", max.cutoff = 3, reduction = "umap")
```

### Module score VIRMA targets

```{r Module score plot}
#Create a module score based on the VIRMA binding RNAs
VIRMA_binding_targets_in_seurat <- as.list(intersect(rownames(Blast_integrated), VIRMA_binding_RNAs))
write.csv(VIRMA_binding_targets_in_seurat, "output/VIRMA_binding_seurat")
Blast_integrated <- AddModuleScore(
 object = Blast_integrated,
 features = VIRMA_binding_targets_in_seurat, #Select only RNAs present in Seurat object
 ctrl = 100,
 name = "VIRMA_features"
)

#Create a feature plot with the module score. 
feature_VIRMA <- FeaturePlot(Blast_integrated, 
                             features = "VIRMA_features1", 
                             label = F, 
                             repel = T, 
                             reduction = "umap") +  
                 scale_colour_gradient2(low="blue", mid="white", high="red")

#Print feature plot together with a dimplot to compare data sets
Dim_VIRMA <- DimPlot(Blast_integrated, reduction = "umap", group.by = "data_set")
feature_VIRMA + Dim_VIRMA
```

### Cell cycle

```{r cell cycle VIRMA}
#As VIRMA regulation should not be limited to cell cycle if it is responsible for a real difference between blastocyst and blastoid, the cell cycle is checked again in this new seurat object. 
Blast_integrated <- CellCycleScoring(Blast_integrated, s.features = s.genes, g2m.features = g2m.genes, set.ident =T)

#Plots plots plots
Idents(Blast_integrated) <- Blast_integrated@meta.data$data_set
Cluster_dim <- DimPlot(Blast_integrated, reduction = "umap") #Plot of different data_sets
Idents(Blast_integrated) <- Blast_integrated@meta.data$Phase
Cell_cycle_dim <- DimPlot(Blast_integrated, reduction = "umap") #Plot of different cell cycles
Cluster_dim + Cell_cycle_dim + feature_VIRMA

#set idents to correct cell layers
Idents(Blast_integrated) <- Blast_integrated@meta.data$cell_layers
```

### VIRMA interactors

```{r Plots of interacting proteins}
#Plot of CPSF5 and CPSF6, related to polyadenylation. https://www.nature.com/articles/s41421-018-0019-0
VlnPlot(Blast_integrated, split.by = "data_set", features = c("CEBPA"))
VlnPlot(Blast_integrated, split.by = "data_set", features = c("NUDT21", "CPSF6"))

#Plot of m6a readers associated with stabilization
VlnPlot(Blast_integrated, split.by = "data_set", features = c("IGF2BP1", "IGF2BP2", "IGF2BP3"))
```

```{r IGF2BP module score}
#To visualize the complete picture of IGF2BPs, a visualization of IGF2BPs as module score is created. 


IGF2BP_list = list(c("IGF2BP1", "IGF2BP2", "IGF2BP3")) #Make a list of all IGF2BPs

Blast_integrated <- AddModuleScore(
  object = Blast_integrated,
 features = IGF2BP_list,
 ctrl = 5,
 name = "IGF2BP_features"
) #Add modulescore to correct seurat object 

#Featureplot of IGF2BP module score
feature_IGF2BP <- FeaturePlot(Blast_integrated, features = "IGF2BP_features1", label = F, repel = T, reduction = "umap") +scale_colour_gradient2(low="blue", mid="white", high="red")
#Split general feature plot on data sets to highlight which cells belong to which group
Dim_IGF2BP <- DimPlot(Blast_integrated, reduction = "umap", group.by = "data_set")

#Print featureplots to highlight the overlap between IGF2BPs, VIRMA and the Blastocyst data set.
feature_IGF2BP +feature_VIRMA +Dim_IGF2BP


#Visualization in form of DotPlot
features <- c("VIRMA", "IGF2BP1", "IGF2BP2", "IGF2BP3")
Idents(Blast_integrated) <- Blast_integrated@meta.data$cell_layers #Set idents to cell layers
DotPlot(Blast_integrated, features = features) + RotatedAxis() +scale_colour_gradient2(low="blue", mid="white", high="red")
```

### Methyl writer complex

```{r Methyl writer complex}
Methyl_writer_complex <- c("VIRMA", "METTL3", "METTL14", "WTAP", "CBLL1", "ZC3H13") #Genes encoding for proteins part of methyl writer complex

#Two visualizations: Heatmap, and DotPlot
DoHeatmap(Blast_integrated, features = features, group.by = "cell_layers", slot = "data", label = F)+scale_fill_gradient(low = "white", high = "black")
DotPlot(Blast_integrated, features = features) + RotatedAxis() +scale_colour_gradient2(low="blue", mid="white", high="red")
FeaturePlot(Blast_integrated, features = "WTAP", reduction = "umap")
```

#### Other METTL genes

```{r other METTLs}
#Other METTL genes could also be of interest. See how they are distributed.
METTL_gene_list_nm <- c("METTL1", "METTL2A", "METTL2B", "METTL3", "METTL4", "METTL5", "METTL6", "METTL8", "METTL14", "METTL15", "METTL16", "METTL19", "METTL25B", "TIA1", "G3BP1", "DDX1") #List DNA/RNA modifying METTLs
METTL_gene_list_pm <- c("METTL9", "METTL10", "METTL11A", "METTL11B", "METTL12", "METTL13", "METTL18", "METTL20",  "METTL21A", "METTL21B", "METTL21C", "METTL21D", "METTL21E", "METTL22", "METTL23") #List of protein modifying METTLs
METTL_gene_list_other <- c("METTL7A", "METTL7B", "METTL17", "METTL25", "METTL26", "METTL27") #List of other METTLs

DoHeatmap(Blast_integrated, features = METTL_gene_list_nm, group.by = "cell_layers", size = 3, slot = "data")+theme(text = element_text(size = 5))+NoLegend() #Heatmap nucleotide modifying

DoHeatmap(Blast_integrated, features = METTL_gene_list_pm, group.by = "cell_layers", size = 3, slot = "data")+theme(text = element_text(size = 5))+NoLegend() #Heatmap protein modifying

DoHeatmap(Blast_integrated, features = METTL_gene_list_other, group.by = "cell_layers", size = 3, slot = "data")+theme(text = element_text(size = 5))+NoLegend() #Heatmap other METTLs modifying

```

## Regulation of VIRMA

### Transcription factors

```{r Potential VIRMA TFs, fig.width= 15}
Potential_virma_TFs <- read.csv("data/TFBS.csv") #Read file that contains TF that could be associated with VIRMA
Potential_virma_TFs <- unlist(Potential_virma_TFs)
DoHeatmap(Blast_integrated, features = Potential_virma_TFs, group.by = "cell_layers", size = 3, slot = "data")+theme(text = element_text(size = 5))+NoLegend() #Heatmap to look for potential candidates

DEGS_All <- read.csv("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/output/DEG_ALL", sep = " ") #Load in DEG values of blastoid vs blastocyst

TF_VIRMA <- DEGS_All[Potential_virma_TFs, ] #Extract rows containing candidate TFs
TF_VIRMA <- TF_VIRMA[order(TF_VIRMA$logFC, decreasing = F, na.last = T),] #Order on logFC 
TF_VIRMA <- rownames(head(TF_VIRMA, 20))  #Extract top FC genes

#Look if patterns are similar to VIRMA in a plot
DotPlot(Blast_integrated, features = c("VIRMA", TF_VIRMA)) + RotatedAxis() +scale_colour_gradient2(low="blue", mid="white", high="red")
```

#### Regression TFs

```{r Potential VIRMA TFs regression}
#A calculation for all top TFs that could interact with the VIRMA gene. To find if one has a high correlation with VIRMA via regression. 
for (gene in TF_VIRMA){
  if (gene %in% rownames(Blastocyst_tpm_te)){
    df <- data.frame(t(Blastocyst_tpm_te[c("VIRMA", gene),])) #Make a df containing VIRMA and a TF of blastocyst
    df <- as.data.frame(sapply(df, as.numeric))
    df[df==0] <- NA #Remove the cells that contain 0 of VIRMA or 0 of the TF
    df[complete.cases(df), ]
    df <- na.omit(df) 
    #Make a plot with a regression line
    plot <- ggplot(df, aes(x= df$VIRMA, y=df[,2]))+
                geom_point()+
                stat_summary(fun.data= df) + 
                geom_smooth(method='lm')+
                stat_poly_eq()+
                ylab(gene)
    print(plot)}
    
   else{print(gene)}
  }

```

# Inherited features

```{r Paternal features}
Paternal_histone_features <- c("KDM3B", "KDM2A", "KDM6A", "KAT6B", "KAT6A", "MSL3", "KMT5C", "PHF8", "NSD1", "KMT5A", "KMT5C", "SUPT7L", "HDAC9", "ASH1L", "NCOA1") #List of paternal features

DoHeatmap(Blast_integrated, features = Paternal_histone_features, group.by = "cell_layers", size = 3, slot = "data")+theme(text = element_text(size = 5))+NoLegend()
VlnPlot(Blast_integrated, features = Paternal_histone_features, group.by = "cell_layers", split.by = "data_set", )
```

### Maternal features

```{r Maternal features, fig.height= 10}
MEGs <- c("AGO2", "ARID1A", "ATG5", "BCAR4", "BCAS2", "BNC1", "BTG4", "CDC20", "CDX2", "CNOT6L", "CTCF", "CTNNB1", "CXXC1", "DCAF13", "DDB1", "DGCR8", "DICER1", "DNMT1", "DNMT3A", "DNMT3L", "DTL", "EED", "EHMT2", "EXOSC10", "EZH2", "GCLM", "H3-3B", "HIRA", "H2AC1", "H2BC1", "HSF1", "HSP90B1", "IGF2BP2", "KDM1A", "KDM4A", "KHDC3L", "KMT2D", "KPNA6", "MAPK3", "MED12", "MGAT1", "MLH3", "NLRP2", "NLRP4", "NLRP5", "NLRP7", "NLRP9", "NPM2", "OOEP", "PADI6", "PATL2", "PDK1", "PLAG1", "PUM1", "RAD9A", "RING1", "RNF2", "SETD2", "SETDB1", "SLBP", "SMARCA4", "TCL1A", "TET3", "TLE6", "TP73", "TRIM28", "TRIP13", "TUBB8", "TUT4", "TUT7", "UBE2", "UCHL1", "UHRF1", "YAP1", "YTHDF2", "ZAR1", " ZBED3", "ZFP36L2", "ZFP57") # list of MEGs from paper: https://www.cell.com/hgg-advances/fulltext/S2666-2477(21)00048-8
DoHeatmap(Blast_integrated, features =  MEGs, group.by = "cell_layers", slot = "data", label = F)+scale_fill_gradient(low = "white", high = "black") #Heatmap of Maternal effect genes


```

## Imprinted genes

```{r fig.height= 15}
Imprinted_genes <- (read.csv("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/Imprinted_genes.csv")) #Load Imprinted_genes targets
Imprinted_genes <- Imprinted_genes$Imprinted_gene_list
DoHeatmap(Blast_integrated, features =  Imprinted_genes, group.by = "times", slot = "data", label = F)+scale_fill_gradient(low = "white", high = "black") #Heatmap of Imprinted_genes
```

```{r fig.width= 15}

Imprinted_genes_wk <- c("PLAGL1", "IGF2", "IGF2R", "UBE3A", "MEST", "H19", "CDKN1C", "PHLDA2", "DLK1", "MEG3", "MKRN3", "GNAS")#Imprinted genes that are associated with specific diseases
DoHeatmap(Blast_integrated_copy, features =  Imprinted_genes_wk, group.by = "times", slot = "data", label = F)+scale_fill_gradient(low = "white", high = "black") #Heatmap of Imprinted_genes
```

# Time points

# Overlap with other targets

```{r ELAVL1 targets}
#To compare to VIRMA targets
ELAVL_targets <- as.data.frame(readxl::read_excel(path = "C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/ELAVL_targets.xlsx")) #Load ELAVL targets

#Use pipeline of AnnotationDbi to convert gene aliases to gene symbol
geneSymbols <- AnnotationDbi::mapIds(org.Hs.eg.db, keys=ELAVL_targets$`ELAVL mRNA targets`, column="SYMBOL", keytype="ALIAS", multiVals="first")
geneSymbols <- as.data.frame(geneSymbols) #Convert to column of dataframe
ELAVL_targets$new_gene_names <- geneSymbols$geneSymbols #Add new gene names to target data frame
ELAVL_targets = ELAVL_targets[!duplicated(ELAVL_targets$new_gene_names),] #Remove targets with no official gene name
#If no official gene names are available. Exclude those from data set
ELAVL_targets <- na.omit(ELAVL_targets)
rownames(ELAVL_targets) <- ELAVL_targets$new_gene_names #Make row names into new gene names and delete the redundant columns
ELAVL_targets <- rownames(ELAVL_targets)
ELAVL_targets <- unlist(ELAVL_targets)
write.csv(ELAVL_targets, "output/ELAVL_targets") #Save ELAVL targets
```

```{r YTHDF2 targets}
#To compare to VIRMA targets
YTHDF2_targets <- as.data.frame(readxl::read_excel(path = "C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/YTHDF2_targets.xlsx")) #Load YTHDF2 targets

#Use pipeline of AnnotationDbi to convert gene aliases to gene symbol
geneSymbols <- AnnotationDbi::mapIds(org.Hs.eg.db, keys=YTHDF2_targets$`YTHDF2 targets`, column="SYMBOL", keytype="ALIAS", multiVals="first")
geneSymbols <- as.data.frame(geneSymbols) #Convert to column of dataframe
YTHDF2_targets$new_gene_names <- geneSymbols$geneSymbols #Add new gene names to target data frame
YTHDF2_targets = YTHDF2_targets[!duplicated(YTHDF2_targets$new_gene_names),] #Remove targets with no official gene name
#If no official gene names are available. Exclude those from data set
YTHDF2_targets <- na.omit(YTHDF2_targets)
rownames(YTHDF2_targets) <- YTHDF2_targets$new_gene_names #Make row names into new gene names and delete the redundant columns
YTHDF2_targets <- rownames(YTHDF2_targets)
YTHDF2_targets <- unlist(YTHDF2_targets)
write.csv(YTHDF2_targets, "output/YTHDF2_targets") #Save YTHDF2 targets
```

```{r IGF2BP targets}
#To compare to VIRMA targets
IGF2BP1_targets <- as.data.frame(read.csv("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/output/IGF2BP1_targets.csv")) #Load IGF2BP1 targets
IGF2BP2_targets <- as.data.frame(read.csv("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/output/IGF2BP2_targets.csv")) #Load IGF2BP2 targets
IGF2BP3_targets <- as.data.frame(read.csv("C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/output/IGF2BP3_targets.csv")) #Load IGF2BP3 targets

IGF2BP_targets <- IGF2BP1_targets$x
IGF2BP_targets <- append(IGF2BP_targets, c(IGF2BP2_targets$x, IGF2BP3_targets$x)) #Combine targets in one structure
```

```{r m6A targets}
#To compare to VIRMA targets
m6A_targets <- as.data.frame(readxl::read_excel(path = "C:/Users/tarau/Documents/Master Medical Epigenomics/Internship 1/Data/R/Blastocyst_Blastoid/data/m6A_targets.xlsx")) #Load m6A_targets targets

#Use pipeline of AnnotationDbi to convert gene aliases to gene symbol
geneSymbols <- AnnotationDbi::mapIds(org.Hs.eg.db, keys=m6A_targets$`HeLa_Ctrl m6A gene`, column="SYMBOL", keytype="ALIAS", multiVals="first")
geneSymbols <- as.data.frame(geneSymbols) #Convert to column of dataframe
m6A_targets$new_gene_names <- geneSymbols$geneSymbols #Add new gene names to target data frame
m6A_targets = m6A_targets[!duplicated(m6A_targets$new_gene_names),] #Remove targets with no official gene name
#If no official gene names are available. Exclude those from data set
m6A_targets <- na.omit(m6A_targets)
rownames(m6A_targets) <- m6A_targets$new_gene_names #Make row names into new gene names and delete the redundant columns
m6A_targets <- rownames(m6A_targets)
m6A_targets <- unlist(m6A_targets)
write.csv(m6A_targets, "output/m6A_targets")
```

```{r fig.width= 12, fig.height=5}

# A method of visualizing the different targets in a volcano plot. Not used in this report anymore. 


# DEG_analysis_m6A <- function(df, group, loglimit, adj_p_value){
#   group <- as.character(group)
#   design <- model.matrix(~group)
#   dge <- DGEList(counts = log(df+1))
#   fit <- lmFit(as.matrix(dge), design)
#   fit <- eBayes(fit, trend = T)
#   table_dge <- topTable(fit, coef=ncol(design), number = length(rownames(df)))
# 
#   table_dge$m6A <- "others"
#   
#   table_dge$m6A[rownames(table_dge) %in% m6A_targets] <- "Target of m6A"
#   #table_dge$VIRMA[rownames(table_dge) %in% VIRMA_binding_RNAs] <- "Target of VIRMA"
# 
# 
#   table_dge$m6A[rownames(table_dge) %in% c("CNOT1", "CNOT2", "CNOT3", "CNOT4", "CNOT6", "CNOT6", "CNOT6L", "CNOT7", "CNOT8", "CNOT9", "CNOT10", "CNOT11")] <- "zCNOT_complex"
# 
#   table_dge$genes <- rownames(table_dge)
#   table_dge$delabel <- NA
#   table_dge$delabel[table_dge$m6A != "others"] <- table_dge$genes[table_dge$m6A != "others"]
#   table_dge <- table_dge[order(table_dge$m6A, decreasing = F, na.last = T),]
#   p <- ggplot(data = table_dge, aes(x= logFC, y=-log10(adj.P.Val), col= m6A, show.legend = F))+
#     NoLegend()+
#     geom_point(show.legend = F)+
#     theme_minimal()+
#     #geom_text_repel(col = "black", na.rm = TRUE, hjust = 1)+
#     geom_vline(xintercept=c(-loglimit, loglimit), col="grey") +
#     geom_hline(yintercept=-log10(adj_p_value), col="grey") +
#     scale_color_manual(values = c("grey", "gold", "purple4"))+
#     ggtitle("m6A")
# 
#   return(p)
# }
# 
# DEG_analysis_VIRMA <- function(df, group, loglimit, adj_p_value){
#   group <- as.character(group)
#   design <- model.matrix(~group)
#   dge <- DGEList(counts = log(df+1))
#   fit <- lmFit(as.matrix(dge), design)
#   fit <- eBayes(fit, trend = T)
#   table_dge <- topTable(fit, coef=ncol(design), number = length(rownames(df)))
# 
#   table_dge$VIRMA <- "others"
#   table_dge$VIRMA[rownames(table_dge) %in% VIRMA_binding_RNAs] <- "Target of VIRMA"
# 
# 
#   table_dge$VIRMA[rownames(table_dge) == "VIRMA"] <- "VIRMA"
# 
#   table_dge$genes <- rownames(table_dge)
#   table_dge$delabel <- NA
#   table_dge$delabel[table_dge$VIRMA != "others"] <- table_dge$genes[table_dge$VIRMA != "others"]
#   table_dge <- table_dge[order(table_dge$VIRMA, decreasing = F, na.last = T),]
#   p <- ggplot(data = table_dge, aes(x= logFC, y=-log10(adj.P.Val), col= VIRMA, show.legend = F))+
#     NoLegend()+
#     geom_point(show.legend = F)+
#     theme_minimal()+
#     #geom_text_repel(col = "black", na.rm = TRUE, hjust = 1)+
#     geom_vline(xintercept=c(-loglimit, loglimit), col="grey") +
#     geom_hline(yintercept=-log10(adj_p_value), col="grey") +
#     scale_color_manual(values = c("grey", "coral","cornflowerblue"))+
#     ggtitle("VIRMA")
# 
#   return(p)
# }
# 
# DEG_analysis_YTHDF2 <- function(df, group, loglimit, adj_p_value){
#   group <- as.character(group)
#   design <- model.matrix(~group)
#   dge <- DGEList(counts = log(df+1))
#   fit <- lmFit(as.matrix(dge), design)
#   fit <- eBayes(fit, trend = T)
#   table_dge <- topTable(fit, coef=ncol(design), number = length(rownames(df)))
# 
#   table_dge$YTHDF2 <- "others"
#   table_dge$YTHDF2[rownames(table_dge) %in% YTHDF2_targets] <- "Target of YTHDF2"
#   table_dge$YTHDF2[rownames(table_dge) == "YTHDF2"] <- "YTHDF2"
# 
#   table_dge$genes <- rownames(table_dge)
#   table_dge$delabel <- NA
#   table_dge$delabel[table_dge$YTHDF2 != "others"] <- table_dge$genes[table_dge$YTHDF2 != "others"]
#   table_dge <- table_dge[order(table_dge$YTHDF2, decreasing = F, na.last = T),]
# 
#   p <- ggplot(data = table_dge, aes(x= logFC, y=-log10(adj.P.Val), col= YTHDF2, show.legend = F))+
#     NoLegend()+
#     geom_point(show.legend = F)+
#     theme_minimal()+
#     #geom_text_repel(col = "black", na.rm = TRUE, hjust = 1)+
#     geom_vline(xintercept=c(-loglimit, loglimit), col="grey") +
#     geom_hline(yintercept=-log10(adj_p_value), col="grey") +
#     scale_color_manual(values = c("grey", "goldenrod1","purple4"))+
#     ggtitle("YTHDF2")
# 
#   return(p)
# }
# 
# DEG_analysis_IGF2BPs <- function(df, group, loglimit, adj_p_value){
#   group <- as.character(group)
#   design <- model.matrix(~group)
#   dge <- DGEList(counts = log(df+1))
#   fit <- lmFit(as.matrix(dge), design)
#   fit <- eBayes(fit, trend = T)
#   table_dge <- topTable(fit, coef=ncol(design), number = length(rownames(df)))
# 
#   table_dge$IGF2BP <- "others"
#   table_dge$IGF2BP[rownames(table_dge) %in% IGF2BP_targets] <- "Target of IGF2BP"
# 
#   table_dge$IGF2BP[rownames(table_dge) %in% c("IGF2BP1", "IGF2BP2", "IGF2BP3")] <- "zIGF2BPs"
# 
#   table_dge$genes <- rownames(table_dge)
#   table_dge$delabel <- NA
#   table_dge$delabel[table_dge$IGF2BP != "others"] <- table_dge$genes[table_dge$IGF2BP != "others"]
#   table_dge <- table_dge[order(table_dge$IGF2BP, decreasing = F, na.last = T),]
# 
#   p <- ggplot(data = table_dge, aes(x= logFC, y=-log10(adj.P.Val), col= IGF2BP, show.legend = F))+
#     NoLegend()+
#     geom_point(show.legend = F)+
#     theme_minimal()+
#     #geom_text_repel(col = "black", na.rm = TRUE, hjust = 1)+
#     geom_vline(xintercept=c(-loglimit, loglimit), col="grey") +
#     geom_hline(yintercept=-log10(adj_p_value), col="grey") +
#     scale_color_manual(values = c("grey", "darkturquoise","firebrick1"))+
#     ggtitle("IGF2BPs")
# 
#   return(p)
# }
# 
# DEG_analysis_IGF2BPs <- function(df, group, loglimit, adj_p_value){
#   group <- as.character(group)
#   design <- model.matrix(~group)
#   dge <- DGEList(counts = log(df+1))
#   fit <- lmFit(as.matrix(dge), design)
#   fit <- eBayes(fit, trend = T)
#   table_dge <- topTable(fit, coef=ncol(design), number = length(rownames(df)))
# 
#   table_dge$IGF2BP <- "others"
#   table_dge$IGF2BP[rownames(table_dge) %in% IGF2BP1_targets$x] <- "Target of IGF2BP1"
#   table_dge$IGF2BP[rownames(table_dge) %in% c("IGF2BP1", "IGF2BP2", "IGF2BP3")] <- "zIGF2BPs"
# 
#   table_dge$genes <- rownames(table_dge)
#   table_dge$delabel <- NA
#   table_dge$delabel[table_dge$IGF2BP != "others"] <- table_dge$genes[table_dge$IGF2BP != "others"]
#   table_dge <- table_dge[order(table_dge$IGF2BP, decreasing = F, na.last = T),]
# 
#   p1 <- ggplot(data = table_dge, aes(x= logFC, y=-log10(adj.P.Val), col= IGF2BP, show.legend = F))+
#     NoLegend()+
#     geom_point(show.legend = F)+
#     theme_minimal()+
#     #geom_text_repel(col = "black", na.rm = TRUE, hjust = 1)+
#     geom_vline(xintercept=c(-loglimit, loglimit), col="grey") +
#     geom_hline(yintercept=-log10(adj_p_value), col="grey") +
#     scale_color_manual(values = c("grey", "springgreen","firebrick1"))+
#     ggtitle("IGF2BP1")
#   
#   table_dge$IGF2BP <- "others"
#   table_dge$IGF2BP[rownames(table_dge) %in% IGF2BP2_targets$x] <- "Target of IGF2BP2"
#   table_dge$IGF2BP[rownames(table_dge) %in% c("IGF2BP1", "IGF2BP2", "IGF2BP3")] <- "zIGF2BPs"
# 
#   table_dge$genes <- rownames(table_dge)
#   table_dge$delabel <- NA
#   table_dge$delabel[table_dge$IGF2BP != "others"] <- table_dge$genes[table_dge$IGF2BP != "others"]
#   table_dge <- table_dge[order(table_dge$IGF2BP, decreasing = F, na.last = T),]
# 
#   p2 <- ggplot(data = table_dge, aes(x= logFC, y=-log10(adj.P.Val), col= IGF2BP, show.legend = F))+
#     NoLegend()+
#     geom_point(show.legend = F)+
#     theme_minimal()+
#     #geom_text_repel(col = "black", na.rm = TRUE, hjust = 1)+
#     geom_vline(xintercept=c(-loglimit, loglimit), col="grey") +
#     geom_hline(yintercept=-log10(adj_p_value), col="grey") +
#     scale_color_manual(values = c("grey", "slateblue","firebrick1"))+
#     ggtitle("IGF2BP2")
# 
#   table_dge$IGF2BP <- "others"
#   table_dge$IGF2BP[rownames(table_dge) %in% IGF2BP3_targets$x] <- "Target of IGF2BP3"
#   table_dge$IGF2BP[rownames(table_dge) %in% c("IGF2BP1", "IGF2BP2", "IGF2BP3")] <- "zIGF2BPs"
# 
#   table_dge$genes <- rownames(table_dge)
#   table_dge$delabel <- NA
#   table_dge$delabel[table_dge$IGF2BP != "others"] <- table_dge$genes[table_dge$IGF2BP != "others"]
#   table_dge <- table_dge[order(table_dge$IGF2BP, decreasing = F, na.last = T),]
# 
#   p3 <- ggplot(data = table_dge, aes(x= logFC, y=-log10(adj.P.Val), col= IGF2BP, show.legend = F))+
#     NoLegend()+
#     geom_point(show.legend = F)+
#     theme_minimal()+
#     #geom_text_repel(col = "black", na.rm = TRUE, hjust = 1)+
#     geom_vline(xintercept=c(-loglimit, loglimit), col="grey") +
#     geom_hline(yintercept=-log10(adj_p_value), col="grey") +
#     scale_color_manual(values = c("grey", "mediumturquoise","firebrick1"))+
#     ggtitle("IGF2BP3")
#   return(p1 | p2 | p3)
# }
# 
# group <- c(rep(c("Blastocyst"), times = 10 ), rep(c("Blastoid"), times =10))
# design <- model.matrix(~group)
# Plot_VIRMA_TE <- DEG_analysis_VIRMA(Blast_tpm_te_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# Plot_m6A_TE <- DEG_analysis_m6A(Blast_tpm_te_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# Plot_YTHDF2_TE <- DEG_analysis_YTHDF2(Blast_tpm_te_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# Plot_IGF2BPs_TE <- DEG_analysis_IGF2BPs(Blast_tpm_te_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# 
# group <- c(rep(c("Blastocyst"), times = 9 ), rep(c("Blastoid"), times =10))
# design <- model.matrix(~group)
# Plot_VIRMA_ICM <- DEG_analysis_VIRMA(Blast_tpm_icm_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# Plot_m6a_ICM <- DEG_analysis_m6A(Blast_tpm_icm_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# Plot_YTHDF2_ICM <- DEG_analysis_YTHDF2(Blast_tpm_icm_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# Plot_IGF2BPs_ICM <- DEG_analysis_IGF2BPs(Blast_tpm_icm_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# 
# group <- c(rep(c("Blastocyst"), times = 5 ), rep(c("Blastoid"), times =10))
# design <- model.matrix(~group)
# Plot_VIRMA_PrE <- DEG_analysis_VIRMA(Blast_tpm_pre_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# Plot_m6A_PrE <- DEG_analysis_m6A(Blast_tpm_pre_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# Plot_YTHDF2_PrE <- DEG_analysis_YTHDF2(Blast_tpm_pre_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# Plot_IGF2BPs_PrE <- DEG_analysis_IGF2BPs(Blast_tpm_pre_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# 
# group <- c(rep(c("Blastocyst"), times = 10 ), rep(c("Blastoid"), times =10))
# design <- model.matrix(~group)
# Plot_VIRMA_all <- DEG_analysis_VIRMA(Blast_tpm_all_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# Plot_m6A_all <- DEG_analysis_m6A(Blast_tpm_all_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# Plot_YTHDF2_all <- DEG_analysis_YTHDF2(Blast_tpm_all_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# Plot_IGF2BPs_all <- DEG_analysis_IGF2BPs(Blast_tpm_all_10, group =group, loglimit = 1.5, adj_p_value = 0.05)
# 
# Plot_m6A_all
# Plot_YTHDF2_all | Plot_IGF2BPs_all | Plot_VIRMA_all
# Plot_IGF2BPs_all
# Plot_YTHDF2_TE | Plot_IGF2BPs_TE | Plot_VIRMA_TE
```

```{r}
Stress_granule_inhabitants <- readxl::read_excel("data/stress_granule_inhabitants.xlsx")
#Use pipeline of AnnotationDbi to convert gene aliases to gene symbol
geneSymbols <- AnnotationDbi::mapIds(org.Hs.eg.db, keys=Stress_granule_inhabitants$Gene_ID, column="SYMBOL", keytype="ALIAS", multiVals="first")
geneSymbols <- as.data.frame(geneSymbols) #Convert to column of dataframe
Stress_granule_inhabitants$new_gene_names <- geneSymbols$geneSymbols #Add new gene names to target data frame
Stress_granule_inhabitants = Stress_granule_inhabitants[!duplicated(Stress_granule_inhabitants$new_gene_names),] #Remove targets with no official gene name
#If no official gene names are available. Exclude those from data set
Stress_granule_inhabitants <- na.omit(Stress_granule_inhabitants)
Stress_granule_inhabitants <- Stress_granule_inhabitants$new_gene_names
Stress_granule_inhabitants <- unlist(Stress_granule_inhabitants)
write.csv(Stress_granule_inhabitants, "output/Stress_granule_inhabitants")
```

# CollecTri

## Setup

```{r Collectri data}

#For collecTri follow instruction (https://saezlab.github.io/decoupleR/articles/tf_sc.html) and run the "run_ulm" function on R using script in data folder
#net <- get_collectri(organism='human', split_complexes=FALSE)
#mat <- as.matrix(Blast_integrated@assays$RNA@layers$counts) #Obtain matrix belonging to Seurat object
#colnames(mat) <- Blast_integrated@meta.data$names
#rownames(mat) <- rownames(Blast_integrated)
#write_rds(x = mat, file = "output/collectri_matrix") #Save the matrix and run in mobaXterm using script in folder.
Collectri_blast_integrated <- readRDS("output/acts_ulm2.RDS") #After running on server get 


#!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!1
#Because the seurat object changed. This data need to be run again to perform subsequent analysis. 
#Code is kept for reproducibility purposes

# # Extract ulm and store it in tfsulm in pbmc
# Blast_integrated[['tfsulm']] <- Collectri_blast_integrated %>%
#   pivot_wider(id_cols = 'source', names_from = 'condition',
#               values_from = 'score') %>%
#   column_to_rownames('source') %>%
#   Seurat::CreateAssayObject(.)
# 
# 
# 
# # Scale the data
# Blast_integrated <- ScaleData(Blast_integrated, assay = "tfsulm")
# Blast_integrated@assays$tfsulm@data <- Blast_integrated@assays$tfsulm@scale.data
# 
# 

```

## Overview

```{r}

# n_tfs <- 50
# #Extract activities from object as a long dataframe
# df <- t(as.matrix(Blast_integrated@assays$tfsulm@data)) %>%
#   as.data.frame() %>%
#   mutate(cluster = Blast_integrated@meta.data$data_set) %>%
#   pivot_longer(cols = -cluster, names_to = "source", values_to = "score") %>%
#   group_by(cluster, source) %>%
#   summarise(mean = mean(score))
# 
# # Get top tfs with more variable means across clusters
# tfs <- df %>%
#   group_by(source) %>%
#   summarise(std = sd(mean)) %>%
#   arrange(-abs(std)) %>%
#   head(n_tfs) %>%
#   pull(source)
# 
# # Subset long data frame to top tfs and transform to wide matrix
# top_acts_mat <- df %>%
#   filter(source %in% tfs) %>%
#   pivot_wider(id_cols = 'cluster', names_from = 'source',
#               values_from = 'mean') %>%
#   column_to_rownames('cluster') %>%
#   as.matrix()
# 
# # Choose color palette
# palette_length = 100
# my_color = colorRampPalette(c("darkblue", "white","red"))(palette_length)
# 
# my_breaks <- c(seq(-3, 0, length.out=ceiling(palette_length/2) + 1),
#                seq(0.05, 3, length.out=floor(palette_length/2)))
# 
# # Plot
# pheatmap(top_acts_mat, border_color = NA, color=my_color, breaks = my_breaks, fontsize = 8) 
```

## Pathways

### YAP Pathway

```{r YAP, fig.width = 12}
# p1 <- DimPlot(Blast_integrated, reduction = "umap", label = TRUE, pt.size = 0.5, group.by = "data_set") + 
#   NoLegend() + ggtitle('Cell types')
# DefaultAssay(object = Blast_integrated) <- "tfsulm"
# p2 <- (FeaturePlot(Blast_integrated, features = c("YAP1"), reduction = "umap") & 
#   scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) +
#   ggtitle('YAP1 activity')
# VlnPlot(Blast_integrated, features = c("YAP1"), split.by = "data_set")
# DefaultAssay(object = Blast_integrated) <- "RNA"
# p3 <- FeaturePlot(Blast_integrated, features = c("YAP1"), reduction = "umap") + ggtitle('YAP1 expression')
# p1 | p3 | p2

```

```{r}
# p1 <- DimPlot(Blast_integrated, reduction = "umap", label = TRUE, pt.size = 0.5, group.by = "data_set") + 
#   NoLegend() + ggtitle('Cell types')
# DefaultAssay(object = Blast_integrated) <- "tfsulm"
# p2 <- (FeaturePlot(Blast_integrated, features = c("TBXT"), reduction = "umap") & 
#   scale_colour_gradient2(low = 'blue', mid = 'white', high = 'red')) +
#   ggtitle('TBXT activity')
# 
# VlnPlot(Blast_integrated, features = "TBXT", split.by = "data_set", group.by = "cell_layers")
# DefaultAssay(object = Blast_integrated) <- "RNA"
# p3 <- FeaturePlot(Blast_integrated, features = c("TBXT"), reduction = "umap") + ggtitle('TBXT expression')
# p1 | p2 | p3


```

```{r}
# Blast_integrated_te <- subset(Blast_integrated, cell_type == "MTE" | cell_type == "TE" | cell_type == "PTE" )
# #Extract activities from object as a long dataframe
# df <- t(as.matrix(Blast_integrated_te@assays$tfsulm@data)) %>%
#   as.data.frame() %>%
#   mutate(cluster = Blast_integrated_te@meta.data$data_set) %>%
#   pivot_longer(cols = -cluster, names_to = "source", values_to = "score") %>%
#   group_by(cluster, source) %>%
#   summarise(mean = mean(score))
# 
# # Get top tfs with more variable means across clusters
# tfs <- df %>%
#   group_by(source) %>%
#   summarise(std = sd(mean)) %>%
#   arrange(-abs(std)) %>%
#   head(n_tfs) %>%
#   pull(source)
# 
# # Subset long data frame to top tfs and transform to wide matrix
# top_acts_mat <- df %>%
#   filter(source %in% tfs) %>%
#   pivot_wider(id_cols = 'cluster', names_from = 'source',
#               values_from = 'mean') %>%
#   column_to_rownames('cluster') %>%
#   as.matrix()
# 
# # Choose color palette
# palette_length = 100
# my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)
# 
# my_breaks <- c(seq(-3, 0, length.out=ceiling(palette_length/2) + 1),
#                seq(0.05, 3, length.out=floor(palette_length/2)))
# 
# # Plot
# pheatmap(top_acts_mat, border_color = NA, color=my_color, breaks = my_breaks, fontsize = 8, main = "Activity of TFs in trophectoderm") 
# 
# Blast_integrated_icm <- subset(Blast_integrated, cell_type == "EPI" | cell_type == "ICM" )
# #Extract activities from object as a long dataframe
# df <- t(as.matrix(Blast_integrated_icm@assays$tfsulm@data)) %>%
#   as.data.frame() %>%
#   mutate(cluster = Blast_integrated_icm@meta.data$data_set) %>%
#   pivot_longer(cols = -cluster, names_to = "source", values_to = "score") %>%
#   group_by(cluster, source) %>%
#   summarise(mean = mean(score))
# 
# # Get top tfs with more variable means across clusters
# tfs <- df %>%
#   group_by(source) %>%
#   summarise(std = sd(mean)) %>%
#   arrange(-abs(std)) %>%
#   head(n_tfs) %>%
#   pull(source)
# 
# # Subset long data frame to top tfs and transform to wide matrix
# top_acts_mat <- df %>%
#   filter(source %in% tfs) %>%
#   pivot_wider(id_cols = 'cluster', names_from = 'source',
#               values_from = 'mean') %>%
#   column_to_rownames('cluster') %>%
#   as.matrix()
# 
# # Choose color palette
# palette_length = 100
# my_color = colorRampPalette(c("Darkblue", "white","red"))(palette_length)
# 
# my_breaks <- c(seq(-3, 0, length.out=ceiling(palette_length/2) + 1),
#                seq(0.05, 3, length.out=floor(palette_length/2)))
# 
# # Plot
# pheatmap(top_acts_mat, border_color = NA, color=my_color, breaks = my_breaks, fontsize = 8,  main = "Activity of TFs in ICM")
```

# Gene ontology

### Downregulated VIRMA targets in blastoid

```{r}
go_VIRMA_all <- read.csv("data/enrichment_DEGS_all_VIRMA.csv") #Open file altered from ShinyGo
go_VIRMA_all <- go_VIRMA_all[order(go_VIRMA_all$Fold_Enrichment, decreasing = TRUE),] #Order the GO terms on basis of fold enrichment from high to low. 
go_VIRMA_all <- go_VIRMA_all[c(1:10),] # Only top 10 
go_VIRMA_all_plot <- ggplot(go_VIRMA_all, 
                            aes(x = reorder(Pathway, go_VIRMA_all$Fold_Enrichment), 
                            y=go_VIRMA_all$Fold_Enrichment, 
                            fill = go_VIRMA_all$minus_log10.fdr.))+ #Add color scale indicating significance 
  geom_col(width = 0.6, col = "black")+
  coord_flip()+ #I want an horizontal bar graph
  ggtitle("GO terms associated with \ndownregulated VIRMA targets in Blastoids")+
  ylab("Fold Enrichment")+
  xlab(NULL)+
  scale_fill_gradient(low = "oldlace", high = "darkred", name = "-log10(FDR)", limits = c(0,70), position = "left")+
  theme_classic(base_size = 20 )+
  scale_y_continuous(expand = c(0, 0), limits = c(0, 4))

go_VIRMA_all_plot
```

### All downregulated genes in blastoid

```{r fig.width = 12, fig.height = 6}
go_all <- read.csv("data/enrichment_all.csv") #open the file from Shiny go
go_all <- head(go_all[order(go_all$Fold_Enrichment, decreasing = TRUE), ], 10) 
plot_go_term_all <- ggplot(go_all, aes(x = reorder(Pathway, go_all$Fold_Enrichment), y=go_all$Fold_Enrichment, fill = go_all$minus_log10.fdr.))+
  geom_col(width = 0.6, col = "black")+
  coord_flip()+
  ggtitle("GO terms associated with \ndownregulated genes in Blastoids")+
  ylab("Fold Enrichment")+
  xlab(NULL)+
  scale_fill_gradient(low = "oldlace", high = "darkred", name = "-log10(FDR)", limits = c(0,70))+
  theme_classic(base_size = 20)+
  scale_y_continuous(expand = c(0, 0), limits = c(0, 2.5))
plot_go_term_all
```

```{r}
RPKM_development <- read.csv("data/RPKM_development.csv") #Open data from https://www.nature.com/articles/nsmb.2660#Sec33
colnames(RPKM_development) <- c("Gene_ID", "Oocyte", "Zygote", "2Cell", "4Cell", "8Cell", "Morula", "Trophectoderm", "Primitive Endoderm", "Epiblast", "hESC-0", "hESC-10")


#Use pipeline of AnnotationDbi to convert gene aliases to gene symbol
geneSymbols <- AnnotationDbi::mapIds(org.Hs.eg.db, keys=RPKM_development$Gene_ID, column="SYMBOL", keytype="ALIAS", multiVals="first")#Convert aliases to official symbols
geneSymbols <- as.data.frame(geneSymbols) #Convert to column of dataframe
RPKM_development$Gene_ID <- geneSymbols$geneSymbols #Add new gene names to target data frame

RPKM_development = RPKM_development[!duplicated(RPKM_development$Gene_ID),] #Remove targets with no official gene name

TPM_development <- RPKM_development #new name for further transformations
TPM_development[,(c(2:12))] <- apply(RPKM_development[,(c(2:12))], 2, function(x) x / sum(as.numeric(x)) * 10^6) #Convert to TPM


TPM_development_stages <- TPM_development[, c(0:10)] #Select for columns that contain only oocyte derived cells.
TPM_development_stages <- TPM_development_stages %>% 
  pivot_longer(cols = 'Oocyte':'Epiblast',
               names_to = "Stage",
               values_to = c("RPKM"),)

RPKM_development <- RPKM_development %>%
  pivot_longer(cols = 'Oocyte':'Epiblast',
               names_to = "Stage",
               values_to = c("RPKM"),)

```

```{r}
TPM_development_stages_VIRMA <- subset(TPM_development_stages, Gene_ID %in% c("YTHDF2", "MATR3", "ELAVL1", "IGF2BP1", "IGF2BP2", "IGF2BP3")) #Subset data to only contain readers


TPM_development_stages_writer_complex <- subset(TPM_development_stages, Gene_ID %in% c("VIRMA", "WTAP", "METTL3", "METTL14", "CBLL1",  "ZC3H13")) #Subset data to only contain writer complex genes


#Set up a different dataframe w/o blastocyst for geomline within ggplot
Line_writer_complex <- subset(TPM_development_stages_writer_complex,  
    TPM_development_stages_writer_complex$Stage == "Oocyte" |
    TPM_development_stages_writer_complex$Stage == "Zygote" |
    TPM_development_stages_writer_complex$Stage == "2Cell" |
    TPM_development_stages_writer_complex$Stage == "4Cell" |
    TPM_development_stages_writer_complex$Stage == "8Cell" | 
    TPM_development_stages_writer_complex$Stage == "Morula")
    

order_stages <- c("Oocyte", "Zygote", "2Cell", "4Cell", "8Cell", "Morula", "Epiblast" , "Primitive Endoderm","Trophectoderm") #Order stages chronologically


plot_writer_complex_stages <- ggplot(TPM_development_stages_writer_complex, aes(x = Stage, y = RPKM, group = Gene_ID, col = Gene_ID)) + #Plot of writer subunits per stage
  geom_point(size = 2) +
  geom_line(linewidth = 1, 
            data = Line_writer_complex) + #This will only plot a line from oocyte to morula
  scale_x_discrete(limits = order_stages) +
  labs(x = NULL, 
       y = "TPM Value", 
       title = "m6A Methyltransferase complex proteins per stage")+
  scale_color_manual(values =c("deeppink",
                               "lightgreen", 
                               "cornflowerblue", 
                               "darkgoldenrod2", 
                               "limegreen", 
                               "orchid", 
                               "grey"))+ #These colors are chosen to represent the figure in the report
  theme_bw()+
  theme(panel.grid = element_blank())+
  facet_wrap(~Gene_ID,
             scales = "free_y", 
             ncol = 3)+
  theme(axis.text.x = element_text(angle = 45, 
                                   vjust = 1, 
                                   hjust=1, 
                                   size = 10), 
        strip.text = element_text(size =15))+
  geom_vline(xintercept = "Morula", 
             linetype = "dashed", 
             col = "grey30")+
  NoLegend()

plot_writer_complex_stages
```

```{r}


TPM_development_stages_readers <- TPM_development_stages_VIRMA
#TPM_development_stages_readers <- TPM_development_stages_readers[TPM_development_stages_readers$Stage == "Zygote" | TPM_development_stages_readers$Stage == "Epiblast"]
TPM_line_readers <- subset(TPM_development_stages_readers, #This subset is made to plot a line in ggplot later
    TPM_development_stages_readers$Stage == "Oocyte" |
    TPM_development_stages_readers$Stage == "Zygote" |
    TPM_development_stages_readers$Stage == "2Cell" |
    TPM_development_stages_readers$Stage == "4Cell" |
    TPM_development_stages_readers$Stage == "8Cell" | 
    TPM_development_stages_readers$Stage == "Morula")
    
readers_stage_plot <- ggplot(TPM_development_stages_readers, aes(x = Stage, 
                                                                 y = RPKM, 
                                                                 group = Gene_ID, 
                                                                 col = Gene_ID)) +
  geom_point(size = 2) +
  geom_line(linewidth = 1, 
            data = TPM_line_readers) + #The subset with only oocyte-morula is chosen
  scale_x_discrete(limits = order_stages) +
  labs(x = NULL, 
       y = "TPM Value", 
       title = "m6A reader proteins per stage")+
  scale_color_manual(values =c("skyblue",
                               "maroon4", 
                               "maroon", 
                               "maroon2", 
                               "khaki", 
                               "midnightblue"))+ #colors are corresponding to figure in report
  theme_bw()+
  theme(panel.grid.major = element_blank())+
  theme(panel.grid.minor = element_blank())+
  facet_wrap(~Gene_ID,
             scales = "free_y", 
             nrow = 2)+
  theme(axis.text.x = element_text(angle = 45,
                                   vjust = 1, 
                                   hjust=1, 
                                   size = 10), 
        strip.text = element_text(size =15))+
  NoLegend()+
  geom_vline(xintercept = "Morula", #Add a line to indicate were the lineages diverge
             linetype = "dashed", 
             col = "grey30")

readers_stage_plot


```

```{r}
DEGs_ALL_df <- readxl::read_excel("output/DEG_ALL_write.xlsx") #Open file that contains list of up and downregulated genes
chromatin_virma <- read.csv("data/chromatin_VIRMA.csv")
chromatin_virma <- unlist(chromatin_virma$Chromatin_VIRMA)



TPM_development_stages_VIRMA_degs <- subset(TPM_development_stages, (Gene_ID %in% DEGs_ALL_df$Blastocyst & Gene_ID %in% VIRMA_binding_RNAs)) #Make as subset containing only genes downregulated in blastoid, and are virma targets

TPM_development_stages_VIRMA_megs <- subset(TPM_development_stages, (Gene_ID %in% MEGs)) #A data frame containing only 

TPM_development_stages_others <- subset(TPM_development_stages, !(Gene_ID %in% DEGs_ALL_df$Blastocyst & Gene_ID %in% VIRMA_binding_RNAs)) # A data frame with no downregulated virma targets


#Mutate three data frames above to be in a longer format. This helps with ggplot
TPM_development_stages_VIRMA_degs <- TPM_development_stages_VIRMA_degs %>%
  group_by(Gene_ID) %>%
  mutate(scaled_RPKM = scale(RPKM, center = TRUE, scale = TRUE)) %>%
  ungroup()

#Mutate three data frames above to be in a longer format. This helps with ggplot
TPM_development_stages_VIRMA_megs <- TPM_development_stages_VIRMA_megs %>%
  group_by(Gene_ID) %>%
  mutate(scaled_RPKM = scale(RPKM, center = TRUE, scale = TRUE)) %>%
  ungroup()

#Mutate three data frames above to be in a longer format. This helps with ggplot
TPM_development_stages_others <- TPM_development_stages_others %>%
  group_by(Gene_ID) %>%
  mutate(scaled_RPKM = scale(RPKM, center = TRUE, scale = TRUE)) %>%
  ungroup()

# Calculate mean and confidence intervals per data set
# Downregulated VIRMA targets
summary_data <- TPM_development_stages_VIRMA_degs %>%
  group_by(Stage) %>%
  summarize(
    mean_scaled_RPKM = mean(scaled_RPKM, na.rm = TRUE),
    se_scaled_RPKM = sd(scaled_RPKM, na.rm = TRUE) / sqrt(n()),
    ci_lower = mean_scaled_RPKM - qt(1 - (0.05 / 2), n() - 1) * se_scaled_RPKM,
    ci_upper = mean_scaled_RPKM + qt(1 - (0.05 / 2), n() - 1) * se_scaled_RPKM
  )
summary_data$virma <- "Downregulated VIRMA targets in blastoid"

# Calculate mean and confidence intervals per data set
#Not downregulated VIRMA targets
summary_data_others <- TPM_development_stages_others %>%
  group_by(Stage) %>%
  summarize(
    mean_scaled_RPKM_other = mean(scaled_RPKM, na.rm = TRUE),
    se_scaled_RPKM = sd(scaled_RPKM, na.rm = TRUE) / sqrt(n()),
    ci_lower_other = mean_scaled_RPKM_other - qt(1 - (0.05 / 2), n() - 1) * se_scaled_RPKM,
    ci_upper_other = mean_scaled_RPKM_other + qt(1 - (0.05 / 2), n() - 1) * se_scaled_RPKM
  )
summary_data_others$virma <- "Not downregulated VIRMA targets in blastoid"

# Calculate mean and confidence intervals per data set
# Maternal effect genes
summary_data_megs <- TPM_development_stages_VIRMA_megs %>%
  group_by(Stage) %>%
  summarize(
    mean_scaled_RPKM_meg = mean(scaled_RPKM, na.rm = TRUE),
    se_scaled_RPKM = sd(scaled_RPKM, na.rm = TRUE) / sqrt(n()),
    ci_lower_meg = mean_scaled_RPKM_meg - qt(1 - (0.05 / 2), n() - 1) * se_scaled_RPKM,
    ci_upper_meg = mean_scaled_RPKM_meg + qt(1 - (0.05 / 2), n() - 1) * se_scaled_RPKM
  )
summary_data_megs$virma <- "Maternal effect genes"

#Join the dataframes together by stage
summary_data_both <- full_join(summary_data, summary_data_others, by= join_by("Stage"))
summary_data_both <- full_join(summary_data_both, summary_data_megs, by= join_by("Stage"))

#Make a subset of again oocyte- Morula
summary_data_both_line <- subset(summary_data_both, 
    summary_data_both$Stage == "Oocyte" |
    summary_data_both$Stage == "Zygote" |
    summary_data_both$Stage == "2Cell" |
    summary_data_both$Stage == "4Cell" |
    summary_data_both$Stage == "8Cell" | 
    summary_data_both$Stage == "Morula")

order_stages <- c("Oocyte", "Zygote", "2Cell", "4Cell", "8Cell", "Morula", "Epiblast", "Primitive Endoderm", "Trophectoderm") #Order stages in right chronological order

#Create the plot
scaled_gene_plot <- ggplot(summary_data_both, aes(x = Stage)) +
  scale_x_discrete(limits = order_stages,) +  
  #geom_ribbon(data = aes(ymin = ci_lower_other, ymax = ci_upper_other, group =3), alpha = 0.2) + #Code for addition of a ribbon
  geom_point(aes(y= mean_scaled_RPKM_other, 
                 group =3, 
                 col = "Not downregulated VIRMA \ntargets in blastoid"), 
             shape = 15)+
  geom_line(data = summary_data_both_line, #Add the line data from oocyte-morula
            aes(y= as.numeric(mean_scaled_RPKM_other), 
                group =3), 
            col = "grey30",
            linewidth = 1) +
  #geom_ribbon(aes(ymin = ci_lower, ymax = ci_upper, group =3), alpha = 0.2, fill = "coral") +
  geom_point(aes(y= as.numeric(mean_scaled_RPKM),
                 group =3, 
                 col = "Downregulated VIRMA \ntargets in blastoids"), 
             shape = 17) +
  geom_line(data = summary_data_both_line, #Add the line data from oocyte-morula
            aes(y= as.numeric(mean_scaled_RPKM), 
                group =3), 
            col = "coral", 
            linewidth = 1) +
  #geom_ribbon(aes(ymin = ci_lower_meg, ymax = ci_upper_meg, group =3), alpha = 0.2, fill = "mediumorchid") +
  geom_point(aes(y= as.numeric(mean_scaled_RPKM_meg), 
                 group =3, 
                 col = "Maternal effect genes"), 
             shape =19) +
  geom_line(data = summary_data_both_line, #Add the line data from oocyte-morula
            aes(y= as.numeric(mean_scaled_RPKM_meg), 
                group =3, 
                guide = T), 
            col = "mediumorchid", 
            linewidth = 1) +  
  theme_light(base_size = 15)+
  xlab(NULL)+
  ylab("Mean scaled \nexpression (TPM) per gene")+
  ggtitle("Trend of gene groups across stages")+
  geom_vline(xintercept = "Morula", 
             linetype = "dashed", 
             col = "grey30")+
  theme(panel.grid.major = element_blank())+
  theme(axis.text.x = element_text(angle = 45, 
                                   hjust=1, 
                                   
                                   colour = "black"))+
  #theme(panel.grid.minor = element_blank())
  scale_color_manual(values = c("coral", 
                                "mediumorchid", 
                                "grey30"), name = NULL)+
  theme(legend.position = "bottom", legend.title = NULL)

scaled_gene_plot

```

# Report figures

## Adjusting seurat object

```{r}
type_order <- c("Naive", "EPI", "ICM", "PrE", "TE", "MTE", "PTE") 
Blast_integrated_copy <- Blast_integrated
Blast_integrated_copy@meta.data$cell_type <- factor(Blast_integrated_copy@meta.data$cell_type, levels= type_order) #Make sure of the order
Blast_integrated_copy@meta.data$cell_layers2 <- Blast_integrated_copy@meta.data$cell_layers
Blast_integrated_copy@meta.data$cell_layers2[ #Combine all TEs
                      Blast_integrated_copy@meta.data$cell_layers2 == "Blastocyst-MTE" |
                      Blast_integrated_copy@meta.data$cell_layers2 == "Blastocyst-PTE" |
                      Blast_integrated_copy@meta.data$cell_layers2 == "Blastocyst-TE"] <- "TE (Blastocyst)"
Blast_integrated_copy@meta.data$cell_layers2[ #Combine EPI and ICM
                      Blast_integrated_copy@meta.data$cell_layers2 == "Blastocyst-EPI" |
                      Blast_integrated_copy@meta.data$cell_layers2 == "Blastocyst-ICM"] <- "EPI (Blastocyst)"
Blast_integrated_copy@meta.data$cell_layers2[ #Combine PRE
                      Blast_integrated_copy@meta.data$cell_layers2 == "Blastocyst-PrE"] <- "PRE (Blastocyst)"


#Do the same for the blastoids
Blast_integrated_copy@meta.data$cell_layers2[
                      Blast_integrated_copy@meta.data$cell_layers2 == "Blastoid-EPI"] <- "EPI (Blastoid)"

Blast_integrated_copy@meta.data$cell_layers2[
                      Blast_integrated_copy@meta.data$cell_layers2 == "Blastoid-TE"|
                      Blast_integrated_copy@meta.data$cell_layers2 == "Blastoid-PTE"] <- "TE (Blastoid)"
Blast_integrated_copy@meta.data$cell_layers2[
                      Blast_integrated_copy@meta.data$cell_layers2 == "Blastoid-PrE"] <- "PRE (Blastoid)"

Blast_integrated_copy@meta.data$cell_layers2[
                      Blast_integrated_copy@meta.data$cell_layers2 == "Blastoid-Naive"] <- "Naive (Blastoid)"


cell_layer2_order <- c("EPI (Blastocyst)", "EPI (Blastoid)", "Naive (Blastoid)", "PRE (Blastocyst)",  "PRE (Blastoid)",  "TE (Blastocyst)", "TE (Blastoid)") #Make sure that the order does not change
Blast_integrated_copy@meta.data$cell_layers2 <- factor(Blast_integrated_copy@meta.data$cell_layers2, levels= cell_layer2_order) #Add order to seurat
```

## Heatmaps

### Methyl writer complex heatmap

```{r Writer complex, fig.height =5, fig.width = 7}
#Make a heatmap to compare the methyl writer complex subunits
Blast_integrated_copy <- ScaleData(Blast_integrated_copy, #Scale data with all data
                                   features = rownames(Blast_integrated_copy))
Methyl_writer_complex <- c("METTL3", 
                           "METTL14", 
                           "VIRMA", 
                           "WTAP", 
                           "ZC3H13", 
                           "CBLL1")

#Make a graoh of the writer complex 
Methyl_writer_complex_hm <- DoHeatmap(Blast_integrated_copy, 
          features = Methyl_writer_complex, 
          group.by = "times", 
          size = 3, 
          angle = 0, 
          hjust = 0.5, 
          vjust = 0.1)+
  scale_fill_gradient2(mid = "gray30", 
                       low = "cornflowerblue", 
                       high = "coral")+  
  guides(color=FALSE)+
  theme(legend.position = "bottom")

```

### Reader heatmap

```{r Reader complex figure}

#A figure containing readers, it will consist out of 2 figures. One stabilizing reader figure and one associated with decay

#Heatmap containing stabilizing readers
heatmap_stabilizers <- DoHeatmap(Blast_integrated_copy, 
                                 group.by = "times", #Per time
                                 features = c("IGF2BP1", "IGF2BP2", "IGF2BP3", "ELAVL1", "MATR3"), 
                                 size = 3, 
                                 angle = 0,
                                 hjust = 0.5,
                                 vjust = 0.1)+
  scale_fill_gradient2(mid = "gray30", 
                       low = "cornflowerblue", 
                       high = "coral")+  
  guides(color=FALSE)+NoLegend()

heatmap_stabilizers

#Heatmap containing the decay associated 
heatmap_decay <- DoHeatmap(Blast_integrated_copy, 
                           group.by = "times", 
                           features = c("YTHDF2"), 
                           size = 2, 
                           angle = 0, 
                           group.bar = F, 
                           group.colors = F)+
  scale_fill_gradient2(mid = "gray30", 
                       low = "cornflowerblue", 
                       high = "coral")+  
  guides(color=FALSE)+
  theme(legend.position = "bottom")

heatmap_decay

#Combine the heatmaps
readers_heatmap <- ggarrange(heatmap_stabilizers, NULL, heatmap_decay, ncol =1, heights = c(15, -3, 7), align = "v")
readers_heatmap
ggarrange(readers_heatmap, Vln_readers, ncol = 2, widths = c(10, 3))
```

### Heatmap markers

```{r fig.width= 6, fig.height =5}
#Heatmap for cell lineage per model
DoHeatmap(Blast_integrated_copy, 
          features = cell_type_markers, 
          group.by = "cell_layers2", 
          size = 4, 
          angle = 45)+
  scale_fill_gradient2(mid = "gray30", 
                       low = "cornflowerblue", 
                       high = "coral")+  
  guides(color=FALSE)
```

```{r fig.width = 2, fig.height= 15}
#As an additional visualization method I want to make small violinplot of normmalized data to put beside my heatmaps. OF both readers as writers


#Same goes for every violinplot
Vln_IGF2BP1 <- VlnPlot(Blast_integrated_copy, group.by ="data_set", features = c("IGF2BP1"), pt.size = 0, cols = c("darkblue", "darkred"))+
  theme(text=element_text(color="black"),axis.text=element_text(color="white"))+
  NoLegend()+
  xlab(NULL)+#minimalistic as possible with no axis
  ylab(NULL)+
  ggtitle(NULL)

Vln_IGF2BP2 <- VlnPlot(Blast_integrated_copy, group.by ="data_set", features = c("IGF2BP2"), pt.size = 0, cols = c("darkblue", "darkred"))+
  theme(text=element_text(color="black"),axis.text=element_text(color="white"))+
  NoLegend()+
  xlab(NULL)+
  ylab(NULL)+
  ggtitle(NULL)

Vln_IGF2BP3 <- VlnPlot(Blast_integrated_copy, group.by ="data_set", features = c("IGF2BP3"), pt.size = 0, cols = c("darkblue", "darkred"))+
  theme(text=element_text(color="black"),axis.text=element_text(color="white"))+
  NoLegend()+
  xlab(NULL)+
  ylab(NULL)+
  ggtitle(NULL)

Vln_ELAVL1 <- VlnPlot(Blast_integrated_copy, group.by ="data_set", features = c("ELAVL1"), pt.size = 0, cols = c("darkblue", "darkred"))+
  theme(text=element_text(color="black"),axis.text=element_text(color="white"))+
  NoLegend()+
  xlab(NULL)+
  ylab(NULL)+
  ggtitle(NULL)

Vln_MATR3 <- VlnPlot(Blast_integrated_copy, group.by ="data_set", features = c("MATR3"), pt.size = 0, cols = c("darkblue", "darkred"))+
  theme(text=element_text(color="black"),axis.text=element_text(color="white"))+
  NoLegend()+
  xlab(NULL)+
  ylab(NULL)+
  ggtitle(NULL)

Vln_YTHDF2 <- VlnPlot(Blast_integrated_copy, group.by ="data_set", features = c("YTHDF2"), pt.size = 0, cols = c("darkblue", "darkred"))+
  theme(text=element_text(color="black"),axis.text=element_text(color="white"))+
  NoLegend()+
  xlab(NULL)+
  ylab(NULL)+
  ggtitle(NULL)

#Combine the plots
Vlns_stabilizers <- ggarrange(Vln_IGF2BP1, 
                              NULL, #ADd NULL as a way to control the size of the blank space between plots
                              Vln_IGF2BP2, 
                              NULL, 
                              Vln_IGF2BP3, 
                              NULL, 
                              Vln_ELAVL1, 
                              NULL, 
                              Vln_MATR3, 
                              NULL, 
                              ncol = 1, 
                              heights = c(3, -1, 3, -1, 3, -1, 3, -1, 3))
#Combine the violin stabilizers witht the decay
Vln_readers <- ggarrange(NULL, 
                         Vlns_stabilizers, 
                         NULL, 
                         Vln_YTHDF2, 
                         ncol = 1, 
                         heights = c(3, 15, -3, 3.5), 
                         widths = c(20, 3, 3, 3))
Vln_readers

```

```{r fig.width = 2, fig.height=15}

#Do the same for the writer complex subunits
Vln_METTL3 <- VlnPlot(Blast_integrated_copy, group.by ="data_set", features = c("METTL3"), pt.size = 0, cols = c("darkblue", "darkred"))+
  theme(text=element_text(color="black"),axis.text=element_text(color="white"))+
  NoLegend()+
  xlab(NULL)+
  ylab(NULL)+
  ggtitle(NULL)

Vln_METTL14 <- VlnPlot(Blast_integrated_copy, group.by ="data_set", features = c("METTL14"), pt.size = 0, cols = c("darkblue", "darkred"))+
  theme(text=element_text(color="black"),axis.text=element_text(color="white"))+
  NoLegend()+
  xlab(NULL)+
  ylab(NULL)+
  ggtitle(NULL)

Vln_VIRMA <- VlnPlot(Blast_integrated_copy, group.by ="data_set", features = c("VIRMA"), pt.size = 0, cols = c("darkblue", "darkred"))+
  theme(text=element_text(color="black"),axis.text=element_text(color="white"))+
  NoLegend()+
  xlab(NULL)+
  ylab(NULL)+
  ggtitle(NULL)

Vln_WTAP <- VlnPlot(Blast_integrated_copy, group.by ="data_set", features = c("WTAP"), pt.size = 0, cols = c("darkblue", "darkred"))+
  theme(text=element_text(color="black"),axis.text=element_text(color="white"))+
  NoLegend()+
  xlab(NULL)+
  ylab(NULL)+
  ggtitle(NULL)

Vln_ZC3H13 <- VlnPlot(Blast_integrated_copy, group.by ="data_set", features = c("ZC3H13"), pt.size = 0, cols = c("darkblue", "darkred"))+
  theme(text=element_text(color="black"),axis.text=element_text(color="white"))+
  NoLegend()+
  xlab(NULL)+
  ylab(NULL)+
  ggtitle(NULL)

Vln_CBLL1 <- VlnPlot(Blast_integrated_copy, group.by ="data_set", features = c("CBLL1"), pt.size = 0, cols = c("darkblue", "darkred"))+
  theme(text=element_text(color="black"),axis.text=element_text(color="white"))+
  NoLegend()+
  xlab(NULL)+
  ylab(NULL)+
  ggtitle(NULL)

#Combine the writer complex subunits
Vlns_writer_complex <- ggarrange(Vln_METTL3, 
                                 Vln_METTL14, 
                                 Vln_VIRMA, 
                                 Vln_WTAP, 
                                 Vln_ZC3H13, 
                                 Vln_CBLL1, 
                                 ncol =1)
Vlns_writer_complex
```

```{r PROGENy}
#Progeny can be used to assess the activation or inhibition of different pathways 
Idents(Blast_integrated_copy) <- "cell_layers2"
Blast_seurat_progeny_clusters <- data.frame(Cell = names(Idents(Blast_integrated_copy)), 
                                           CellType = as.character(Idents(Blast_integrated_copy)), 
                                           stringsAsFactors = F)

#Use PROGENy workflow as described in their vignette. 
Blast_seurat_progeny <- progeny(Blast_integrated_copy, 
                                       scale = F, 
                                       organism = "Human", 
                                       top = 500, 
                                       perm = 1, 
                                       return_assay = T)

Blast_seurat_progeny <- ScaleData(Blast_seurat_progeny, assay = "progeny") #Make a seperate assay in seurat

#Make a new dataframe for progeny
BCP_df <- as.data.frame(t(GetAssayData(Blast_seurat_progeny, slot = "data", assay = "progeny")))%>%
  rownames_to_column("Cell") %>%
  gather(Pathway, Activity, -Cell)

#Transform data according to vignette
BCP_df <- inner_join(BCP_df, Blast_seurat_progeny_clusters)
sum_bcp_df <- BCP_df %>%
  group_by(Pathway, CellType) %>%
  summarise(avg = mean(Activity), std=sd(Activity))

sum_bcp_df <- sum_bcp_df %>%
  dplyr::select(-std) %>%
  spread(Pathway, avg) %>%
  data.frame(row.names = 1, check.names = F, stringsAsFactors = F)

#Choose colors. Standard to PROGENy, can be altered if preferred
paletteLength = 100
myColor = colorRampPalette(c("cornflowerblue", "gray30","coral"))(paletteLength)

progenyBreaks = c(seq(min(sum_bcp_df), 0, 
                      length.out=ceiling(paletteLength/2) + 1),
                  seq(max(sum_bcp_df)/paletteLength, 
                      max(sum_bcp_df), 
                      length.out=floor(paletteLength/2)))
progeny_hmap = pheatmap(t(sum_bcp_df),fontsize=14, 
                        fontsize_row = 10, 
                        cluster_cols = F,
                        color=myColor, breaks = progenyBreaks, 
                        main = "Pathway analysis Blastoid vs. Blastocyst", angle_col = 0, fontsize_col = 10,                        treeheight_col = 0,  border_color = NA)


#Plot the progeny in complex heatmap as it is prettier than pretty heatmap

progeny_hmap <- ComplexHeatmap::pheatmap(t(sum_bcp_df),
                                          fontsize=14, 
                                          fontsize_row = 13, 
                                          cluster_cols = F,
                                          color=myColor, 
                                          breaks = progenyBreaks,
                                          main = "Pathway analysis cell types in blastoid and blastocyst", 
                                          fontsize_col = 13,   
                                          treeheight_col = 0,  
                                          border_color = "black",
                                          column_names_side = c("top"),
                                          angle_col = "45", 
                                          cellwidth = 80
                                          )
```

```{r VIRMA target module score}

#Comparison for one specific cell type that may influence VIRMA target module scores
# Adjust the comparisons as needed
comparisons_celltypes <- list( c("EPI (Blastocyst)", "EPI (Blastoid)"), c("PRE (Blastocyst)", "PRE (Blastoid)"), c("TE (Blastocyst)", "TE (Blastoid)")) #Make a list for comparisons to test with t-test or similar



#Make a plot comparing the cell types to one another
VIRMA_module_vln_cell_types <- VlnPlot(Blast_integrated_copy, 
                                       features = "VIRMA_features1", 
                                       split.by = "data_set", 
                                       group.by = "cell_layers2", 
                                       cols= c("darkblue", "darkred"))+
  #ggtitle("VIRMA features across cell types")+ 
  ggtitle("Cell types")+
  theme(plot.title = element_text(size=12))+
  NoLegend()+xlab(NULL)+
  stat_compare_means(comparisons = comparisons_celltypes, 
                     method = "wilcox.test", 
                     label = "p.signif")+ 
  stat_compare_means(label.y = 50)+ 
  ylim(-5,11)+
    ylab("Calculated VIRMA \nmodule score")
VIRMA_module_vln_cell_types


#Make a plot comparing the datasets to eachother
VIRMA_module_vln_ds <- VlnPlot(Blast_integrated_copy, 
                               features = "VIRMA_features1", 
                               group.by = "data_set", 
                               cols= c("darkblue", "darkred"))+
  #ggtitle("Difference in VIRMA \nfeatures between \nblastocyst and blastoid")+ 
  theme(plot.title = element_text(size=12))+
  NoLegend()+
  xlab(NULL)+
  stat_compare_means(comparisons = list(c("Blastocyst", "Blastoid")), 
                     method = "wilcox.test", 
                     label = "p.signif")+ 
  stat_compare_means(label.y = 50)+ 
  ylim(-5,10) + 
  ggtitle("Model")+
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5))+
  ylab("Calculated VIRMA \nmodule score")
VIRMA_module_vln_ds

# Adjust the comparisons as needed
comparisons_times <- list( c("S1", "S2"), c("S2", "S3"), c("Naive", "24h"),  c("24h", "60h"), c("60h", "96h"), c("S1", "96h"))


#Make a plot comparing the stages to eachother
VIRMA_module_vln_times <-  VlnPlot(Blast_integrated_copy, features = "VIRMA_features1", split.by = "data_set", group.by = "times", cols= c("darkblue", "darkred"))+
  #ggtitle("VIRMA features across time points")+
  ggtitle("Time points")+
  NoLegend()+
  xlab(NULL)+ 
  theme(plot.title = element_text(size=12)) + 
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5))+
  stat_compare_means(comparisons = comparisons_times, 
                     method = "wilcox.test", 
                     label = "p.signif")+ 
  stat_compare_means(label.y = 50)+ 
  ylim(-5,12) +
  ylab("Calculated VIRMA \nmodule score")

VIRMA_module_vln_times

#Combine the timeplot and celltype plot
vln_module_ct_times <- ggarrange(VIRMA_module_vln_times, 
                                 NULL, 
                                 VIRMA_module_vln_cell_types,
                                 ncol =1, 
                                 heights = c(40, 5, 50), 
                                 widths = c(5,5, 5))

vln_module_ct_times


#Combine the combined time/stage plot with the model plot
vln_module_ct_times_ds <- ggarrange(VIRMA_module_vln_ds, 
                                    vln_module_ct_times, 
                                    ncol =2, 
                                    heights = c(200, 200), 
                                    widths = c(150, 300))

vln_module_ct_times_ds
```

### Heatmap potential transcription factors

```{r}
#Make heatmap about the most downregulated TFs in blastoid that could regulate VIRMA
DoHeatmap(Blast_integrated_copy, 
          features = TF_VIRMA, 
          slot = "data", 
          group.by = "cell_layers2", 
          size = 2, 
          angle = 45, 
          group.bar = T)+
  scale_fill_gradient2(mid = "gray30", low = "cornflowerblue", high = "coral")+  guides(color=FALSE)
```

## Volcanoplots

```{r}
#PLot volcanoplots of DEGs together for a figure
DEG_plot_te | DEG_plot_icm | DEG_plot_pre | DEG_plot_all

#Plot together via ggarrange, combine them
DEG_plot_panels <- ggarrange(DEG_plot_te + theme_classic(base_size = 13),
                             DEG_plot_icm + theme_classic(base_size = 13),
                             DEG_plot_pre + theme_classic(base_size = 13),
                             DEG_plot_all + theme_classic(base_size = 13))
DEG_plot_panels
```

### Fold enrichment

```{r}
FEV <- readxl::read_excel("data/FoldEnrichmentVIRMA.xlsx") #Data determined from David database about fold enrichment

#FEV$`Cell type` <- factor(FEV$`Cell type`, levels = c("ALL", "PRE", "EPI", "TE"))
FEV$`Cell type` <- factor(FEV$`Cell type`, levels = c("Primitive Endoderm", "Epiblast", "Trophectoderm",  "All celltypes")) #Order the axis


FEV_plot <- ggplot(FEV, 
                   aes(x = FEV$"Fold Enrichment", 
                       y= FEV$"Cell type", 
                       fill =FEV$`-log10(FDR)`))+
  geom_col(width = c(.4, .4, .4, 1), #make only the ' all ' column bigger for extra salience
           col = "black")+
  #coord_flip()+
  xlab("Fold Enrichment")+
  ylab(NULL)+
  scale_fill_gradient(low = "oldlace", 
                      high = "darkred", 
                      name = "-log10(FDR)", 
                      limits = c(0,200))+
  theme_classic(base_size = 15)+
  #scale_y_continuous(expand = c(0, 0))+
  scale_x_continuous(expand = c(0, 0), 
                     limits = c(0, 2.5))+
  theme(axis.text.y = element_text(size = c(15, 15, 15, 20), 
                                   angle = 0, 
                                   vjust = 0.5, 
                                   hjust=1), 
        axis.title.x=element_text(size=15))+
  theme(legend.position = "right")+
  theme(legend.title = element_text(size = 15), 
        legend.text = element_text(size = 10), 
        legend.key.size = unit(0.3, "cm"))+
  ggtitle("BioGRID enrichment for \nVIRMA RNA targets")
FEV_plot

#Combine the VIRMA volcanoplots together for a pretty plot
virma_plots_arranged <- ggarrange(DEG_plot_te_VIRMA+theme_classic(base_size = 15), 
                                  DEG_plot_icm_VIRMA+theme_classic(base_size = 15), 
                                  DEG_plot_pre_VIRMA+theme_classic(base_size = 15), 
                                  DEG_plot_all_VIRMA+theme_classic(base_size = 15), ncol = 4)
virma_plots_arranged
```

# Specific plot for Klaas

```{r}
#klaas wanted figures containing only S2 and 96h violinplots.
Idents(Blast_integrated_copy) <- "times"
#Substract only 96h and S2
Blast_integrated_S2_96h <- subset(Blast_integrated_copy, idents = c("96h", "S2"))
Idents(Blast_integrated_S2_96h) <- "cell_layers2"
#Delete naive cells from this structure
Blast_integrated_S2_96h <- subset(Blast_integrated_S2_96h, idents = c("Naive (Blastoid)"), invert = TRUE)

#Find Downregulated VIRMA targets
downregulated_virma_targets <- intersect(DEGs_ALL_df$Blastocyst, VIRMA_binding_RNAs)


#Plot in heatmap
DoHeatmap(Blast_integrated_S2_96h, 
          features= downregulated_virma_targets, 
          slot = "data", 
          size = 3, 
          angle = 45, 
          group.bar = T)+
  scale_fill_gradient2(mid = "white", low = "black", high = "red", midpoint = 3)+ 
  guides(color=FALSE)+
  theme(axis.text.y = element_text(size = 0))
```

```{r}
#Klaas wanted to know about seemingly two groups present in the EPI when looking at module score. 
Blast_integrated_S2_96h_dg <- Blast_integrated_S2_96h[downregulated_virma_targets,]

#Add module score to divide data later by module score
Blast_integrated_S2_96h <- AddModuleScore(
 object = Blast_integrated_S2_96h,
 features = downregulated_virma_targets, #Select only RNAs present in Seurat object
 name = "Downregulated_VIRMA_features"
)

#Plot in violinplot to check
VlnPlot(Blast_integrated_S2_96h, 
        features = cell_type_markers, 
        group.by = "cell_layers2", 
        split.by = "data_set", 
        cols = c("darkblue", "darkred"))+
  NoLegend()

```

```{r}
comparisons_phases <- list( c("G1", "G2M"), c("G1", "S"), c("G2M", "S"))

list_phase2 <- NULL

for (i in 1:length(Blast_integrated_copy@meta.data$Phase)){  if (Blast_integrated_copy@meta.data$Phase[i] == "S"){ if (Blast_integrated_copy@meta.data$data_set[i] == "Blastocyst"){  list_phase2 <- append(list_phase2, "S (Blastocyst)")}  else {list_phase2 <- append(list_phase2, "S (Blastoid)")}  }  if (Blast_integrated_copy@meta.data$Phase[i] == "G1"){ if (Blast_integrated_copy@meta.data$data_set[i] == "Blastocyst"){  list_phase2 <- append(list_phase2, "G1 (Blastocyst))")}  else {list_phase2 <- append(list_phase2, "G1 (Blastoid)")}  }  if (Blast_integrated_copy@meta.data$Phase[i] == "G2M"){ if (Blast_integrated_copy@meta.data$data_set[i] == "Blastocyst"){ list_phase2 <- append(list_phase2, "G2M (Blastocyst)")} else {list_phase2 <- append(list_phase2, "G2M (Blastoid)")}} }

print(list_phase2) Blast_integrated_copy@meta.data$Phase2 <- list_phase2

VIRMA_module_phase <- VlnPlot(Blast_integrated_copy, features = "VIRMA_features1", group.by = "data_set", cols= c("darkblue", "darkred"))+ 
  ggtitle("VIRMA features across time points")+ 
  ggtitle(NULL)+ 
  NoLegend()+
  xlab(NULL)+ 
  theme(plot.title = element_text(size=12)) +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5))+ 
  stat_compare_means(method = "anova", label = "p.signif")+ 
  stat_compare_means(label.y = 50)+ 
  facet_wrap(~Blast_integrated_copy@meta.data$Phase)+ 
  ylim(-5,12) + 
  ylab("Calculated VIRMA \nmodule score")

VIRMA_module_phase
Blast_integrated_copy$Phase <- factor(Blast_integrated_copy$Phase, levels = c("G2M", "S", "G1"))
VIRMA_module_phase <- VlnPlot(Blast_integrated_copy, features = "VIRMA_features1", group.by = "Phase")+ 
  ggtitle("VIRMA features across time points")+ 
  ggtitle(NULL)+ 
  NoLegend()+
  xlab(NULL)+ 
  theme(plot.title = element_text(size=12)) +
  theme(axis.text.x = element_text(angle = 0, hjust = 0.5))+ 
  stat_compare_means(method = "anova", label = "p.signif")+ 
  stat_compare_means(label.y = 50)+ 
  facet_wrap(~Blast_integrated_copy@meta.data$data_set)+ 
  ylab("Calculated VIRMA \nmodule score")

VIRMA_module_phase
```

## Stress granules

```{r fig.width = 3, fig.height= 6}
stress_granule_matrix_down <- matrix(c(593,991,1794,14524), nrow = 2, ncol =2)
fisher_down <- fisher.test(stress_granule_matrix_down)

stress_granule_matrix_up <- matrix(c(19,1571,296,16022), nrow = 2, ncol =2)
fisher_up <- fisher.test(stress_granule_matrix_up)


stress_granule_matrix_total <- data.frame(c((1584/17902*100), (593/2387*100), (13/309*100)), c("Genome", "Downregulated", "Upregulated"), ncol = 2, nrow = 3)
colnames(stress_granule_matrix_total) <- c("Relative amount of genes", "Category")

SG_plot <- ggplot(stress_granule_matrix_total, aes(x= factor(Category, c("Genome", "Downregulated", "Upregulated")), y= `Relative amount of genes`))+
  geom_col(width = 0.5, fill = c("grey", "darkblue", "darkred"))+
  geom_text(aes(label = c(NA, "<2.2e16","0.088")), vjust = -1, size = 3)+
  geom_text(aes(label = c(NA, "p-value =","p-value =")), vjust = -2.5, size = 3)+
  scale_y_continuous(expand = c(0, 0), limits = c(0, 30))+
  ylab("Relative amount of genes \nin stress granule(%)")+
  xlab(NULL)+
  theme_classic(base_size = 15)
```

```{r}
#Plot other stress granule markers
VlnPlot(Blast_integrated_copy, 
        features = c("TIA1", "G3BP1", "TIAL1"), 
        cols= c("darkblue", "darkred"), 
        group.by = "cell_layers2", 
        split.by = "data_set")
```

# The END!

```{r}

#Thanks for reading all the way to the end.
#The conclusion of this research project is that the blastoid has one major limitation, the downregulation of VIRMA targets. This is seen in the VIRMA targets itself, but also with the writer complex and the readers.


#Will be continued.....

```
